{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "63c8357d-aca0-43ac-ad76-d8954b6d133d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1b91bd6f-b591-40e3-ad7e-c4d43ca38fa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['BC60', 'BC66', 'BC70', 'BLC1', 'CC107', 'CC127-2', 'CC18', 'CC46', 'CC5', 'CC82', 'CC94', 'CC96', 'CC99', 'GC39', 'GC41', 'LS1', 'LS6', 'M14', 'M21', 'M22', 'NC5', 'OC14', 'OC33', 'OC40', 'OC42', 'OC48', 'OC50', 'OC54', 'OT1', 'PC12', 'PC14', 'PC21', 'PC32', 'SkC6', 'StS1', 'UC2']\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/mnt/c/Users/Komar/Lab/OutPutDir/LINE/AGCT/')\n",
    "print(os.listdir())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "48d2948a-2b26-472a-bcab-aa6d7dbc20ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC60 ---------------------------------------------------\n",
      "metatable shape: (1907, 32)\n",
      "sample file names: ['BC60-L_S106_ME_L001__', 'MN-BC60-L_S108_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 828   including test: 0\n",
      "somatic and quasi-somatic: (1079, 32)   including test: 0 \n",
      "\n",
      "BC60-L_S106_ME_L001__ 441   test: 0\n",
      "MN-BC60-L_S108_ME_L001__ 638   test: 0\n",
      "overall num of somatic insertions:  1079 \n",
      "\n",
      "BC60-L_S106_ME_L001__  was: 441  become: 36\n",
      "BC60-L_S106_ME_L001__ reads were written\n",
      "MN-BC60-L_S108_ME_L001__  was: 638  become: 71\n",
      "MN-BC60-L_S108_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC60-L_S106_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.02s/it]\n",
      "BC60-L_S106_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.55s/it]\n",
      "MN-BC60-L_S108_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 7/7 [00:19<00:00,  2.81s/it]\n",
      "MN-BC60-L_S108_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 7/7 [00:25<00:00,  3.63s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:14<00:00,  7.18s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 83\n",
      "unique names: 23\n",
      "after: 35 \n",
      "\n",
      "before: 251\n",
      "unique names: 35\n",
      "after: 61 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC60-L_S106_ME_L001__: 100%|████████████████████████████████████████████████████████████| 35/35 [05:38<00:00,  9.68s/it]\n",
      "MN-BC60-L_S108_ME_L001__: 100%|█████████████████████████████████████████████████████████| 61/61 [10:22<00:00, 10.20s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BC66 ---------------------------------------------------\n",
      "metatable shape: (1927, 32)\n",
      "sample file names: ['BC66-L_S102_ME_L001__', 'MN-BC66-L_S104_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 842   including test: 0\n",
      "somatic and quasi-somatic: (1085, 32)   including test: 0 \n",
      "\n",
      "BC66-L_S102_ME_L001__ 382   test: 0\n",
      "MN-BC66-L_S104_ME_L001__ 703   test: 0\n",
      "overall num of somatic insertions:  1085 \n",
      "\n",
      "BC66-L_S102_ME_L001__  was: 382  become: 36\n",
      "BC66-L_S102_ME_L001__ reads were written\n",
      "MN-BC66-L_S104_ME_L001__  was: 703  become: 68\n",
      "MN-BC66-L_S104_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC66-L_S102_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.72s/it]\n",
      "BC66-L_S102_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.63s/it]\n",
      "MN-BC66-L_S104_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 6/6 [00:17<00:00,  2.96s/it]\n",
      "MN-BC66-L_S104_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 6/6 [00:22<00:00,  3.83s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:11<00:00,  5.94s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 112\n",
      "unique names: 21\n",
      "after: 34 \n",
      "\n",
      "before: 82\n",
      "unique names: 33\n",
      "after: 52 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC66-L_S102_ME_L001__: 100%|████████████████████████████████████████████████████████████| 34/34 [04:12<00:00,  7.43s/it]\n",
      "MN-BC66-L_S104_ME_L001__: 100%|█████████████████████████████████████████████████████████| 52/52 [06:39<00:00,  7.68s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BC70 ---------------------------------------------------\n",
      "metatable shape: (1139, 32)\n",
      "sample file names: ['BC70-L_S142_ME_L001__', 'MN-BC70-L_S144_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 178   including test: 0\n",
      "somatic and quasi-somatic: (961, 32)   including test: 0 \n",
      "\n",
      "BC70-L_S142_ME_L001__ 197   test: 0\n",
      "MN-BC70-L_S144_ME_L001__ 764   test: 0\n",
      "overall num of somatic insertions:  961 \n",
      "\n",
      "BC70-L_S142_ME_L001__  was: 197  become: 26\n",
      "BC70-L_S142_ME_L001__ reads were written\n",
      "MN-BC70-L_S144_ME_L001__  was: 764  become: 45\n",
      "MN-BC70-L_S144_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC70-L_S142_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.08s/it]\n",
      "BC70-L_S142_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.33s/it]\n",
      "MN-BC70-L_S144_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.37s/it]\n",
      "MN-BC70-L_S144_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 4/4 [00:13<00:00,  3.29s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.00s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 187\n",
      "unique names: 18\n",
      "after: 48 \n",
      "\n",
      "before: 132\n",
      "unique names: 35\n",
      "after: 42 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC70-L_S142_ME_L001__: 100%|████████████████████████████████████████████████████████████| 48/48 [06:34<00:00,  8.21s/it]\n",
      "MN-BC70-L_S144_ME_L001__: 100%|█████████████████████████████████████████████████████████| 42/42 [06:31<00:00,  9.32s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BLC1 ---------------------------------------------------\n",
      "metatable shape: (1684, 32)\n",
      "sample file names: ['BLC1-L_S74_ME_L001__', 'MN-BLC1-L_S76_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 749   including test: 0\n",
      "somatic and quasi-somatic: (935, 32)   including test: 0 \n",
      "\n",
      "BLC1-L_S74_ME_L001__ 634   test: 0\n",
      "MN-BLC1-L_S76_ME_L001__ 301   test: 0\n",
      "overall num of somatic insertions:  935 \n",
      "\n",
      "BLC1-L_S74_ME_L001__  was: 634  become: 69\n",
      "BLC1-L_S74_ME_L001__ reads were written\n",
      "MN-BLC1-L_S76_ME_L001__  was: 301  become: 19\n",
      "MN-BLC1-L_S76_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLC1-L_S74_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:16<00:00,  2.82s/it]\n",
      "BLC1-L_S74_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:22<00:00,  3.71s/it]\n",
      "MN-BLC1-L_S76_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.43s/it]\n",
      "MN-BLC1-L_S76_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.32s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 281\n",
      "unique names: 31\n",
      "after: 57 \n",
      "\n",
      "before: 315\n",
      "unique names: 13\n",
      "after: 27 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLC1-L_S74_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 57/57 [08:10<00:00,  8.61s/it]\n",
      "MN-BLC1-L_S76_ME_L001__: 100%|██████████████████████████████████████████████████████████| 27/27 [02:47<00:00,  6.19s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC107 ---------------------------------------------------\n",
      "metatable shape: (1337, 32)\n",
      "sample file names: ['CC107-L_S130_ME_L001__', 'MN-CC107-L_S132_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 573   including test: 0\n",
      "somatic and quasi-somatic: (764, 32)   including test: 0 \n",
      "\n",
      "CC107-L_S130_ME_L001__ 417   test: 0\n",
      "MN-CC107-L_S132_ME_L001__ 347   test: 0\n",
      "overall num of somatic insertions:  764 \n",
      "\n",
      "CC107-L_S130_ME_L001__  was: 417  become: 34\n",
      "CC107-L_S130_ME_L001__ reads were written\n",
      "MN-CC107-L_S132_ME_L001__  was: 347  become: 25\n",
      "MN-CC107-L_S132_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC107-L_S130_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.49s/it]\n",
      "CC107-L_S130_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.36s/it]\n",
      "MN-CC107-L_S132_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.56s/it]\n",
      "MN-CC107-L_S132_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.84s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.88s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 85\n",
      "unique names: 24\n",
      "after: 35 \n",
      "\n",
      "before: 141\n",
      "unique names: 15\n",
      "after: 25 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC107-L_S130_ME_L001__: 100%|███████████████████████████████████████████████████████████| 35/35 [05:51<00:00, 10.03s/it]\n",
      "MN-CC107-L_S132_ME_L001__: 100%|████████████████████████████████████████████████████████| 25/25 [03:24<00:00,  8.18s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC127-2 ---------------------------------------------------\n",
      "metatable shape: (1724, 32)\n",
      "sample file names: ['CC127-2-L_S6_ME_L001__', 'MN-CC127-L_S8_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 593   including test: 0\n",
      "somatic and quasi-somatic: (1131, 32)   including test: 0 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__ 802   test: 0\n",
      "MN-CC127-L_S8_ME_L001__ 329   test: 0\n",
      "overall num of somatic insertions:  1131 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__  was: 802  become: 106\n",
      "CC127-2-L_S6_ME_L001__ reads were written\n",
      "MN-CC127-L_S8_ME_L001__  was: 329  become: 36\n",
      "MN-CC127-L_S8_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC127-2-L_S6_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 10/10 [00:29<00:00,  2.90s/it]\n",
      "CC127-2-L_S6_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 10/10 [00:37<00:00,  3.75s/it]\n",
      "MN-CC127-L_S8_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.88s/it]\n",
      "MN-CC127-L_S8_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:11<00:00,  3.78s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:23<00:00, 11.53s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 494\n",
      "unique names: 67\n",
      "after: 121 \n",
      "\n",
      "before: 227\n",
      "unique names: 22\n",
      "after: 47 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC127-2-L_S6_ME_L001__: 100%|█████████████████████████████████████████████████████████| 121/121 [18:19<00:00,  9.08s/it]\n",
      "MN-CC127-L_S8_ME_L001__: 100%|██████████████████████████████████████████████████████████| 47/47 [04:55<00:00,  6.28s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC18 ---------------------------------------------------\n",
      "metatable shape: (1557, 32)\n",
      "sample file names: ['CC18-L_S118_ME_L001__', 'MN-CC18-L_S120_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 646   including test: 0\n",
      "somatic and quasi-somatic: (911, 32)   including test: 0 \n",
      "\n",
      "CC18-L_S118_ME_L001__ 385   test: 0\n",
      "MN-CC18-L_S120_ME_L001__ 526   test: 0\n",
      "overall num of somatic insertions:  911 \n",
      "\n",
      "CC18-L_S118_ME_L001__  was: 385  become: 31\n",
      "CC18-L_S118_ME_L001__ reads were written\n",
      "MN-CC18-L_S120_ME_L001__  was: 526  become: 60\n",
      "MN-CC18-L_S120_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC18-L_S118_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.64s/it]\n",
      "CC18-L_S118_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.38s/it]\n",
      "MN-CC18-L_S120_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.53s/it]\n",
      "MN-CC18-L_S120_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 6/6 [00:20<00:00,  3.39s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.37s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 48\n",
      "unique names: 17\n",
      "after: 30 \n",
      "\n",
      "before: 117\n",
      "unique names: 41\n",
      "after: 57 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC18-L_S118_ME_L001__: 100%|████████████████████████████████████████████████████████████| 30/30 [03:50<00:00,  7.68s/it]\n",
      "MN-CC18-L_S120_ME_L001__: 100%|█████████████████████████████████████████████████████████| 57/57 [09:57<00:00, 10.49s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC46 ---------------------------------------------------\n",
      "metatable shape: (1448, 32)\n",
      "sample file names: ['CC46-L_S134_ME_L001__', 'MN-CC46-L_S136_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 541   including test: 0\n",
      "somatic and quasi-somatic: (907, 32)   including test: 0 \n",
      "\n",
      "CC46-L_S134_ME_L001__ 422   test: 0\n",
      "MN-CC46-L_S136_ME_L001__ 485   test: 0\n",
      "overall num of somatic insertions:  907 \n",
      "\n",
      "CC46-L_S134_ME_L001__  was: 422  become: 28\n",
      "CC46-L_S134_ME_L001__ reads were written\n",
      "MN-CC46-L_S136_ME_L001__  was: 485  become: 40\n",
      "MN-CC46-L_S136_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC46-L_S134_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.06s/it]\n",
      "CC46-L_S134_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.99s/it]\n",
      "MN-CC46-L_S136_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.19s/it]\n",
      "MN-CC46-L_S136_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 4/4 [00:16<00:00,  4.23s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.86s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 387\n",
      "unique names: 17\n",
      "after: 46 \n",
      "\n",
      "before: 56\n",
      "unique names: 24\n",
      "after: 32 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC46-L_S134_ME_L001__: 100%|████████████████████████████████████████████████████████████| 46/46 [05:05<00:00,  6.64s/it]\n",
      "MN-CC46-L_S136_ME_L001__: 100%|█████████████████████████████████████████████████████████| 32/32 [04:48<00:00,  9.02s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC5 ---------------------------------------------------\n",
      "metatable shape: (1677, 32)\n",
      "sample file names: ['CC5-L_S146_ME_L001__', 'MN-CC5-L_S148_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 600   including test: 0\n",
      "somatic and quasi-somatic: (1077, 32)   including test: 0 \n",
      "\n",
      "CC5-L_S146_ME_L001__ 427   test: 0\n",
      "MN-CC5-L_S148_ME_L001__ 650   test: 0\n",
      "overall num of somatic insertions:  1077 \n",
      "\n",
      "CC5-L_S146_ME_L001__  was: 427  become: 31\n",
      "CC5-L_S146_ME_L001__ reads were written\n",
      "MN-CC5-L_S148_ME_L001__  was: 650  become: 57\n",
      "MN-CC5-L_S148_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC5-L_S146_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.68s/it]\n",
      "CC5-L_S146_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.41s/it]\n",
      "MN-CC5-L_S148_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.28s/it]\n",
      "MN-CC5-L_S148_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 5/5 [00:15<00:00,  3.18s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.33s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 206\n",
      "unique names: 20\n",
      "after: 42 \n",
      "\n",
      "before: 51\n",
      "unique names: 29\n",
      "after: 41 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC5-L_S146_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 42/42 [05:30<00:00,  7.87s/it]\n",
      "MN-CC5-L_S148_ME_L001__: 100%|██████████████████████████████████████████████████████████| 41/41 [05:42<00:00,  8.36s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC82 ---------------------------------------------------\n",
      "metatable shape: (2129, 32)\n",
      "sample file names: ['CC82-L_S10_ME_L001__', 'MN-CC82-L_S12_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 690   including test: 0\n",
      "somatic and quasi-somatic: (1439, 32)   including test: 0 \n",
      "\n",
      "CC82-L_S10_ME_L001__ 1051   test: 0\n",
      "MN-CC82-L_S12_ME_L001__ 388   test: 0\n",
      "overall num of somatic insertions:  1439 \n",
      "\n",
      "CC82-L_S10_ME_L001__  was: 1051  become: 101\n",
      "CC82-L_S10_ME_L001__ reads were written\n",
      "MN-CC82-L_S12_ME_L001__  was: 388  become: 38\n",
      "MN-CC82-L_S12_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC82-L_S10_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 10/10 [00:26<00:00,  2.66s/it]\n",
      "CC82-L_S10_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 10/10 [00:33<00:00,  3.37s/it]\n",
      "MN-CC82-L_S12_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.44s/it]\n",
      "MN-CC82-L_S12_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.29s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:23<00:00, 11.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 160\n",
      "unique names: 59\n",
      "after: 89 \n",
      "\n",
      "before: 282\n",
      "unique names: 22\n",
      "after: 50 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC82-L_S10_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 89/89 [15:03<00:00, 10.15s/it]\n",
      "MN-CC82-L_S12_ME_L001__: 100%|██████████████████████████████████████████████████████████| 50/50 [06:15<00:00,  7.51s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC94 ---------------------------------------------------\n",
      "metatable shape: (1011, 32)\n",
      "sample file names: ['CC94-L_S14_ME_L001__', 'MN-CC94-L_S16_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 265   including test: 0\n",
      "somatic and quasi-somatic: (746, 32)   including test: 0 \n",
      "\n",
      "CC94-L_S14_ME_L001__ 644   test: 0\n",
      "MN-CC94-L_S16_ME_L001__ 102   test: 0\n",
      "overall num of somatic insertions:  746 \n",
      "\n",
      "CC94-L_S14_ME_L001__  was: 644  become: 56\n",
      "CC94-L_S14_ME_L001__ reads were written\n",
      "MN-CC94-L_S16_ME_L001__  was: 102  become: 12\n",
      "MN-CC94-L_S16_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC94-L_S14_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:12<00:00,  2.52s/it]\n",
      "CC94-L_S14_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:18<00:00,  3.76s/it]\n",
      "MN-CC94-L_S16_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.75s/it]\n",
      "MN-CC94-L_S16_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  4.00s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:10<00:00,  5.50s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 120\n",
      "unique names: 28\n",
      "after: 45 \n",
      "\n",
      "before: 273\n",
      "unique names: 7\n",
      "after: 27 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC94-L_S14_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 45/45 [06:31<00:00,  8.71s/it]\n",
      "MN-CC94-L_S16_ME_L001__: 100%|██████████████████████████████████████████████████████████| 27/27 [02:48<00:00,  6.25s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC96 ---------------------------------------------------\n",
      "metatable shape: (2429, 32)\n",
      "sample file names: ['CC96-L_S66_ME_L001__', 'MN-CC96-L_S68_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 955   including test: 0\n",
      "somatic and quasi-somatic: (1474, 32)   including test: 0 \n",
      "\n",
      "CC96-L_S66_ME_L001__ 596   test: 0\n",
      "MN-CC96-L_S68_ME_L001__ 878   test: 0\n",
      "overall num of somatic insertions:  1474 \n",
      "\n",
      "CC96-L_S66_ME_L001__  was: 596  become: 66\n",
      "CC96-L_S66_ME_L001__ reads were written\n",
      "MN-CC96-L_S68_ME_L001__  was: 878  become: 143\n",
      "MN-CC96-L_S68_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC96-L_S66_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:16<00:00,  2.76s/it]\n",
      "CC96-L_S66_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:22<00:00,  3.69s/it]\n",
      "MN-CC96-L_S68_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 14/14 [00:39<00:00,  2.82s/it]\n",
      "MN-CC96-L_S68_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 14/14 [00:50<00:00,  3.59s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:22<00:00, 11.23s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 51\n",
      "unique names: 36\n",
      "after: 47 \n",
      "\n",
      "before: 332\n",
      "unique names: 96\n",
      "after: 128 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC96-L_S66_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 47/47 [06:58<00:00,  8.91s/it]\n",
      "MN-CC96-L_S68_ME_L001__: 100%|████████████████████████████████████████████████████████| 128/128 [22:23<00:00, 10.50s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC99 ---------------------------------------------------\n",
      "metatable shape: (1338, 32)\n",
      "sample file names: ['CC99-L_S70_ME_L001__', 'MN-CC99-L_S72_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 551   including test: 0\n",
      "somatic and quasi-somatic: (787, 32)   including test: 0 \n",
      "\n",
      "CC99-L_S70_ME_L001__ 459   test: 0\n",
      "MN-CC99-L_S72_ME_L001__ 328   test: 0\n",
      "overall num of somatic insertions:  787 \n",
      "\n",
      "CC99-L_S70_ME_L001__  was: 459  become: 43\n",
      "CC99-L_S70_ME_L001__ reads were written\n",
      "MN-CC99-L_S72_ME_L001__  was: 328  become: 26\n",
      "MN-CC99-L_S72_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC99-L_S70_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.32s/it]\n",
      "CC99-L_S70_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 4/4 [00:13<00:00,  3.26s/it]\n",
      "MN-CC99-L_S72_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.46s/it]\n",
      "MN-CC99-L_S72_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.86s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.43s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 130\n",
      "unique names: 25\n",
      "after: 38 \n",
      "\n",
      "before: 140\n",
      "unique names: 11\n",
      "after: 26 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC99-L_S70_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 38/38 [05:19<00:00,  8.41s/it]\n",
      "MN-CC99-L_S72_ME_L001__: 100%|██████████████████████████████████████████████████████████| 26/26 [03:12<00:00,  7.39s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GC39 ---------------------------------------------------\n",
      "metatable shape: (1935, 32)\n",
      "sample file names: ['GC39-L_S62_ME_L001__', 'MN-GC39-L_S64_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 889   including test: 0\n",
      "somatic and quasi-somatic: (1046, 32)   including test: 0 \n",
      "\n",
      "GC39-L_S62_ME_L001__ 524   test: 0\n",
      "MN-GC39-L_S64_ME_L001__ 522   test: 0\n",
      "overall num of somatic insertions:  1046 \n",
      "\n",
      "GC39-L_S62_ME_L001__  was: 524  become: 65\n",
      "GC39-L_S62_ME_L001__ reads were written\n",
      "MN-GC39-L_S64_ME_L001__  was: 522  become: 56\n",
      "MN-GC39-L_S64_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC39-L_S62_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.22s/it]\n",
      "GC39-L_S62_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:18<00:00,  3.05s/it]\n",
      "MN-GC39-L_S64_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 5/5 [00:12<00:00,  2.58s/it]\n",
      "MN-GC39-L_S64_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 5/5 [00:17<00:00,  3.60s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.98s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 355\n",
      "unique names: 25\n",
      "after: 51 \n",
      "\n",
      "before: 95\n",
      "unique names: 37\n",
      "after: 55 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC39-L_S62_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 51/51 [07:16<00:00,  8.56s/it]\n",
      "MN-GC39-L_S64_ME_L001__: 100%|██████████████████████████████████████████████████████████| 55/55 [06:52<00:00,  7.51s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GC41 ---------------------------------------------------\n",
      "metatable shape: (1578, 32)\n",
      "sample file names: ['GC41-L_S154_ME_L001__', 'MN-GC41-L_S156_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 738   including test: 0\n",
      "somatic and quasi-somatic: (840, 32)   including test: 0 \n",
      "\n",
      "GC41-L_S154_ME_L001__ 397   test: 0\n",
      "MN-GC41-L_S156_ME_L001__ 443   test: 0\n",
      "overall num of somatic insertions:  840 \n",
      "\n",
      "GC41-L_S154_ME_L001__  was: 397  become: 35\n",
      "GC41-L_S154_ME_L001__ reads were written\n",
      "MN-GC41-L_S156_ME_L001__  was: 443  become: 35\n",
      "MN-GC41-L_S156_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC41-L_S154_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.25s/it]\n",
      "GC41-L_S154_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.17s/it]\n",
      "MN-GC41-L_S156_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.41s/it]\n",
      "MN-GC41-L_S156_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.13s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.67s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 148\n",
      "unique names: 18\n",
      "after: 36 \n",
      "\n",
      "before: 38\n",
      "unique names: 21\n",
      "after: 28 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC41-L_S154_ME_L001__: 100%|████████████████████████████████████████████████████████████| 36/36 [04:08<00:00,  6.90s/it]\n",
      "MN-GC41-L_S156_ME_L001__: 100%|█████████████████████████████████████████████████████████| 28/28 [03:38<00:00,  7.81s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LS1 ---------------------------------------------------\n",
      "metatable shape: (698, 32)\n",
      "sample file names: ['LS1-L_S138_ME_L001__', 'MN-LS1-L_S140_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 71   including test: 0\n",
      "somatic and quasi-somatic: (627, 32)   including test: 0 \n",
      "\n",
      "LS1-L_S138_ME_L001__ 559   test: 0\n",
      "MN-LS1-L_S140_ME_L001__ 68   test: 0\n",
      "overall num of somatic insertions:  627 \n",
      "\n",
      "LS1-L_S138_ME_L001__  was: 559  become: 30\n",
      "LS1-L_S138_ME_L001__ reads were written\n",
      "MN-LS1-L_S140_ME_L001__  was: 68  become: 7\n",
      "MN-LS1-L_S140_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS1-L_S138_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:05<00:00,  1.98s/it]\n",
      "LS1-L_S138_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.71s/it]\n",
      "MN-LS1-L_S140_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "MN-LS1-L_S140_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.59s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 86\n",
      "unique names: 22\n",
      "after: 38 \n",
      "\n",
      "before: 82\n",
      "unique names: 5\n",
      "after: 13 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS1-L_S138_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 38/38 [04:34<00:00,  7.22s/it]\n",
      "MN-LS1-L_S140_ME_L001__: 100%|██████████████████████████████████████████████████████████| 13/13 [01:13<00:00,  5.62s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LS6 ---------------------------------------------------\n",
      "metatable shape: (1556, 32)\n",
      "sample file names: ['LS6-L_S46_ME_L001__', 'MN-LS6-L_S48_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 691   including test: 0\n",
      "somatic and quasi-somatic: (865, 32)   including test: 0 \n",
      "\n",
      "LS6-L_S46_ME_L001__ 551   test: 0\n",
      "MN-LS6-L_S48_ME_L001__ 314   test: 0\n",
      "overall num of somatic insertions:  865 \n",
      "\n",
      "LS6-L_S46_ME_L001__  was: 551  become: 52\n",
      "LS6-L_S46_ME_L001__ reads were written\n",
      "MN-LS6-L_S48_ME_L001__  was: 314  become: 32\n",
      "MN-LS6-L_S48_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS6-L_S46_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.15s/it]\n",
      "LS6-L_S46_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 5/5 [00:14<00:00,  2.89s/it]\n",
      "MN-LS6-L_S48_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.42s/it]\n",
      "MN-LS6-L_S48_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.16s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.79s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 157\n",
      "unique names: 28\n",
      "after: 52 \n",
      "\n",
      "before: 29\n",
      "unique names: 14\n",
      "after: 22 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS6-L_S46_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 52/52 [06:54<00:00,  7.98s/it]\n",
      "MN-LS6-L_S48_ME_L001__: 100%|███████████████████████████████████████████████████████████| 22/22 [02:33<00:00,  6.99s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M14 ---------------------------------------------------\n",
      "metatable shape: (1812, 32)\n",
      "sample file names: ['M14-L_S94_ME_L001__', 'MN-M14-L_S96_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 822   including test: 0\n",
      "somatic and quasi-somatic: (990, 32)   including test: 0 \n",
      "\n",
      "M14-L_S94_ME_L001__ 421   test: 0\n",
      "MN-M14-L_S96_ME_L001__ 569   test: 0\n",
      "overall num of somatic insertions:  990 \n",
      "\n",
      "M14-L_S94_ME_L001__  was: 421  become: 30\n",
      "M14-L_S94_ME_L001__ reads were written\n",
      "MN-M14-L_S96_ME_L001__  was: 569  become: 49\n",
      "MN-M14-L_S96_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M14-L_S94_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.52s/it]\n",
      "M14-L_S94_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.22s/it]\n",
      "MN-M14-L_S96_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.29s/it]\n",
      "MN-M14-L_S96_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.01s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.82s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 130\n",
      "unique names: 17\n",
      "after: 39 \n",
      "\n",
      "before: 35\n",
      "unique names: 21\n",
      "after: 27 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M14-L_S94_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 39/39 [03:53<00:00,  5.98s/it]\n",
      "MN-M14-L_S96_ME_L001__: 100%|███████████████████████████████████████████████████████████| 27/27 [04:02<00:00,  8.97s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M21 ---------------------------------------------------\n",
      "metatable shape: (1729, 32)\n",
      "sample file names: ['M21-L_S82_ME_L001__', 'MN-M21-L_S84_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 823   including test: 0\n",
      "somatic and quasi-somatic: (906, 32)   including test: 0 \n",
      "\n",
      "M21-L_S82_ME_L001__ 461   test: 0\n",
      "MN-M21-L_S84_ME_L001__ 445   test: 0\n",
      "overall num of somatic insertions:  906 \n",
      "\n",
      "M21-L_S82_ME_L001__  was: 461  become: 49\n",
      "M21-L_S82_ME_L001__ reads were written\n",
      "MN-M21-L_S84_ME_L001__  was: 445  become: 39\n",
      "MN-M21-L_S84_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M21-L_S82_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 4/4 [00:10<00:00,  2.52s/it]\n",
      "M21-L_S82_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 4/4 [00:13<00:00,  3.45s/it]\n",
      "MN-M21-L_S84_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.65s/it]\n",
      "MN-M21-L_S84_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.43s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.33s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 47\n",
      "unique names: 34\n",
      "after: 43 \n",
      "\n",
      "before: 58\n",
      "unique names: 24\n",
      "after: 34 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M21-L_S82_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 43/43 [05:32<00:00,  7.73s/it]\n",
      "MN-M21-L_S84_ME_L001__: 100%|███████████████████████████████████████████████████████████| 34/34 [04:56<00:00,  8.71s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M22 ---------------------------------------------------\n",
      "metatable shape: (2134, 32)\n",
      "sample file names: ['M22-L_S98_ME_L001__', 'MN-M22-L_S100_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 990   including test: 0\n",
      "somatic and quasi-somatic: (1144, 32)   including test: 0 \n",
      "\n",
      "M22-L_S98_ME_L001__ 555   test: 0\n",
      "MN-M22-L_S100_ME_L001__ 589   test: 0\n",
      "overall num of somatic insertions:  1144 \n",
      "\n",
      "M22-L_S98_ME_L001__  was: 555  become: 60\n",
      "M22-L_S98_ME_L001__ reads were written\n",
      "MN-M22-L_S100_ME_L001__  was: 589  become: 43\n",
      "MN-M22-L_S100_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M22-L_S98_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.31s/it]\n",
      "M22-L_S98_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:19<00:00,  3.18s/it]\n",
      "MN-M22-L_S100_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.32s/it]\n",
      "MN-M22-L_S100_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 4/4 [00:11<00:00,  3.00s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.07s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 43\n",
      "unique names: 35\n",
      "after: 43 \n",
      "\n",
      "before: 503\n",
      "unique names: 24\n",
      "after: 53 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M22-L_S98_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 43/43 [07:30<00:00, 10.47s/it]\n",
      "MN-M22-L_S100_ME_L001__: 100%|██████████████████████████████████████████████████████████| 53/53 [06:57<00:00,  7.88s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NC5 ---------------------------------------------------\n",
      "metatable shape: (1479, 32)\n",
      "sample file names: ['MN-NC5-L_S88_ME_L001__', 'NC5-L_S86_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 580   including test: 0\n",
      "somatic and quasi-somatic: (899, 32)   including test: 0 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__ 236   test: 0\n",
      "NC5-L_S86_ME_L001__ 663   test: 0\n",
      "overall num of somatic insertions:  899 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__  was: 236  become: 28\n",
      "MN-NC5-L_S88_ME_L001__ reads were written\n",
      "NC5-L_S86_ME_L001__  was: 663  become: 69\n",
      "NC5-L_S86_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-NC5-L_S88_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.68s/it]\n",
      "MN-NC5-L_S88_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.95s/it]\n",
      "NC5-L_S86_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.57s/it]\n",
      "NC5-L_S86_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:19<00:00,  3.25s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.79s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 94\n",
      "unique names: 13\n",
      "after: 28 \n",
      "\n",
      "before: 90\n",
      "unique names: 39\n",
      "after: 55 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-NC5-L_S88_ME_L001__: 100%|███████████████████████████████████████████████████████████| 28/28 [04:07<00:00,  8.85s/it]\n",
      "NC5-L_S86_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 55/55 [07:56<00:00,  8.66s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC14 ---------------------------------------------------\n",
      "metatable shape: (1712, 32)\n",
      "sample file names: ['MN-OC14-L_S112_ME_L001__', 'OC14-L_S110_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 764   including test: 0\n",
      "somatic and quasi-somatic: (948, 32)   including test: 0 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__ 503   test: 0\n",
      "OC14-L_S110_ME_L001__ 445   test: 0\n",
      "overall num of somatic insertions:  948 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__  was: 503  become: 56\n",
      "MN-OC14-L_S112_ME_L001__ reads were written\n",
      "OC14-L_S110_ME_L001__  was: 445  become: 44\n",
      "OC14-L_S110_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC14-L_S112_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.09s/it]\n",
      "MN-OC14-L_S112_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 5/5 [00:13<00:00,  2.76s/it]\n",
      "OC14-L_S110_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.48s/it]\n",
      "OC14-L_S110_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.22s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.57s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 91\n",
      "unique names: 33\n",
      "after: 58 \n",
      "\n",
      "before: 251\n",
      "unique names: 27\n",
      "after: 54 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC14-L_S112_ME_L001__: 100%|█████████████████████████████████████████████████████████| 58/58 [08:10<00:00,  8.45s/it]\n",
      "OC14-L_S110_ME_L001__: 100%|████████████████████████████████████████████████████████████| 54/54 [06:50<00:00,  7.60s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC33 ---------------------------------------------------\n",
      "metatable shape: (815, 32)\n",
      "sample file names: ['MN-OC33-L_S116_ME_L001__', 'OC33-L_S114_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 242   including test: 0\n",
      "somatic and quasi-somatic: (573, 32)   including test: 0 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__ 213   test: 0\n",
      "OC33-L_S114_ME_L001__ 360   test: 0\n",
      "overall num of somatic insertions:  573 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__  was: 213  become: 13\n",
      "MN-OC33-L_S116_ME_L001__ reads were written\n",
      "OC33-L_S114_ME_L001__  was: 360  become: 27\n",
      "OC33-L_S114_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC33-L_S116_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.96s/it]\n",
      "MN-OC33-L_S116_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.65s/it]\n",
      "OC33-L_S114_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.45s/it]\n",
      "OC33-L_S114_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.29s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.20s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 9\n",
      "unique names: 8\n",
      "after: 9 \n",
      "\n",
      "before: 25\n",
      "unique names: 20\n",
      "after: 25 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC33-L_S116_ME_L001__: 100%|███████████████████████████████████████████████████████████| 9/9 [01:35<00:00, 10.63s/it]\n",
      "OC33-L_S114_ME_L001__: 100%|████████████████████████████████████████████████████████████| 25/25 [03:50<00:00,  9.21s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC40 ---------------------------------------------------\n",
      "metatable shape: (1450, 32)\n",
      "sample file names: ['MN-OC40-L_S80_ME_L001__', 'OC40-L_S78_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 595   including test: 0\n",
      "somatic and quasi-somatic: (855, 32)   including test: 0 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__ 272   test: 0\n",
      "OC40-L_S78_ME_L001__ 583   test: 0\n",
      "overall num of somatic insertions:  855 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__  was: 272  become: 19\n",
      "MN-OC40-L_S80_ME_L001__ reads were written\n",
      "OC40-L_S78_ME_L001__  was: 583  become: 56\n",
      "OC40-L_S78_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC40-L_S80_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.43s/it]\n",
      "MN-OC40-L_S80_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.11s/it]\n",
      "OC40-L_S78_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.15s/it]\n",
      "OC40-L_S78_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:14<00:00,  2.82s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.13s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 21\n",
      "unique names: 6\n",
      "after: 10 \n",
      "\n",
      "before: 75\n",
      "unique names: 35\n",
      "after: 54 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC40-L_S80_ME_L001__: 100%|██████████████████████████████████████████████████████████| 10/10 [01:26<00:00,  8.61s/it]\n",
      "OC40-L_S78_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 54/54 [07:50<00:00,  8.71s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC42 ---------------------------------------------------\n",
      "metatable shape: (382, 32)\n",
      "sample file names: ['MN-OC42-L_S40_ME_L001__', 'OC42-L_S38_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 30   including test: 0\n",
      "somatic and quasi-somatic: (352, 32)   including test: 0 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__ 152   test: 0\n",
      "OC42-L_S38_ME_L001__ 200   test: 0\n",
      "overall num of somatic insertions:  352 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__  was: 152  become: 12\n",
      "MN-OC42-L_S40_ME_L001__ reads were written\n",
      "OC42-L_S38_ME_L001__  was: 200  become: 15\n",
      "OC42-L_S38_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC42-L_S40_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.97s/it]\n",
      "MN-OC42-L_S40_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.42s/it]\n",
      "OC42-L_S38_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.95s/it]\n",
      "OC42-L_S38_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.81s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.34it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 127\n",
      "unique names: 5\n",
      "after: 10 \n",
      "\n",
      "before: 15\n",
      "unique names: 12\n",
      "after: 15 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC42-L_S40_ME_L001__: 100%|██████████████████████████████████████████████████████████| 10/10 [01:11<00:00,  7.13s/it]\n",
      "OC42-L_S38_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 15/15 [02:54<00:00, 11.63s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC48 ---------------------------------------------------\n",
      "metatable shape: (1798, 32)\n",
      "sample file names: ['MN-OC48-L_S44_ME_L001__', 'OC48-L_S42_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 699   including test: 0\n",
      "somatic and quasi-somatic: (1099, 32)   including test: 0 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__ 365   test: 0\n",
      "OC48-L_S42_ME_L001__ 734   test: 0\n",
      "overall num of somatic insertions:  1099 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__  was: 365  become: 34\n",
      "MN-OC48-L_S44_ME_L001__ reads were written\n",
      "OC48-L_S42_ME_L001__  was: 734  become: 62\n",
      "OC48-L_S42_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC48-L_S44_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.42s/it]\n",
      "MN-OC48-L_S44_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.27s/it]\n",
      "OC48-L_S42_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:14<00:00,  2.48s/it]\n",
      "OC48-L_S42_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:19<00:00,  3.27s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.02s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 213\n",
      "unique names: 19\n",
      "after: 31 \n",
      "\n",
      "before: 110\n",
      "unique names: 39\n",
      "after: 60 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC48-L_S44_ME_L001__: 100%|██████████████████████████████████████████████████████████| 31/31 [03:55<00:00,  7.59s/it]\n",
      "OC48-L_S42_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 60/60 [09:38<00:00,  9.65s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC50 ---------------------------------------------------\n",
      "metatable shape: (1873, 32)\n",
      "sample file names: ['MN-OC50-L_S128_ME_L001__', 'OC50-L_S126_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 841   including test: 0\n",
      "somatic and quasi-somatic: (1032, 32)   including test: 0 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__ 430   test: 0\n",
      "OC50-L_S126_ME_L001__ 602   test: 0\n",
      "overall num of somatic insertions:  1032 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__  was: 430  become: 43\n",
      "MN-OC50-L_S128_ME_L001__ reads were written\n",
      "OC50-L_S126_ME_L001__  was: 602  become: 59\n",
      "OC50-L_S126_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC50-L_S128_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.35s/it]\n",
      "MN-OC50-L_S128_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 4/4 [00:11<00:00,  2.89s/it]\n",
      "OC50-L_S126_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.16s/it]\n",
      "OC50-L_S126_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 5/5 [00:16<00:00,  3.24s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.12s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 366\n",
      "unique names: 25\n",
      "after: 47 \n",
      "\n",
      "before: 49\n",
      "unique names: 36\n",
      "after: 47 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC50-L_S128_ME_L001__: 100%|█████████████████████████████████████████████████████████| 47/47 [06:38<00:00,  8.48s/it]\n",
      "OC50-L_S126_ME_L001__: 100%|████████████████████████████████████████████████████████████| 47/47 [07:55<00:00, 10.11s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC54 ---------------------------------------------------\n",
      "metatable shape: (1878, 32)\n",
      "sample file names: ['MN-OC54-L_S56_ME_L001__', 'OC54-L_S54_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 854   including test: 0\n",
      "somatic and quasi-somatic: (1024, 32)   including test: 0 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__ 453   test: 0\n",
      "OC54-L_S54_ME_L001__ 571   test: 0\n",
      "overall num of somatic insertions:  1024 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__  was: 453  become: 33\n",
      "MN-OC54-L_S56_ME_L001__ reads were written\n",
      "OC54-L_S54_ME_L001__  was: 571  become: 57\n",
      "OC54-L_S54_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC54-L_S56_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.43s/it]\n",
      "MN-OC54-L_S56_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.91s/it]\n",
      "OC54-L_S54_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.40s/it]\n",
      "OC54-L_S54_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:15<00:00,  3.18s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.94s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 71\n",
      "unique names: 16\n",
      "after: 28 \n",
      "\n",
      "before: 276\n",
      "unique names: 32\n",
      "after: 59 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC54-L_S56_ME_L001__: 100%|██████████████████████████████████████████████████████████| 28/28 [03:37<00:00,  7.78s/it]\n",
      "OC54-L_S54_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 59/59 [06:48<00:00,  6.92s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OT1 ---------------------------------------------------\n",
      "metatable shape: (1157, 32)\n",
      "sample file names: ['MN-OT1-L_S152_ME_L001__', 'OT1-L_S150_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 505   including test: 0\n",
      "somatic and quasi-somatic: (652, 32)   including test: 0 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__ 303   test: 0\n",
      "OT1-L_S150_ME_L001__ 349   test: 0\n",
      "overall num of somatic insertions:  652 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__  was: 303  become: 20\n",
      "MN-OT1-L_S152_ME_L001__ reads were written\n",
      "OT1-L_S150_ME_L001__  was: 349  become: 31\n",
      "OT1-L_S150_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OT1-L_S152_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.85s/it]\n",
      "MN-OT1-L_S152_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.86s/it]\n",
      "OT1-L_S150_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.06s/it]\n",
      "OT1-L_S150_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.59s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:02<00:00,  1.28s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 33\n",
      "unique names: 9\n",
      "after: 17 \n",
      "\n",
      "before: 283\n",
      "unique names: 23\n",
      "after: 43 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OT1-L_S152_ME_L001__: 100%|██████████████████████████████████████████████████████████| 17/17 [02:35<00:00,  9.12s/it]\n",
      "OT1-L_S150_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 43/43 [06:02<00:00,  8.42s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC12 ---------------------------------------------------\n",
      "metatable shape: (2150, 32)\n",
      "sample file names: ['MN-PC12-L_S24_ME_L001__', 'PC12-L_S22_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 541   including test: 0\n",
      "somatic and quasi-somatic: (1609, 32)   including test: 0 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__ 342   test: 0\n",
      "PC12-L_S22_ME_L001__ 1267   test: 0\n",
      "overall num of somatic insertions:  1609 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__  was: 342  become: 28\n",
      "MN-PC12-L_S24_ME_L001__ reads were written\n",
      "PC12-L_S22_ME_L001__  was: 1267  become: 120\n",
      "PC12-L_S22_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC12-L_S24_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.37s/it]\n",
      "MN-PC12-L_S24_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.17s/it]\n",
      "PC12-L_S22_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 12/12 [00:27<00:00,  2.29s/it]\n",
      "PC12-L_S22_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 12/12 [00:36<00:00,  3.03s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:12<00:00,  6.25s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 11\n",
      "unique names: 9\n",
      "after: 11 \n",
      "\n",
      "before: 286\n",
      "unique names: 83\n",
      "after: 136 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC12-L_S24_ME_L001__: 100%|██████████████████████████████████████████████████████████| 11/11 [01:16<00:00,  7.00s/it]\n",
      "PC12-L_S22_ME_L001__: 100%|███████████████████████████████████████████████████████████| 136/136 [20:02<00:00,  8.84s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC14 ---------------------------------------------------\n",
      "metatable shape: (1567, 32)\n",
      "sample file names: ['MN-PC14-L_S36_ME_L001__', 'PC14-L_S34_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 346   including test: 0\n",
      "somatic and quasi-somatic: (1221, 32)   including test: 0 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__ 206   test: 0\n",
      "PC14-L_S34_ME_L001__ 1015   test: 0\n",
      "overall num of somatic insertions:  1221 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__  was: 206  become: 18\n",
      "MN-PC14-L_S36_ME_L001__ reads were written\n",
      "PC14-L_S34_ME_L001__  was: 1015  become: 100\n",
      "PC14-L_S34_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC14-L_S36_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.49s/it]\n",
      "MN-PC14-L_S36_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.05s/it]\n",
      "PC14-L_S34_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 10/10 [00:24<00:00,  2.42s/it]\n",
      "PC14-L_S34_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 10/10 [00:31<00:00,  3.18s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.66s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 228\n",
      "unique names: 10\n",
      "after: 38 \n",
      "\n",
      "before: 160\n",
      "unique names: 67\n",
      "after: 92 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC14-L_S36_ME_L001__: 100%|██████████████████████████████████████████████████████████| 38/38 [04:18<00:00,  6.81s/it]\n",
      "PC14-L_S34_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 92/92 [16:10<00:00, 10.55s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC21 ---------------------------------------------------\n",
      "metatable shape: (1596, 32)\n",
      "sample file names: ['MN-PC21-L_S20_ME_L001__', 'PC21-L_S18_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 660   including test: 0\n",
      "somatic and quasi-somatic: (936, 32)   including test: 0 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__ 597   test: 0\n",
      "PC21-L_S18_ME_L001__ 339   test: 0\n",
      "overall num of somatic insertions:  936 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__  was: 597  become: 61\n",
      "MN-PC21-L_S20_ME_L001__ reads were written\n",
      "PC21-L_S18_ME_L001__  was: 339  become: 35\n",
      "PC21-L_S18_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC21-L_S20_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.29s/it]\n",
      "MN-PC21-L_S20_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 6/6 [00:19<00:00,  3.18s/it]\n",
      "PC21-L_S18_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.24s/it]\n",
      "PC21-L_S18_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.01s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:12<00:00,  6.26s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 215\n",
      "unique names: 36\n",
      "after: 72 \n",
      "\n",
      "before: 70\n",
      "unique names: 23\n",
      "after: 36 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC21-L_S20_ME_L001__: 100%|██████████████████████████████████████████████████████████| 72/72 [09:16<00:00,  7.72s/it]\n",
      "PC21-L_S18_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 36/36 [05:09<00:00,  8.61s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC32 ---------------------------------------------------\n",
      "metatable shape: (1964, 32)\n",
      "sample file names: ['MN-PC32-L_S32_ME_L001__', 'PC32-L_S30_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 178   including test: 0\n",
      "somatic and quasi-somatic: (1786, 32)   including test: 0 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__ 244   test: 0\n",
      "PC32-L_S30_ME_L001__ 1542   test: 0\n",
      "overall num of somatic insertions:  1786 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__  was: 244  become: 22\n",
      "MN-PC32-L_S32_ME_L001__ reads were written\n",
      "PC32-L_S30_ME_L001__  was: 1542  become: 146\n",
      "PC32-L_S30_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC32-L_S32_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.93s/it]\n",
      "MN-PC32-L_S32_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.51s/it]\n",
      "PC32-L_S30_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 14/14 [00:36<00:00,  2.59s/it]\n",
      "PC32-L_S30_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 14/14 [00:48<00:00,  3.45s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:15<00:00,  7.87s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 9\n",
      "unique names: 9\n",
      "after: 9 \n",
      "\n",
      "before: 194\n",
      "unique names: 89\n",
      "after: 140 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC32-L_S32_ME_L001__: 100%|████████████████████████████████████████████████████████████| 9/9 [01:21<00:00,  9.08s/it]\n",
      "PC32-L_S30_ME_L001__: 100%|███████████████████████████████████████████████████████████| 140/140 [22:45<00:00,  9.76s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SkC6 ---------------------------------------------------\n",
      "metatable shape: (1896, 32)\n",
      "sample file names: ['MN-SkC6-L_S60_ME_L001__', 'SkC6-L_S58_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 804   including test: 0\n",
      "somatic and quasi-somatic: (1092, 32)   including test: 0 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__ 582   test: 0\n",
      "SkC6-L_S58_ME_L001__ 510   test: 0\n",
      "overall num of somatic insertions:  1092 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__  was: 582  become: 63\n",
      "MN-SkC6-L_S60_ME_L001__ reads were written\n",
      "SkC6-L_S58_ME_L001__  was: 510  become: 59\n",
      "SkC6-L_S58_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-SkC6-L_S60_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 6/6 [00:14<00:00,  2.43s/it]\n",
      "MN-SkC6-L_S60_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 6/6 [00:18<00:00,  3.15s/it]\n",
      "SkC6-L_S58_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.05s/it]\n",
      "SkC6-L_S58_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:14<00:00,  3.00s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.71s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 39\n",
      "unique names: 34\n",
      "after: 39 \n",
      "\n",
      "before: 150\n",
      "unique names: 37\n",
      "after: 41 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-SkC6-L_S60_ME_L001__: 100%|██████████████████████████████████████████████████████████| 39/39 [05:56<00:00,  9.13s/it]\n",
      "SkC6-L_S58_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 41/41 [08:52<00:00, 12.99s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "StS1 ---------------------------------------------------\n",
      "metatable shape: (1933, 32)\n",
      "sample file names: ['MN-StS1-L_S92_ME_L001__', 'StS1-L_S90_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 808   including test: 0\n",
      "somatic and quasi-somatic: (1125, 32)   including test: 0 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__ 473   test: 0\n",
      "StS1-L_S90_ME_L001__ 652   test: 0\n",
      "overall num of somatic insertions:  1125 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__  was: 473  become: 40\n",
      "MN-StS1-L_S92_ME_L001__ reads were written\n",
      "StS1-L_S90_ME_L001__  was: 652  become: 59\n",
      "StS1-L_S90_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-StS1-L_S92_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 4/4 [00:14<00:00,  3.74s/it]\n",
      "MN-StS1-L_S92_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 4/4 [00:11<00:00,  2.97s/it]\n",
      "StS1-L_S90_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:10<00:00,  2.10s/it]\n",
      "StS1-L_S90_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:14<00:00,  2.88s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.25s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 75\n",
      "unique names: 16\n",
      "after: 28 \n",
      "\n",
      "before: 242\n",
      "unique names: 36\n",
      "after: 68 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-StS1-L_S92_ME_L001__: 100%|██████████████████████████████████████████████████████████| 28/28 [03:52<00:00,  8.30s/it]\n",
      "StS1-L_S90_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 68/68 [09:53<00:00,  8.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "UC2 ---------------------------------------------------\n",
      "metatable shape: (1364, 32)\n",
      "sample file names: ['MN-UC2-L_S124_ME_L001__', 'UC2-L_S122_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 604   including test: 0\n",
      "somatic and quasi-somatic: (760, 32)   including test: 0 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__ 254   test: 0\n",
      "UC2-L_S122_ME_L001__ 506   test: 0\n",
      "overall num of somatic insertions:  760 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__  was: 254  become: 22\n",
      "MN-UC2-L_S124_ME_L001__ reads were written\n",
      "UC2-L_S122_ME_L001__  was: 506  become: 37\n",
      "UC2-L_S122_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-UC2-L_S124_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.16s/it]\n",
      "MN-UC2-L_S124_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.70s/it]\n",
      "UC2-L_S122_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.41s/it]\n",
      "UC2-L_S122_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.15s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.87s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 23\n",
      "unique names: 11\n",
      "after: 23 \n",
      "\n",
      "before: 117\n",
      "unique names: 24\n",
      "after: 39 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-UC2-L_S124_ME_L001__: 100%|██████████████████████████████████████████████████████████| 23/23 [02:55<00:00,  7.61s/it]\n",
      "UC2-L_S122_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 39/39 [05:25<00:00,  8.36s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "dirs = os.listdir() #[8:]\n",
    "k = 0\n",
    "avg_time = 0\n",
    "for dirname in dirs:\n",
    "    print(dirname, '---------------------------------------------------')\n",
    "    curr_dir = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/AGCT/' + dirname\n",
    "    if not os.path.isdir(curr_dir):\n",
    "        continue\n",
    "    os.chdir(curr_dir)\n",
    "    DIR = os.getcwd() + '/nofiltration'\n",
    "    t = time.time()\n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir)\n",
    "    # 2 Filtration\n",
    "    writedown_filtered(DIR, file_names, df_somatic_list)\n",
    "    # 3 UCSC BLAT parsing\n",
    "    blat_parsing(DIR, file_names)\n",
    "    # 4 BLAT analysis\n",
    "    blat_analysis(DIR, file_names)\n",
    "    # 5 RepeatMasker parsing\n",
    "    blat_res, RM_res = get_10best_alings(DIR, file_names)\n",
    "    RM_parsing(DIR, file_names, blat_res, RM_res)\n",
    "    # 6 Merging\n",
    "    merging(DIR, file_names)\n",
    "    k += 1\n",
    "    if k == 1:\n",
    "        avg_time = int(time.time() - t)\n",
    "    avg_time = (avg_time + int(time.time() - t)) / 2\n",
    "    test = telegram_bot_sendtext(f\"Postprocessing done for {dirname}_AGCT\\n\" + f\"Duration: {int(time.time() - t)} sec\\n\" + \n",
    "                                 f\"Time to finish (approx): {round(avg_time * (len(dirs) - k) / 3600, 1)} hours\")     "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db2dbb9b-3e55-4447-97e7-506a91f3733c",
   "metadata": {},
   "source": [
    "### Output data filtration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "854976dc-f742-4bb7-8642-b650a576d7dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def get_somatic(path, show=True):\n",
    "    # read pipeline output\n",
    "    res = pd.read_table(path + '/metatable_ts.txt')\n",
    "    if show:\n",
    "        print('metatable shape:', res.shape)\n",
    "    #print(res.head())\n",
    "\n",
    "    # make list of file names\n",
    "    file_names = []\n",
    "    for col_name in res.columns:\n",
    "        if col_name.find('_NUM_READS') != -1:\n",
    "            file_names.append(col_name[:-10])\n",
    "    if show:\n",
    "        print('sample file names:', file_names)\n",
    "    \n",
    "    # checking for null occurrences \n",
    "    df_0 = res.copy(deep=True)\n",
    "    for i in range(len(file_names)):\n",
    "        df_0 = df_0[df_0[file_names[i] + '_NUM_READS'] == 0]\n",
    "    if show:\n",
    "        print('null occurrences:', df_0.shape[0])\n",
    "    \n",
    "    # not somatic (germline) insertions (exists in all samples)\n",
    "    df_germ = res.copy(deep=True)\n",
    "    for i in range(len(file_names)):\n",
    "        df_germ = df_germ[df_germ[file_names[i] + '_NUM_READS'] > 0]\n",
    "    if show:\n",
    "        print('all not somatic:', df_germ.shape[0],  '  including test:', df_germ[df_germ['READNAME'].str.contains('ARTIFICIAL')].shape[0])\n",
    "    \n",
    "    # somatic and quasi-somatic (exists for example in 3/4 samples) insertions\n",
    "    df_qsom = res.copy(deep=True)\n",
    "    df_qsom = df_qsom.drop(df_germ.index, axis=0)\n",
    "    if show:\n",
    "        print('somatic and quasi-somatic:', df_qsom.shape, '  including test:', df_qsom[df_qsom['READNAME'].str.contains('ARTIFICIAL')].shape[0], '\\n')\n",
    "    \n",
    "    # insertions specific only for this tissue/sample(file) \n",
    "    df_somatic_list = []\n",
    "    overall_num = 0\n",
    "    for i in range(len(file_names)):\n",
    "        df = df_qsom.copy(deep=True)\n",
    "        df = df[df[file_names[i] + '_NUM_READS'] > 0]\n",
    "        for j in range(1, len(file_names)):\n",
    "            df = df[df[file_names[i - j] + '_NUM_READS'] == 0]\n",
    "        if show:\n",
    "            print(file_names[i], df.shape[0], '  test:', df[df['READNAME'].str.contains('ARTIFICIAL')].shape[0])\n",
    "        overall_num += df.shape[0]\n",
    "        df_somatic_list.append(df.copy(deep=True))\n",
    "    if show:\n",
    "        print('overall num of somatic insertions: ', overall_num, '\\n')\n",
    "    return file_names, df_somatic_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "50dc1107-a2f3-49bc-b811-c5ae6289b393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# preparing for copying into blat and filtering \n",
    "def writedown_filtered(DIR, file_names, df_somatic_list):\n",
    "    if not os.path.exists(DIR + '/filtred'):\n",
    "        os.makedirs(DIR + '/filtred')\n",
    "    \n",
    "    for i in range(len(file_names)):\n",
    "        # filtration\n",
    "        df_tmp = df_somatic_list[i].copy(deep=True)\n",
    "        #df_tmp = df_tmp[df_tmp[file_names[i] + '_NUM_READS'] > 1].copy(deep=True)\n",
    "        df_tmp = df_tmp[df_tmp['R1'].str.len() > 25].copy(deep=True)  # because Blat can aling seq > 25nt\n",
    "        df_tmp = df_tmp[df_tmp['R2'].str.len() > 25].copy(deep=True)\n",
    "        df_tmp = df_tmp[df_tmp['MDMATCH'] < (52+6+14 - 5)].copy(deep=True)  # template_switch_filtration: 12 primer + 6 dALU + 14 flank = 32\n",
    "        print(file_names[i], ' was:', df_somatic_list[i].shape[0], ' become:', df_tmp.shape[0])\n",
    "    \n",
    "        df_tmp.to_csv(DIR + '/filtred/' + f'{file_names[i]}_filtered.csv', index=False)\n",
    "        print(f'{file_names[i]} reads were written')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4bec163-6591-4b65-b8e4-39d72b250720",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### UCSC Blat parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e2b30646-5aad-4efa-bec0-32b37eaf7db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement(char):\n",
    "    if char == 'A':\n",
    "        return 'T'\n",
    "    if char == 'T':\n",
    "        return 'A'\n",
    "    if char == 'C':\n",
    "        return 'G'\n",
    "    if char == 'G':\n",
    "        return 'C'\n",
    "    if char == 'N':\n",
    "        return 'N'\n",
    "\n",
    "def reverse_comp(seq):\n",
    "    seq_ls = list(seq)\n",
    "    new_seq = []\n",
    "    for i in reversed(seq_ls):\n",
    "        new_seq.append(complement(i))\n",
    "    return \"\".join(new_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "660b628e-5486-464d-8562-6231870c1c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reads_fa(reads_list, reads_names, R_n=1):\n",
    "    line = ''\n",
    "    for read, name in zip(reads_list, reads_names):\n",
    "        # if R_n == 1:  # for Alu 5' end\n",
    "        if R_n == 2:  # for LINE 3' end\n",
    "            line = line + '>' + name + f'_R{R_n}' + '\\n' + reverse_comp(read) + '\\n'  # because if insertion is \"+\" then R1 is rev_comp of + strand dna\n",
    "        else:\n",
    "            line = line + '>' + name + f'_R{R_n}' + '\\n' + read + '\\n'\n",
    "    return line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e54449ea-8362-44b4-b448-57d336351ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_blat_output(driver, filtered_flanks, i, k, j, length=10):\n",
    "    # set flank coords and submit\n",
    "    range_input = driver.find_element(By.NAME, \"userSeq\")\n",
    "    range_input.clear()\n",
    "    #push 10 reads at once\n",
    "    read_list = filtered_flanks[i][f'R{k}'].tolist()[j*10 : j*10 + length]\n",
    "    name_list = filtered_flanks[i]['READNAME'].tolist()[j*10 : j*10 + length]\n",
    "    range_input.send_keys(get_reads_fa(read_list, name_list, R_n=k))\n",
    "    \n",
    "    button = driver.find_element(By.NAME, \"Submit\")\n",
    "    button.click()\n",
    "    # get results \n",
    "    blat_out = ''\n",
    "    try:\n",
    "        #print('waiting')\n",
    "        element = WebDriverWait(driver, 60).until(\n",
    "                            EC.presence_of_element_located((By.TAG_NAME, \"pre\")) \n",
    "                            )\n",
    "        blat_out = element.text\n",
    "    except:\n",
    "        print(f'exception for {file_names[i]}')\n",
    "    # get back to request form       \n",
    "    driver.back()\n",
    "    return blat_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3b1ba3e-b957-4e9f-a196-317bbded61e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import trange\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.wait import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.support.select import Select\n",
    "\n",
    "def blat_parsing(DIR, file_names):\n",
    "    service = Service(executable_path=\"/usr/bin/chromedriver\")\n",
    "    chrome_options = Options()\n",
    "    chrome_options.add_argument('--headless')\n",
    "    chrome_options.add_argument('--no-sandbox')\n",
    "    chrome_options.add_argument('--disable-dev-shm-usage')\n",
    "    driver = webdriver.Chrome(service=service, options=chrome_options)\n",
    "    \n",
    "    filtered_flanks = []\n",
    "    for i in range(len(file_names)):\n",
    "        df_tmp = pd.read_csv(DIR + '/filtred/' + f'{file_names[i]}_filtered.csv')\n",
    "        filtered_flanks.append(df_tmp.copy(deep=True))\n",
    "    \n",
    "    # go to the UCSC Blat\n",
    "    driver.get(\"https://genome.ucsc.edu/cgi-bin/hgBlat?hgsid=1711647570_KRW9QfFHAHaje8cp8sWT4EtpGupE&command=start\")\n",
    "    print(driver.title)\n",
    "    \n",
    "    # set Hg38\n",
    "    selects = driver.find_elements(By.TAG_NAME, \"select\")\n",
    "    for i in selects:\n",
    "        if i.get_property(\"name\") == \"db\":\n",
    "            element = i\n",
    "            break\n",
    "    dropdown = Select(element)\n",
    "    dropdown.select_by_visible_text(\"Dec. 2013 (GRCh38/hg38)\")\n",
    "    \n",
    "    if not os.path.exists(DIR + '/blat_output'):\n",
    "        os.makedirs(DIR + '/blat_output')\n",
    "    \n",
    "    for i in range(len(filtered_flanks)):\n",
    "        for k in [1, 2]:\n",
    "            output = ''\n",
    "            for j in trange(int(filtered_flanks[i].shape[0] / 10), desc=f'{file_names[i]}_R{k}'):  \n",
    "                output = output + '\\n' + get_blat_output(driver, filtered_flanks, i, k, j)\n",
    "            if filtered_flanks[i].shape[0] % 10 != 0:\n",
    "                output = output + '\\n' + get_blat_output(driver, filtered_flanks, i, k, int(filtered_flanks[i].shape[0] / 10), \n",
    "                                                         length=filtered_flanks[i].shape[0] % 10)\n",
    "                \n",
    "            with open(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out_tmp.txt', 'w') as f:\n",
    "                f.write(output)\n",
    "                #print('res written')\n",
    "    \n",
    "    driver.quit()\n",
    "\n",
    "    for i in range(len(filtered_flanks)):\n",
    "        for k in [1, 2]:\n",
    "            blat_out = pd.DataFrame(columns=['QUERY', 'SCORE', 'QSTART', 'QEND', 'QSIZE', 'IDENTITY', 'CHROM', 'STRAND', 'START', 'END', 'SPAN'])\n",
    "            with open(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out_tmp.txt', 'r') as f:\n",
    "                output = f.read()\n",
    "            output = output.split('\\n')\n",
    "            for line in output:\n",
    "                    if line.count('browser'):\n",
    "                        tmp = line.split(' ')\n",
    "                        tmp = ' '.join(tmp).split()\n",
    "                        blat_out.loc[len(blat_out.index)] = [tmp[2], tmp[3], tmp[4], tmp[5], tmp[6], tmp[7], tmp[8], tmp[9], tmp[10], tmp[11], tmp[12]]\n",
    "            blat_out.to_csv(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out.csv', index=False) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81a7be26-c0e6-4f9b-965b-5109ac089de3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Blat analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7251a234-dbfa-471c-8fd1-e468b75f6aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import trange\n",
    "\n",
    "def blat_merge_and_filtration(filename):\n",
    "\n",
    "    FRAGMENT_LEN = 1000  # from electroforeses flanks will be from ~60 to ~585\n",
    "\n",
    "    #chomosome list\n",
    "    chromosomes = list(range(1, 23))\n",
    "    chromosomes.append('X')\n",
    "    chromosomes.append('Y')\n",
    "    chromosomes = ['chr' + str(x) for x in chromosomes]\n",
    "    \n",
    "    df1 = pd.read_table(filename + '_R1_blat_out.csv', sep=',')\n",
    "    #df1 = df1.drop(['Unnamed: 0'], axis=1)\n",
    "    \n",
    "    df2 = pd.read_table(filename + '_R2_blat_out.csv', sep=',')\n",
    "    #df2 = df2.drop(['Unnamed: 0'], axis=1)\n",
    "    \n",
    "    read_names = df1['QUERY'].unique()\n",
    "    merget_df = pd.DataFrame(columns=['QUERY', \n",
    "                                      'SCORE_1', 'QSIZE_1', 'QSTART_1', 'QEND_1', 'IDENTITY_1', 'CHROM_1', 'STRAND_1', 'START_1', 'END_1', 'SPAN_1',\n",
    "                                      'SCORE_2', 'QSIZE_2', 'QSTART_2', 'QEND_2', 'IDENTITY_2', 'CHROM_2', 'STRAND_2', 'START_2', 'END_2', 'SPAN_2'\n",
    "                                     ]\n",
    "                            )\n",
    "    for name in read_names:\n",
    "        df1_1 = df1[df1['QUERY'].str.contains(name)].copy(deep=True)\n",
    "        df2_1 = df2[df2['QUERY'].str.contains(name[:-3])].copy(deep=True)\n",
    "        for i, row_1 in df1_1.iterrows():\n",
    "            tmp_df = df2_1[df2_1['CHROM'] == row_1['CHROM']]\n",
    "            for i, row_tmp in tmp_df.iterrows():\n",
    "                if abs(row_tmp['START'] - row_1['START']) <= FRAGMENT_LEN and abs(row_tmp['END'] - row_1['END']) <= FRAGMENT_LEN:\n",
    "                    merget_df.loc[len(merget_df.index)] = [row_1['QUERY'][:-3], \n",
    "                                                           row_1['SCORE'], row_1['QSIZE'], row_1['QSTART'], row_1['QEND'], row_1['IDENTITY'],\n",
    "                                                           row_1['CHROM'], row_1['STRAND'], row_1['START'], row_1['END'], row_1['SPAN'],\n",
    "                                                           row_tmp['SCORE'], row_tmp['QSIZE'], row_tmp['QSTART'], row_tmp['QEND'], row_tmp['IDENTITY'], \n",
    "                                                           row_tmp['CHROM'], row_tmp['STRAND'], row_tmp['START'], row_tmp['END'], row_tmp['SPAN'] \n",
    "                                                          ]\n",
    "    #print(name[:-3], df1_1.shape[0], df2_1.shape[0], merget_df.shape[0])\n",
    "    \n",
    "    #drop alt chr\n",
    "    merget_df.drop(merget_df[ ~merget_df['CHROM_1'].isin(chromosomes) ].index, axis=0, inplace=True)\n",
    "    #drop matches with different directions \n",
    "    merget_df = merget_df[merget_df['STRAND_1'] == merget_df['STRAND_2']].copy(deep=True)\n",
    "\n",
    "    # get only high score\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df = merget_df[merget_df['SCORE_1'] > 0.9 * merget_df['QSIZE_1']].copy(deep=True)\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df = merget_df[merget_df['SCORE_2'] > 0.9 * merget_df['QSIZE_2']].copy(deep=True)\n",
    "\n",
    "    # drop aligns which are not close to RE 5'end \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['QEND_1'] <= merget_df['QSIZE_1'] - 3) ].index, axis=0, inplace=True)   \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['QSTART_1'] >= 3) ].index, axis=0, inplace=True)\n",
    "\n",
    "    # drop strange aligns: \"+\" R2_start >= R1_end; \"-\" R1_start >= R2_end\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['START_2'] >= merget_df['END_1']) ].index, axis=0, inplace=True)   \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['START_1'] >= merget_df['END_2']) ].index, axis=0, inplace=True)\n",
    "        \n",
    "    # calc flank metrics and insertion coord \n",
    "    if merget_df.shape[0] > 0:\n",
    "        for i, row in merget_df.iterrows():\n",
    "            merget_df.loc[i, 'F_CHR'] = row['CHROM_1']\n",
    "            start = int(min(row['START_1'], row['START_2'], row['END_1'], row['END_2']))\n",
    "            end = int(max(row['START_1'], row['START_2'], row['END_1'], row['END_2']))\n",
    "            merget_df.loc[i, 'F_START'] = start\n",
    "            merget_df.loc[i, 'F_END'] = end\n",
    "            merget_df.loc[i, 'F_LEN'] = int(end - start) + 1\n",
    "            if row['STRAND_1'] == '+':\n",
    "                merget_df.loc[i, 'INS_COORD'] = end\n",
    "            else:\n",
    "                merget_df.loc[i, 'INS_COORD'] = start - 1\n",
    "            \n",
    "    return merget_df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "257ca7e7-032a-4492-a2e8-6a659b1641b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def blat_analysis(DIR, file_names):\n",
    "    if not os.path.exists(DIR + '/blat_filtred'):\n",
    "        os.makedirs(DIR + '/blat_filtred')\n",
    "    \n",
    "    for i in trange(len(file_names), desc='blat analysis:'):\n",
    "        df_tmp = blat_merge_and_filtration(DIR + '/blat_output/' + file_names[i])\n",
    "        df_tmp.to_csv(DIR + '/blat_filtred/' + f'{file_names[i]}_blat_filtred.csv', index=False)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3128433c-7822-4342-bb60-f966cd19ecf3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### RepeatMasker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "39de57f6-133d-4996-bb6b-9fd374e765a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get 5 best aligns for each read\n",
    "import pandas as pd\n",
    "\n",
    "def get_10best_alings(DIR, file_names):\n",
    "    blat_res = []\n",
    "    RM_res = []\n",
    "    for name in file_names:\n",
    "        blat_res.append(pd.read_table(DIR + '/blat_filtred/' + f'{name}_blat_filtred.csv', sep=','))\n",
    "        RM_res.append(pd.DataFrame(columns=['QUERY', 'EXTENSION', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', 'REPEAT_NAME', 'R_STRAND', 'R_CHR', \n",
    "                                            'R_START', 'R_END', 'INS_OVERLAP', 'DESC', 'F_LEN', 'R_LEN', 'F_START', 'F_END',\n",
    "                                            'SCORE_1', 'QSIZE_1', 'SCORE_2', 'QSIZE_2', 'CHR_1', 'START_1', 'END_1', 'CHR_2', 'START_2', 'END_2', \n",
    "                                            ]))\n",
    "\n",
    "    for blat_res_part in blat_res:\n",
    "        print('before:', blat_res_part.shape[0])\n",
    "        amount_list = blat_res_part.groupby(['QUERY']).count().sort_values(by=['QUERY'])['SCORE_1'].tolist()\n",
    "        read_names = blat_res_part['QUERY'].unique()\n",
    "        read_names.sort()\n",
    "        print('unique names:', len(read_names))\n",
    "        for num, name in zip(amount_list, read_names):\n",
    "            if num > 5:\n",
    "                indexes = blat_res_part[blat_res_part['QUERY'].str.contains(name)].index.tolist()[5:]\n",
    "                blat_res_part.drop(indexes, axis=0, inplace=True)\n",
    "                \n",
    "        print('after:', blat_res_part.shape[0], '\\n')\n",
    "\n",
    "    return blat_res, RM_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3bfd31e-1ec0-4191-aee1-2ed383ae2017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return distanse from the insertion to the repeat nearest end and description \n",
    "# if they overlap - positive number, else - negative\n",
    "def calc_overlap(ins_coord, r_start, r_end):\n",
    "    ins_coord = int(ins_coord)\n",
    "    r_start = int(r_start)\n",
    "    r_end = int(r_end)\n",
    "    if r_start < ins_coord:\n",
    "        if r_end < ins_coord:\n",
    "            return r_end - ins_coord, 'ins after the repeat'  # negative - insertion after the repeat\n",
    "        else:\n",
    "            return min(r_end - ins_coord, ins_coord - r_start), 'ins in the repeat'  # positive - insertion in the repeat\n",
    "    else:\n",
    "        # r_start >= ins_coord\n",
    "        return ins_coord - r_start, 'ins before the repeat'  # negative - insertion before the repeat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a898c34-d74d-47c4-9b7e-b931db426167",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import trange\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.wait import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.support.select import Select\n",
    "\n",
    "def RM_parsing(DIR, file_names, blat_res, RM_res):\n",
    "    service = Service(executable_path=\"/usr/bin/chromedriver\")\n",
    "    chrome_options = Options()\n",
    "    chrome_options.add_argument('--headless')\n",
    "    chrome_options.add_argument('--no-sandbox')\n",
    "    chrome_options.add_argument('--disable-dev-shm-usage')\n",
    "    driver = webdriver.Chrome(service=service, options=chrome_options)\n",
    "    \n",
    "    # go to the RepeatMasker\n",
    "    driver.get(\"https://www.repeatmasker.org/cgi-bin/AnnotationRequest\")\n",
    "    print(driver.title)\n",
    "    \n",
    "    # set Hg38\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(\n",
    "            EC.presence_of_element_located((By.NAME, \"genome\"))\n",
    "        )\n",
    "        dropdown = Select(element)\n",
    "        dropdown.select_by_visible_text(\"Human - Dec 2013 - hg38\")\n",
    "    except:\n",
    "        print(\"can't select genome\")\n",
    "    \n",
    "    \n",
    "    if not os.path.exists(DIR + '/RepeatMasker'):\n",
    "        os.makedirs(DIR + '/RepeatMasker')\n",
    "    \n",
    "    # flank coords annotating\n",
    "    j = 0\n",
    "    EXTENSION = 20\n",
    "    for blat_res_part, RM_res_part in zip(blat_res, RM_res):\n",
    "        blat_res_part = blat_res_part.reset_index()\n",
    "        for i in trange(blat_res_part.shape[0], desc=f'{file_names[j]}'):\n",
    "            # set flank coords and submit\n",
    "            try:\n",
    "                range_input = WebDriverWait(driver, 15).until(\n",
    "                              EC.presence_of_element_located((By.NAME, \"range\")) \n",
    "                              )\n",
    "                #driver.find_element(By.NAME, \"range\")\n",
    "                range_input.clear()\n",
    "    \n",
    "                low_coord = int(blat_res_part.loc[i, 'INS_COORD'] - EXTENSION)\n",
    "                high_coord = int(blat_res_part.loc[i, 'INS_COORD'] + EXTENSION)\n",
    "                chr = blat_res_part.loc[i, 'F_CHR']\n",
    "                coords = f\"{chr}:{low_coord}-{high_coord}\" \n",
    "                #print(coords)\n",
    "                range_input.send_keys(coords)\n",
    "               \n",
    "                button = driver.find_element(By.NAME, \".submit\")\n",
    "                button.click()\n",
    "            except:\n",
    "                print(\"can't pass coords\", blat_res_part.loc[i, 'QUERY'])\n",
    "        \n",
    "            # get results \n",
    "            try:\n",
    "                element = WebDriverWait(driver, 15).until(\n",
    "                          EC.presence_of_element_located((By.TAG_NAME, \"pre\")) \n",
    "                          )\n",
    "                #link = element.find_element(By.TAG_NAME, \"a\").get_attribute(\"href\")\n",
    "                output = element.text.split('\\n')[2:]\n",
    "                #print(output)\n",
    "                for line in output:\n",
    "                    tmp_line = line.split(' ')\n",
    "                    tmp_line = ' '.join(tmp_line).split()\n",
    "                    #print(tmp_line)\n",
    "                    overlapping = calc_overlap(blat_res_part.loc[i, 'INS_COORD'], tmp_line[5], tmp_line[6])\n",
    "                    #print(overlapping)\n",
    "                    RM_res_part.loc[len(RM_res_part.index)] = [blat_res_part.loc[i, 'QUERY'], EXTENSION, blat_res_part.loc[i, 'F_CHR'],\n",
    "                                                               blat_res_part.loc[i, 'INS_COORD'], blat_res_part.loc[i, 'STRAND_1'], tmp_line[9], \n",
    "                                                               tmp_line[8], tmp_line[4], tmp_line[5], tmp_line[6], overlapping[0], overlapping[1], \n",
    "                                                               blat_res_part.loc[i, 'F_LEN'], int(tmp_line[6]) - int(tmp_line[5]) + 1, \n",
    "                                                               blat_res_part.loc[i, 'F_START'], blat_res_part.loc[i, 'F_END'],\n",
    "                                                               blat_res_part.loc[i, 'SCORE_1'], blat_res_part.loc[i, 'QSIZE_1'], \n",
    "                                                               blat_res_part.loc[i, 'SCORE_2'], blat_res_part.loc[i, 'QSIZE_2'],\n",
    "                                                               blat_res_part.loc[i, 'CHROM_1'], blat_res_part.loc[i, 'START_1'], \n",
    "                                                               blat_res_part.loc[i, 'END_1'], blat_res_part.loc[i, 'CHROM_2'], \n",
    "                                                               blat_res_part.loc[i, 'START_2'], blat_res_part.loc[i, 'END_2'],\n",
    "                                                              ]\n",
    "            except:\n",
    "                #print('no matching')\n",
    "                RM_res_part.loc[len(RM_res_part.index)] = [blat_res_part.loc[i, 'QUERY'], EXTENSION, blat_res_part.loc[i, 'F_CHR'],\n",
    "                                                           blat_res_part.loc[i, 'INS_COORD'], blat_res_part.loc[i, 'STRAND_1'], '-', '-', \n",
    "                                                           '-', 0, 0, 0, 'no repeat', blat_res_part.loc[i, 'F_LEN'],\n",
    "                                                           0, blat_res_part.loc[i, 'F_START'], blat_res_part.loc[i, 'F_END'],\n",
    "                                                           blat_res_part.loc[i, 'SCORE_1'], blat_res_part.loc[i, 'QSIZE_1'], \n",
    "                                                           blat_res_part.loc[i, 'SCORE_2'], blat_res_part.loc[i, 'QSIZE_2'],\n",
    "                                                           blat_res_part.loc[i, 'CHROM_1'], blat_res_part.loc[i, 'START_1'], \n",
    "                                                           blat_res_part.loc[i, 'END_1'], blat_res_part.loc[i, 'CHROM_2'], \n",
    "                                                           blat_res_part.loc[i, 'START_2'], blat_res_part.loc[i, 'END_2'],\n",
    "                                                          ]\n",
    "            # get back to request form\n",
    "            driver.back()\n",
    "        RM_res_part.to_csv(DIR + '/RepeatMasker/' + f'{file_names[j]}_RM_res.csv', index=False)\n",
    "        j += 1\n",
    "    \n",
    "    #quit from site\n",
    "    driver.quit()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4412bf-df02-4c68-8bde-2c7cff267be6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Merging with pipeline info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "698dba1a-3f78-4ad1-86e9-509a6e2f66cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging with pipeline info\n",
    "def attach_pipeline_info(filename, df_res_part):\n",
    "    filtered_flank = pd.read_csv(DIR + '/filtred/' + f'{filename}_filtered.csv')\n",
    "    \n",
    "    right_df = filtered_flank[['READNAME', 'CHR', 'POS', 'INS_STRAND', f'{filename}_NUM_READS', 'R1', 'R2']].copy(deep=True)\n",
    "    left_df = df_res_part.copy(deep=True)\n",
    "    \n",
    "    return left_df.merge(right_df, left_on = 'QUERY', right_on = 'READNAME').drop(columns='READNAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6f9c1b72-05d8-4cbd-b02d-99a99e69d61b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_dist(x1, x2):\n",
    "    j = 0\n",
    "    for i in range(len(x1)):\n",
    "        if x1[i] != x2[i]:\n",
    "            j = j + 1\n",
    "    return j\n",
    "\n",
    "# return length of overlapping part\n",
    "def find_overlap_len(str1, str2):\n",
    "    if len(str1) < len(str2):\n",
    "        str_tmp = str1\n",
    "        str1 = str2\n",
    "        str2 = str_tmp\n",
    "    if len(str1) >= len(str2):\n",
    "        delta = len(str1) - len(str2)\n",
    "        for i in range(len(str1)):\n",
    "            if delta - i >= 0:\n",
    "                part1 = str1[i:len(str2) + i]\n",
    "                part2 = str2\n",
    "            else:\n",
    "                part1 = str1[i:]\n",
    "                part2 = str2[0:len(str2) + delta - i]\n",
    "            # print(part1, part2)\n",
    "            if hamming_dist(part1, part2) == 0:\n",
    "                #print(part1, len(part1))\n",
    "                return len(part1)\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c8f7b0ca-e77f-434e-911b-fb373afac65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging with pipeline info\n",
    "# reads overlap calculation\n",
    "def merging(DIR, file_names):\n",
    "    RM_res_list = []\n",
    "    i = 0\n",
    "    for name in file_names:\n",
    "        RM_res_list.append(pd.read_table(DIR + '/RepeatMasker/' + f'{name}_RM_res.csv', sep=','))\n",
    "        RM_res_list[i] = attach_pipeline_info(name, RM_res_list[i])\n",
    "        for j, row in RM_res_list[i].iterrows():\n",
    "            RM_res_list[i].loc[j, 'READS_OVERLAP'] = find_overlap_len(row['R1'], reverse_comp(row['R2']))\n",
    "        RM_res_list[i].to_csv(DIR + '/RepeatMasker/' + f'{name}_RM_res_merged.csv', index=False)\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5705f834-5d69-4b35-baca-1a0ba86510b7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "2e639e5c-4b65-49e3-bfee-6bf25aefb4a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def text_replace(text):\n",
    "    if len(text) == 0:\n",
    "        return ' '\n",
    "    text = text.replace('_', '\\_')\n",
    "    text = text.replace('*', '\\*')\n",
    "    text = text.replace('[', '\\[')\n",
    "    text = text.replace(']', '\\]')\n",
    "    text = text.replace('(', '\\(')\n",
    "    text = text.replace(')', '\\)')\n",
    "    text = text.replace('~', '\\~')\n",
    "    text = text.replace('`', '\\`')\n",
    "    text = text.replace('>', '\\>')\n",
    "    text = text.replace('#', '\\#')\n",
    "    text = text.replace('+', '\\+')\n",
    "    text = text.replace('-', '\\-')\n",
    "    text = text.replace('=', '\\=')\n",
    "    text = text.replace('|', '\\|')\n",
    "    text = text.replace('{', '\\{')\n",
    "    text = text.replace('}', '\\}')\n",
    "    text = text.replace('!', '\\!')\n",
    "    return text\n",
    "\n",
    "def telegram_bot_sendtext(bot_message):\n",
    "    bot_message = text_replace(bot_message)\n",
    "    BOT_TOKEN = \"6608136916:AAHB_1JsP8lsFLD7Uh4WoiajrUFCZToQ2cQ\"\n",
    "    admin_id = str(360072308)\n",
    "    send_text = 'https://api.telegram.org/bot' + BOT_TOKEN + '/sendMessage?chat_id=' + admin_id + '&parse_mode=Markdown&text=' + bot_message\n",
    "    response = requests.get(send_text)\n",
    "    return response.json()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "761b34db-4dbb-4e0f-8b66-48ffdec1590b",
   "metadata": {},
   "source": [
    "### Drop questionable insertions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c27a97c4-861b-4ad1-9ea3-e2dc61c49957",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "\n",
    "def is_questionable(df_row, is_line):\n",
    "    # no repeat - ok\n",
    "    if df_row['DESC'] == 'no repeat':\n",
    "        return False\n",
    "    if is_line:\n",
    "        # LINE\n",
    "        # ins in the repeat\n",
    "        if df_row['DESC'] == 'ins in the repeat':\n",
    "            # repeat is in L1HS, <=L1PA7 => drop\n",
    "            if df_row['REPEAT_NAME'] == 'L1HS' or \\\n",
    "               ( bool(re.search('L1PA[1-7]', df_row['REPEAT_NAME'])) and (not bool(re.search('L1PA[1-9][0-9]', df_row['REPEAT_NAME']))) ):\n",
    "                return True\n",
    "        # ins near the repeat\n",
    "        else:\n",
    "            # near <=L1PA7 or L1HS in 20nt => drop\n",
    "            if ( df_row['REPEAT_NAME'] == 'L1HS' or \\\n",
    "                (bool(re.search('L1PA[1-7]', df_row['REPEAT_NAME'])) and (not bool(re.search('L1PA[1-9][0-9]', df_row['REPEAT_NAME'])))) \\\n",
    "               ) and abs(df_row['INS_OVERLAP']) <= 20:\n",
    "                return True\n",
    "    else:\n",
    "        # Alu\n",
    "        # ins in the repeat\n",
    "        if df_row['DESC'] == 'ins in the repeat':\n",
    "            # repeat is in L1HS, <=L1PA7 or Alu => drop\n",
    "            if df_row['REPEAT_NAME'] == 'L1HS' or \\\n",
    "               ( bool(re.search('L1PA[1-7]', df_row['REPEAT_NAME'])) and (not bool(re.search('L1PA[1-9][0-9]', df_row['REPEAT_NAME']))) ) or \\\n",
    "               bool(re.search('Alu', df_row['REPEAT_NAME'])):\n",
    "                return True\n",
    "        # ins near the repeat\n",
    "        else:\n",
    "            # near Alu in 5nt => drop\n",
    "            if bool(re.search('Alu', df_row['REPEAT_NAME'])) and abs(df_row['INS_OVERLAP']) <= 5:\n",
    "                return True\n",
    "    return False\n",
    "\n",
    "\n",
    "\n",
    "def drop_questionable(DIR, file_names, is_line):\n",
    "    \n",
    "    for name in file_names:\n",
    "        df = pd.read_table(DIR + '/RepeatMasker/' + f'{name}_RM_res_merged.csv', sep=',')\n",
    "\n",
    "        read_names = df['QUERY'].unique()\n",
    "        print(name, ' candidates num:', len(read_names))\n",
    "        num = 0\n",
    "        for r_name in read_names:\n",
    "            df_tmp = df[df['QUERY'] == r_name].copy(deep=True)\n",
    "            # uniaue match\n",
    "            if len(df_tmp['INS_COORD_B'].unique()) == 1:\n",
    "                # 1 repeat or no repeat\n",
    "                if df_tmp.shape[0] == 1:\n",
    "                    if is_questionable(df_tmp.iloc[0], is_line):\n",
    "                        df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                        num += 1\n",
    "                        #print(r_name)\n",
    "                # 2 repeats in the area of the ins\n",
    "                else:\n",
    "                    for i, row in df_tmp.iterrows():\n",
    "                        if is_questionable(row, is_line):\n",
    "                            df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                            num += 1\n",
    "                            #print(r_name)\n",
    "                            break\n",
    "            # not unique => drop\n",
    "            else:\n",
    "                df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                num += 1\n",
    "                #print(r_name)\n",
    "        print(name, ' dropped:', num)\n",
    "        df.to_csv(DIR + '/result/' + f'{name}_result.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e7cb77e8-6adb-41d0-950b-ce6bf9ea62f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC60 ---------------------------------------------------\n",
      "metatable shape: (1907, 32)\n",
      "sample file names: ['BC60-L_S106_ME_L001__', 'MN-BC60-L_S108_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 828   including test: 0\n",
      "somatic and quasi-somatic: (1079, 32)   including test: 0 \n",
      "\n",
      "BC60-L_S106_ME_L001__ 441   test: 0\n",
      "MN-BC60-L_S108_ME_L001__ 638   test: 0\n",
      "overall num of somatic insertions:  1079 \n",
      "\n",
      "BC60-L_S106_ME_L001__  candidates num: 23\n",
      "BC60-L_S106_ME_L001__  dropped: 4\n",
      "MN-BC60-L_S108_ME_L001__  candidates num: 35\n",
      "MN-BC60-L_S108_ME_L001__  dropped: 8\n",
      "BC66 ---------------------------------------------------\n",
      "metatable shape: (1927, 32)\n",
      "sample file names: ['BC66-L_S102_ME_L001__', 'MN-BC66-L_S104_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 842   including test: 0\n",
      "somatic and quasi-somatic: (1085, 32)   including test: 0 \n",
      "\n",
      "BC66-L_S102_ME_L001__ 382   test: 0\n",
      "MN-BC66-L_S104_ME_L001__ 703   test: 0\n",
      "overall num of somatic insertions:  1085 \n",
      "\n",
      "BC66-L_S102_ME_L001__  candidates num: 21\n",
      "BC66-L_S102_ME_L001__  dropped: 4\n",
      "MN-BC66-L_S104_ME_L001__  candidates num: 33\n",
      "MN-BC66-L_S104_ME_L001__  dropped: 9\n",
      "BC70 ---------------------------------------------------\n",
      "metatable shape: (1139, 32)\n",
      "sample file names: ['BC70-L_S142_ME_L001__', 'MN-BC70-L_S144_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 178   including test: 0\n",
      "somatic and quasi-somatic: (961, 32)   including test: 0 \n",
      "\n",
      "BC70-L_S142_ME_L001__ 197   test: 0\n",
      "MN-BC70-L_S144_ME_L001__ 764   test: 0\n",
      "overall num of somatic insertions:  961 \n",
      "\n",
      "BC70-L_S142_ME_L001__  candidates num: 18\n",
      "BC70-L_S142_ME_L001__  dropped: 13\n",
      "MN-BC70-L_S144_ME_L001__  candidates num: 35\n",
      "MN-BC70-L_S144_ME_L001__  dropped: 5\n",
      "BLC1 ---------------------------------------------------\n",
      "metatable shape: (1684, 32)\n",
      "sample file names: ['BLC1-L_S74_ME_L001__', 'MN-BLC1-L_S76_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 749   including test: 0\n",
      "somatic and quasi-somatic: (935, 32)   including test: 0 \n",
      "\n",
      "BLC1-L_S74_ME_L001__ 634   test: 0\n",
      "MN-BLC1-L_S76_ME_L001__ 301   test: 0\n",
      "overall num of somatic insertions:  935 \n",
      "\n",
      "BLC1-L_S74_ME_L001__  candidates num: 31\n",
      "BLC1-L_S74_ME_L001__  dropped: 11\n",
      "MN-BLC1-L_S76_ME_L001__  candidates num: 13\n",
      "MN-BLC1-L_S76_ME_L001__  dropped: 6\n",
      "CC107 ---------------------------------------------------\n",
      "metatable shape: (1337, 32)\n",
      "sample file names: ['CC107-L_S130_ME_L001__', 'MN-CC107-L_S132_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 573   including test: 0\n",
      "somatic and quasi-somatic: (764, 32)   including test: 0 \n",
      "\n",
      "CC107-L_S130_ME_L001__ 417   test: 0\n",
      "MN-CC107-L_S132_ME_L001__ 347   test: 0\n",
      "overall num of somatic insertions:  764 \n",
      "\n",
      "CC107-L_S130_ME_L001__  candidates num: 24\n",
      "CC107-L_S130_ME_L001__  dropped: 5\n",
      "MN-CC107-L_S132_ME_L001__  candidates num: 15\n",
      "MN-CC107-L_S132_ME_L001__  dropped: 4\n",
      "CC127-2 ---------------------------------------------------\n",
      "metatable shape: (1724, 32)\n",
      "sample file names: ['CC127-2-L_S6_ME_L001__', 'MN-CC127-L_S8_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 593   including test: 0\n",
      "somatic and quasi-somatic: (1131, 32)   including test: 0 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__ 802   test: 0\n",
      "MN-CC127-L_S8_ME_L001__ 329   test: 0\n",
      "overall num of somatic insertions:  1131 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__  candidates num: 67\n",
      "CC127-2-L_S6_ME_L001__  dropped: 16\n",
      "MN-CC127-L_S8_ME_L001__  candidates num: 22\n",
      "MN-CC127-L_S8_ME_L001__  dropped: 8\n",
      "CC18 ---------------------------------------------------\n",
      "metatable shape: (1557, 32)\n",
      "sample file names: ['CC18-L_S118_ME_L001__', 'MN-CC18-L_S120_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 646   including test: 0\n",
      "somatic and quasi-somatic: (911, 32)   including test: 0 \n",
      "\n",
      "CC18-L_S118_ME_L001__ 385   test: 0\n",
      "MN-CC18-L_S120_ME_L001__ 526   test: 0\n",
      "overall num of somatic insertions:  911 \n",
      "\n",
      "CC18-L_S118_ME_L001__  candidates num: 17\n",
      "CC18-L_S118_ME_L001__  dropped: 4\n",
      "MN-CC18-L_S120_ME_L001__  candidates num: 41\n",
      "MN-CC18-L_S120_ME_L001__  dropped: 4\n",
      "CC46 ---------------------------------------------------\n",
      "metatable shape: (1448, 32)\n",
      "sample file names: ['CC46-L_S134_ME_L001__', 'MN-CC46-L_S136_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 541   including test: 0\n",
      "somatic and quasi-somatic: (907, 32)   including test: 0 \n",
      "\n",
      "CC46-L_S134_ME_L001__ 422   test: 0\n",
      "MN-CC46-L_S136_ME_L001__ 485   test: 0\n",
      "overall num of somatic insertions:  907 \n",
      "\n",
      "CC46-L_S134_ME_L001__  candidates num: 17\n",
      "CC46-L_S134_ME_L001__  dropped: 10\n",
      "MN-CC46-L_S136_ME_L001__  candidates num: 24\n",
      "MN-CC46-L_S136_ME_L001__  dropped: 3\n",
      "CC5 ---------------------------------------------------\n",
      "metatable shape: (1677, 32)\n",
      "sample file names: ['CC5-L_S146_ME_L001__', 'MN-CC5-L_S148_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 600   including test: 0\n",
      "somatic and quasi-somatic: (1077, 32)   including test: 0 \n",
      "\n",
      "CC5-L_S146_ME_L001__ 427   test: 0\n",
      "MN-CC5-L_S148_ME_L001__ 650   test: 0\n",
      "overall num of somatic insertions:  1077 \n",
      "\n",
      "CC5-L_S146_ME_L001__  candidates num: 20\n",
      "CC5-L_S146_ME_L001__  dropped: 8\n",
      "MN-CC5-L_S148_ME_L001__  candidates num: 29\n",
      "MN-CC5-L_S148_ME_L001__  dropped: 6\n",
      "CC82 ---------------------------------------------------\n",
      "metatable shape: (2129, 32)\n",
      "sample file names: ['CC82-L_S10_ME_L001__', 'MN-CC82-L_S12_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 690   including test: 0\n",
      "somatic and quasi-somatic: (1439, 32)   including test: 0 \n",
      "\n",
      "CC82-L_S10_ME_L001__ 1051   test: 0\n",
      "MN-CC82-L_S12_ME_L001__ 388   test: 0\n",
      "overall num of somatic insertions:  1439 \n",
      "\n",
      "CC82-L_S10_ME_L001__  candidates num: 59\n",
      "CC82-L_S10_ME_L001__  dropped: 12\n",
      "MN-CC82-L_S12_ME_L001__  candidates num: 22\n",
      "MN-CC82-L_S12_ME_L001__  dropped: 7\n",
      "CC94 ---------------------------------------------------\n",
      "metatable shape: (1011, 32)\n",
      "sample file names: ['CC94-L_S14_ME_L001__', 'MN-CC94-L_S16_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 265   including test: 0\n",
      "somatic and quasi-somatic: (746, 32)   including test: 0 \n",
      "\n",
      "CC94-L_S14_ME_L001__ 644   test: 0\n",
      "MN-CC94-L_S16_ME_L001__ 102   test: 0\n",
      "overall num of somatic insertions:  746 \n",
      "\n",
      "CC94-L_S14_ME_L001__  candidates num: 28\n",
      "CC94-L_S14_ME_L001__  dropped: 6\n",
      "MN-CC94-L_S16_ME_L001__  candidates num: 7\n",
      "MN-CC94-L_S16_ME_L001__  dropped: 5\n",
      "CC96 ---------------------------------------------------\n",
      "metatable shape: (2429, 32)\n",
      "sample file names: ['CC96-L_S66_ME_L001__', 'MN-CC96-L_S68_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 955   including test: 0\n",
      "somatic and quasi-somatic: (1474, 32)   including test: 0 \n",
      "\n",
      "CC96-L_S66_ME_L001__ 596   test: 0\n",
      "MN-CC96-L_S68_ME_L001__ 878   test: 0\n",
      "overall num of somatic insertions:  1474 \n",
      "\n",
      "CC96-L_S66_ME_L001__  candidates num: 36\n",
      "CC96-L_S66_ME_L001__  dropped: 6\n",
      "MN-CC96-L_S68_ME_L001__  candidates num: 96\n",
      "MN-CC96-L_S68_ME_L001__  dropped: 13\n",
      "CC99 ---------------------------------------------------\n",
      "metatable shape: (1338, 32)\n",
      "sample file names: ['CC99-L_S70_ME_L001__', 'MN-CC99-L_S72_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 551   including test: 0\n",
      "somatic and quasi-somatic: (787, 32)   including test: 0 \n",
      "\n",
      "CC99-L_S70_ME_L001__ 459   test: 0\n",
      "MN-CC99-L_S72_ME_L001__ 328   test: 0\n",
      "overall num of somatic insertions:  787 \n",
      "\n",
      "CC99-L_S70_ME_L001__  candidates num: 25\n",
      "CC99-L_S70_ME_L001__  dropped: 4\n",
      "MN-CC99-L_S72_ME_L001__  candidates num: 11\n",
      "MN-CC99-L_S72_ME_L001__  dropped: 5\n",
      "GC39 ---------------------------------------------------\n",
      "metatable shape: (1935, 32)\n",
      "sample file names: ['GC39-L_S62_ME_L001__', 'MN-GC39-L_S64_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 889   including test: 0\n",
      "somatic and quasi-somatic: (1046, 32)   including test: 0 \n",
      "\n",
      "GC39-L_S62_ME_L001__ 524   test: 0\n",
      "MN-GC39-L_S64_ME_L001__ 522   test: 0\n",
      "overall num of somatic insertions:  1046 \n",
      "\n",
      "GC39-L_S62_ME_L001__  candidates num: 25\n",
      "GC39-L_S62_ME_L001__  dropped: 8\n",
      "MN-GC39-L_S64_ME_L001__  candidates num: 37\n",
      "MN-GC39-L_S64_ME_L001__  dropped: 7\n",
      "GC41 ---------------------------------------------------\n",
      "metatable shape: (1578, 32)\n",
      "sample file names: ['GC41-L_S154_ME_L001__', 'MN-GC41-L_S156_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 738   including test: 0\n",
      "somatic and quasi-somatic: (840, 32)   including test: 0 \n",
      "\n",
      "GC41-L_S154_ME_L001__ 397   test: 0\n",
      "MN-GC41-L_S156_ME_L001__ 443   test: 0\n",
      "overall num of somatic insertions:  840 \n",
      "\n",
      "GC41-L_S154_ME_L001__  candidates num: 18\n",
      "GC41-L_S154_ME_L001__  dropped: 5\n",
      "MN-GC41-L_S156_ME_L001__  candidates num: 21\n",
      "MN-GC41-L_S156_ME_L001__  dropped: 5\n",
      "LS1 ---------------------------------------------------\n",
      "metatable shape: (698, 32)\n",
      "sample file names: ['LS1-L_S138_ME_L001__', 'MN-LS1-L_S140_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 71   including test: 0\n",
      "somatic and quasi-somatic: (627, 32)   including test: 0 \n",
      "\n",
      "LS1-L_S138_ME_L001__ 559   test: 0\n",
      "MN-LS1-L_S140_ME_L001__ 68   test: 0\n",
      "overall num of somatic insertions:  627 \n",
      "\n",
      "LS1-L_S138_ME_L001__  candidates num: 22\n",
      "LS1-L_S138_ME_L001__  dropped: 7\n",
      "MN-LS1-L_S140_ME_L001__  candidates num: 5\n",
      "MN-LS1-L_S140_ME_L001__  dropped: 2\n",
      "LS6 ---------------------------------------------------\n",
      "metatable shape: (1556, 32)\n",
      "sample file names: ['LS6-L_S46_ME_L001__', 'MN-LS6-L_S48_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 691   including test: 0\n",
      "somatic and quasi-somatic: (865, 32)   including test: 0 \n",
      "\n",
      "LS6-L_S46_ME_L001__ 551   test: 0\n",
      "MN-LS6-L_S48_ME_L001__ 314   test: 0\n",
      "overall num of somatic insertions:  865 \n",
      "\n",
      "LS6-L_S46_ME_L001__  candidates num: 28\n",
      "LS6-L_S46_ME_L001__  dropped: 6\n",
      "MN-LS6-L_S48_ME_L001__  candidates num: 14\n",
      "MN-LS6-L_S48_ME_L001__  dropped: 2\n",
      "M14 ---------------------------------------------------\n",
      "metatable shape: (1812, 32)\n",
      "sample file names: ['M14-L_S94_ME_L001__', 'MN-M14-L_S96_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 822   including test: 0\n",
      "somatic and quasi-somatic: (990, 32)   including test: 0 \n",
      "\n",
      "M14-L_S94_ME_L001__ 421   test: 0\n",
      "MN-M14-L_S96_ME_L001__ 569   test: 0\n",
      "overall num of somatic insertions:  990 \n",
      "\n",
      "M14-L_S94_ME_L001__  candidates num: 17\n",
      "M14-L_S94_ME_L001__  dropped: 7\n",
      "MN-M14-L_S96_ME_L001__  candidates num: 21\n",
      "MN-M14-L_S96_ME_L001__  dropped: 2\n",
      "M21 ---------------------------------------------------\n",
      "metatable shape: (1729, 32)\n",
      "sample file names: ['M21-L_S82_ME_L001__', 'MN-M21-L_S84_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 823   including test: 0\n",
      "somatic and quasi-somatic: (906, 32)   including test: 0 \n",
      "\n",
      "M21-L_S82_ME_L001__ 461   test: 0\n",
      "MN-M21-L_S84_ME_L001__ 445   test: 0\n",
      "overall num of somatic insertions:  906 \n",
      "\n",
      "M21-L_S82_ME_L001__  candidates num: 34\n",
      "M21-L_S82_ME_L001__  dropped: 4\n",
      "MN-M21-L_S84_ME_L001__  candidates num: 24\n",
      "MN-M21-L_S84_ME_L001__  dropped: 4\n",
      "M22 ---------------------------------------------------\n",
      "metatable shape: (2134, 32)\n",
      "sample file names: ['M22-L_S98_ME_L001__', 'MN-M22-L_S100_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 990   including test: 0\n",
      "somatic and quasi-somatic: (1144, 32)   including test: 0 \n",
      "\n",
      "M22-L_S98_ME_L001__ 555   test: 0\n",
      "MN-M22-L_S100_ME_L001__ 589   test: 0\n",
      "overall num of somatic insertions:  1144 \n",
      "\n",
      "M22-L_S98_ME_L001__  candidates num: 35\n",
      "M22-L_S98_ME_L001__  dropped: 7\n",
      "MN-M22-L_S100_ME_L001__  candidates num: 24\n",
      "MN-M22-L_S100_ME_L001__  dropped: 9\n",
      "NC5 ---------------------------------------------------\n",
      "metatable shape: (1479, 32)\n",
      "sample file names: ['MN-NC5-L_S88_ME_L001__', 'NC5-L_S86_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 580   including test: 0\n",
      "somatic and quasi-somatic: (899, 32)   including test: 0 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__ 236   test: 0\n",
      "NC5-L_S86_ME_L001__ 663   test: 0\n",
      "overall num of somatic insertions:  899 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__  candidates num: 13\n",
      "MN-NC5-L_S88_ME_L001__  dropped: 5\n",
      "NC5-L_S86_ME_L001__  candidates num: 39\n",
      "NC5-L_S86_ME_L001__  dropped: 7\n",
      "OC14 ---------------------------------------------------\n",
      "metatable shape: (1712, 32)\n",
      "sample file names: ['MN-OC14-L_S112_ME_L001__', 'OC14-L_S110_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 764   including test: 0\n",
      "somatic and quasi-somatic: (948, 32)   including test: 0 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__ 503   test: 0\n",
      "OC14-L_S110_ME_L001__ 445   test: 0\n",
      "overall num of somatic insertions:  948 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__  candidates num: 33\n",
      "MN-OC14-L_S112_ME_L001__  dropped: 9\n",
      "OC14-L_S110_ME_L001__  candidates num: 27\n",
      "OC14-L_S110_ME_L001__  dropped: 8\n",
      "OC33 ---------------------------------------------------\n",
      "metatable shape: (815, 32)\n",
      "sample file names: ['MN-OC33-L_S116_ME_L001__', 'OC33-L_S114_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 242   including test: 0\n",
      "somatic and quasi-somatic: (573, 32)   including test: 0 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__ 213   test: 0\n",
      "OC33-L_S114_ME_L001__ 360   test: 0\n",
      "overall num of somatic insertions:  573 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__  candidates num: 8\n",
      "MN-OC33-L_S116_ME_L001__  dropped: 2\n",
      "OC33-L_S114_ME_L001__  candidates num: 20\n",
      "OC33-L_S114_ME_L001__  dropped: 3\n",
      "OC40 ---------------------------------------------------\n",
      "metatable shape: (1450, 32)\n",
      "sample file names: ['MN-OC40-L_S80_ME_L001__', 'OC40-L_S78_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 595   including test: 0\n",
      "somatic and quasi-somatic: (855, 32)   including test: 0 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__ 272   test: 0\n",
      "OC40-L_S78_ME_L001__ 583   test: 0\n",
      "overall num of somatic insertions:  855 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__  candidates num: 6\n",
      "MN-OC40-L_S80_ME_L001__  dropped: 1\n",
      "OC40-L_S78_ME_L001__  candidates num: 35\n",
      "OC40-L_S78_ME_L001__  dropped: 9\n",
      "OC42 ---------------------------------------------------\n",
      "metatable shape: (382, 32)\n",
      "sample file names: ['MN-OC42-L_S40_ME_L001__', 'OC42-L_S38_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 30   including test: 0\n",
      "somatic and quasi-somatic: (352, 32)   including test: 0 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__ 152   test: 0\n",
      "OC42-L_S38_ME_L001__ 200   test: 0\n",
      "overall num of somatic insertions:  352 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__  candidates num: 5\n",
      "MN-OC42-L_S40_ME_L001__  dropped: 2\n",
      "OC42-L_S38_ME_L001__  candidates num: 12\n",
      "OC42-L_S38_ME_L001__  dropped: 2\n",
      "OC48 ---------------------------------------------------\n",
      "metatable shape: (1798, 32)\n",
      "sample file names: ['MN-OC48-L_S44_ME_L001__', 'OC48-L_S42_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 699   including test: 0\n",
      "somatic and quasi-somatic: (1099, 32)   including test: 0 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__ 365   test: 0\n",
      "OC48-L_S42_ME_L001__ 734   test: 0\n",
      "overall num of somatic insertions:  1099 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__  candidates num: 19\n",
      "MN-OC48-L_S44_ME_L001__  dropped: 5\n",
      "OC48-L_S42_ME_L001__  candidates num: 39\n",
      "OC48-L_S42_ME_L001__  dropped: 8\n",
      "OC50 ---------------------------------------------------\n",
      "metatable shape: (1873, 32)\n",
      "sample file names: ['MN-OC50-L_S128_ME_L001__', 'OC50-L_S126_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 841   including test: 0\n",
      "somatic and quasi-somatic: (1032, 32)   including test: 0 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__ 430   test: 0\n",
      "OC50-L_S126_ME_L001__ 602   test: 0\n",
      "overall num of somatic insertions:  1032 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__  candidates num: 25\n",
      "MN-OC50-L_S128_ME_L001__  dropped: 8\n",
      "OC50-L_S126_ME_L001__  candidates num: 36\n",
      "OC50-L_S126_ME_L001__  dropped: 6\n",
      "OC54 ---------------------------------------------------\n",
      "metatable shape: (1878, 32)\n",
      "sample file names: ['MN-OC54-L_S56_ME_L001__', 'OC54-L_S54_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 854   including test: 0\n",
      "somatic and quasi-somatic: (1024, 32)   including test: 0 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__ 453   test: 0\n",
      "OC54-L_S54_ME_L001__ 571   test: 0\n",
      "overall num of somatic insertions:  1024 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__  candidates num: 16\n",
      "MN-OC54-L_S56_ME_L001__  dropped: 4\n",
      "OC54-L_S54_ME_L001__  candidates num: 32\n",
      "OC54-L_S54_ME_L001__  dropped: 8\n",
      "OT1 ---------------------------------------------------\n",
      "metatable shape: (1157, 32)\n",
      "sample file names: ['MN-OT1-L_S152_ME_L001__', 'OT1-L_S150_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 505   including test: 0\n",
      "somatic and quasi-somatic: (652, 32)   including test: 0 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__ 303   test: 0\n",
      "OT1-L_S150_ME_L001__ 349   test: 0\n",
      "overall num of somatic insertions:  652 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__  candidates num: 9\n",
      "MN-OT1-L_S152_ME_L001__  dropped: 3\n",
      "OT1-L_S150_ME_L001__  candidates num: 23\n",
      "OT1-L_S150_ME_L001__  dropped: 7\n",
      "PC12 ---------------------------------------------------\n",
      "metatable shape: (2150, 32)\n",
      "sample file names: ['MN-PC12-L_S24_ME_L001__', 'PC12-L_S22_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 541   including test: 0\n",
      "somatic and quasi-somatic: (1609, 32)   including test: 0 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__ 342   test: 0\n",
      "PC12-L_S22_ME_L001__ 1267   test: 0\n",
      "overall num of somatic insertions:  1609 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__  candidates num: 9\n",
      "MN-PC12-L_S24_ME_L001__  dropped: 2\n",
      "PC12-L_S22_ME_L001__  candidates num: 83\n",
      "PC12-L_S22_ME_L001__  dropped: 15\n",
      "PC14 ---------------------------------------------------\n",
      "metatable shape: (1567, 32)\n",
      "sample file names: ['MN-PC14-L_S36_ME_L001__', 'PC14-L_S34_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 346   including test: 0\n",
      "somatic and quasi-somatic: (1221, 32)   including test: 0 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__ 206   test: 0\n",
      "PC14-L_S34_ME_L001__ 1015   test: 0\n",
      "overall num of somatic insertions:  1221 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__  candidates num: 10\n",
      "MN-PC14-L_S36_ME_L001__  dropped: 7\n",
      "PC14-L_S34_ME_L001__  candidates num: 67\n",
      "PC14-L_S34_ME_L001__  dropped: 8\n",
      "PC21 ---------------------------------------------------\n",
      "metatable shape: (1596, 32)\n",
      "sample file names: ['MN-PC21-L_S20_ME_L001__', 'PC21-L_S18_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 660   including test: 0\n",
      "somatic and quasi-somatic: (936, 32)   including test: 0 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__ 597   test: 0\n",
      "PC21-L_S18_ME_L001__ 339   test: 0\n",
      "overall num of somatic insertions:  936 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__  candidates num: 36\n",
      "MN-PC21-L_S20_ME_L001__  dropped: 13\n",
      "PC21-L_S18_ME_L001__  candidates num: 23\n",
      "PC21-L_S18_ME_L001__  dropped: 4\n",
      "PC32 ---------------------------------------------------\n",
      "metatable shape: (1964, 32)\n",
      "sample file names: ['MN-PC32-L_S32_ME_L001__', 'PC32-L_S30_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 178   including test: 0\n",
      "somatic and quasi-somatic: (1786, 32)   including test: 0 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__ 244   test: 0\n",
      "PC32-L_S30_ME_L001__ 1542   test: 0\n",
      "overall num of somatic insertions:  1786 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__  candidates num: 9\n",
      "MN-PC32-L_S32_ME_L001__  dropped: 0\n",
      "PC32-L_S30_ME_L001__  candidates num: 89\n",
      "PC32-L_S30_ME_L001__  dropped: 17\n",
      "SkC6 ---------------------------------------------------\n",
      "metatable shape: (1896, 32)\n",
      "sample file names: ['MN-SkC6-L_S60_ME_L001__', 'SkC6-L_S58_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 804   including test: 0\n",
      "somatic and quasi-somatic: (1092, 32)   including test: 0 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__ 582   test: 0\n",
      "SkC6-L_S58_ME_L001__ 510   test: 0\n",
      "overall num of somatic insertions:  1092 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__  candidates num: 34\n",
      "MN-SkC6-L_S60_ME_L001__  dropped: 6\n",
      "SkC6-L_S58_ME_L001__  candidates num: 37\n",
      "SkC6-L_S58_ME_L001__  dropped: 1\n",
      "StS1 ---------------------------------------------------\n",
      "metatable shape: (1933, 32)\n",
      "sample file names: ['MN-StS1-L_S92_ME_L001__', 'StS1-L_S90_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 808   including test: 0\n",
      "somatic and quasi-somatic: (1125, 32)   including test: 0 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__ 473   test: 0\n",
      "StS1-L_S90_ME_L001__ 652   test: 0\n",
      "overall num of somatic insertions:  1125 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__  candidates num: 16\n",
      "MN-StS1-L_S92_ME_L001__  dropped: 4\n",
      "StS1-L_S90_ME_L001__  candidates num: 36\n",
      "StS1-L_S90_ME_L001__  dropped: 9\n",
      "UC2 ---------------------------------------------------\n",
      "metatable shape: (1364, 32)\n",
      "sample file names: ['MN-UC2-L_S124_ME_L001__', 'UC2-L_S122_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 604   including test: 0\n",
      "somatic and quasi-somatic: (760, 32)   including test: 0 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__ 254   test: 0\n",
      "UC2-L_S122_ME_L001__ 506   test: 0\n",
      "overall num of somatic insertions:  760 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__  candidates num: 11\n",
      "MN-UC2-L_S124_ME_L001__  dropped: 5\n",
      "UC2-L_S122_ME_L001__  candidates num: 24\n",
      "UC2-L_S122_ME_L001__  dropped: 5\n"
     ]
    }
   ],
   "source": [
    "DIR = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/AGCT/'\n",
    "os.chdir(DIR)\n",
    "\n",
    "dirs = os.listdir()\n",
    "for dirname in dirs:\n",
    "    print(dirname, '---------------------------------------------------')\n",
    "    curr_dir = DIR + dirname \n",
    "    if not os.path.isdir(curr_dir):\n",
    "        continue\n",
    "    os.chdir(curr_dir)\n",
    "    \n",
    "    if not os.path.exists(curr_dir + '/nofiltration/result'):\n",
    "        os.makedirs(curr_dir + '/nofiltration/result')\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir)\n",
    "    # drop\n",
    "    drop_questionable(curr_dir + '/nofiltration', file_names, is_line=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea975fec-b6ac-439a-9eb9-67904652447c",
   "metadata": {},
   "source": [
    "### Tumor norm matches "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0620e0b4-2fed-48d3-937e-7cecd37d1aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# вставки в раке Алю1 - вставки в норме Рса; \n",
    "# Вставки в раке Рса - вставки в норме Алю1, для того же пациента\n",
    "def check_matches_tumor_mn(curr_dir1, curr_dir2, file_names, dirname, filter, del_matches):\n",
    "\n",
    "    window = 20\n",
    "    num_matched = [dirname]\n",
    "\n",
    "    if file_names[0].find('MN-') != -1:\n",
    "        tmp = str(file_names[1])\n",
    "        file_names[1] = str(file_names[0])\n",
    "        file_names[0] = str(tmp)\n",
    "    # 0 - tumor, 1 - matched norm\n",
    "    \n",
    "    filtration = ''\n",
    "    if not filter:\n",
    "        filtration = '/nofiltration'\n",
    "    #\n",
    "    df_tumor_Alu1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{file_names[0]}_result.csv', sep=',')\n",
    "    df_tumor_Rsa1 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{file_names[0]}_result.csv', sep=',')\n",
    "    df_norm_Alu1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{file_names[1]}_result.csv', sep=',')\n",
    "    df_norm_Rsa1 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{file_names[1]}_result.csv', sep=',')\n",
    "\n",
    "    df_t_Alu1_mn_Rsa1 = pd.DataFrame(columns=df_tumor_Alu1.columns)\n",
    "    for i, row1 in df_tumor_Alu1.iterrows():\n",
    "        for j, row2 in df_norm_Rsa1.iterrows():\n",
    "            if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= window):\n",
    "                if del_matches:\n",
    "                    df_tumor_Alu1.drop(i, axis=0, inplace=True)\n",
    "                    df_norm_Rsa1.drop(j, axis=0, inplace=True)\n",
    "                df_t_Alu1_mn_Rsa1.loc[len(df_t_Alu1_mn_Rsa1.index)] = row1\n",
    "\n",
    "    df_t_Rsa1_mn_Alu1 = pd.DataFrame(columns=df_tumor_Alu1.columns)\n",
    "    for i, row1 in df_tumor_Rsa1.iterrows():\n",
    "        for j, row2 in df_norm_Alu1.iterrows():\n",
    "            if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= window):\n",
    "                if del_matches:\n",
    "                    df_tumor_Rsa1.drop(i, axis=0, inplace=True)\n",
    "                    df_norm_Alu1.drop(j, axis=0, inplace=True)\n",
    "                df_t_Rsa1_mn_Alu1.loc[len(df_t_Rsa1_mn_Alu1.index)] = row1\n",
    "\n",
    "    if del_matches:\n",
    "        df_tumor_Alu1.to_csv(curr_dir1 + filtration + '/result/' + f'{file_names[0]}_result.csv', index=False)\n",
    "        df_tumor_Rsa1.to_csv(curr_dir2 + filtration + '/result/' + f'{file_names[0]}_result.csv', index=False)\n",
    "        df_norm_Alu1.to_csv(curr_dir1 + filtration + '/result/' + f'{file_names[1]}_result.csv', index=False)\n",
    "        df_norm_Rsa1.to_csv(curr_dir2 + filtration + '/result/' + f'{file_names[1]}_result.csv', index=False)\n",
    "    \n",
    "    num_matched.append(len(df_t_Alu1_mn_Rsa1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_t_Rsa1_mn_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_tumor_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_tumor_Rsa1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_norm_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_norm_Rsa1['QUERY'].unique()))\n",
    "    \n",
    "    return num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fabd6f39-47e3-480c-8ee3-454054513023",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>TUMOR_Alu1_MN_HaeIII</th>\n",
       "      <th>TUMOR_HaeIII_MN_Alu1</th>\n",
       "      <th>TUMOR_Alu1</th>\n",
       "      <th>TUMOR_HaeIII</th>\n",
       "      <th>MN_Alu1</th>\n",
       "      <th>MN_HaeIII</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BC60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BC66</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC70</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CC107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CC127-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>38</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CC18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>37</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CC46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CC82</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CC94</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CC96</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "      <td>83</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CC99</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GC39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>30</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GC41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LS6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OC14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>24</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OC33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OC40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OC42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OC48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OC50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OC54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OT1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PC12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PC14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PC21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PC32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>34</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SkC6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>14</td>\n",
       "      <td>28</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>StS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>UC2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NAME  TUMOR_Alu1_MN_HaeIII  TUMOR_HaeIII_MN_Alu1  TUMOR_Alu1  \\\n",
       "0      BC60                     0                     0          19   \n",
       "1      BC66                     0                     0          17   \n",
       "2      BC70                     0                     0           5   \n",
       "3      BLC1                     0                     0          20   \n",
       "4     CC107                     0                     0          19   \n",
       "5   CC127-2                     0                     0          51   \n",
       "6      CC18                     0                     0          13   \n",
       "7      CC46                     0                     0           7   \n",
       "8       CC5                     0                     0          12   \n",
       "9      CC82                     0                     0          47   \n",
       "10     CC94                     0                     0          22   \n",
       "11     CC96                     0                     0          30   \n",
       "12     CC99                     0                     0          21   \n",
       "13     GC39                     0                     0          17   \n",
       "14     GC41                     0                     0          13   \n",
       "15      LS1                     0                     0          15   \n",
       "16      LS6                     0                     0          22   \n",
       "17      M14                     0                     0          10   \n",
       "18      M21                     0                     0          29   \n",
       "19      M22                     0                     0          28   \n",
       "20      NC5                     0                     0          32   \n",
       "21     OC14                     0                     0          19   \n",
       "22     OC33                     0                     0          17   \n",
       "23     OC40                     0                     0          26   \n",
       "24     OC42                     0                     0           9   \n",
       "25     OC48                     0                     0          31   \n",
       "26     OC50                     0                     0          30   \n",
       "27     OC54                     0                     0          24   \n",
       "28      OT1                     0                     0          16   \n",
       "29     PC12                     0                     0          68   \n",
       "30     PC14                     0                     0          59   \n",
       "31     PC21                     0                     0          19   \n",
       "32     PC32                     0                     0          72   \n",
       "33     SkC6                     0                     0          36   \n",
       "34     StS1                     0                     0          27   \n",
       "35      UC2                     0                     0          19   \n",
       "\n",
       "    TUMOR_HaeIII  MN_Alu1  MN_HaeIII  \n",
       "0             12       27         18  \n",
       "1             18       24         12  \n",
       "2              3       30         20  \n",
       "3             10        7          9  \n",
       "4             11       11          3  \n",
       "5             38       14          6  \n",
       "6             10       37         34  \n",
       "7              3       21          9  \n",
       "8             22       23          8  \n",
       "9             34       15          2  \n",
       "10             8        2          0  \n",
       "11            20       83         48  \n",
       "12            18        6          9  \n",
       "13            13       30         19  \n",
       "14             7       16          9  \n",
       "15             5        3          1  \n",
       "16             9       12          6  \n",
       "17            15       19         16  \n",
       "18            15       19         10  \n",
       "19             8       15         14  \n",
       "20             6        8          9  \n",
       "21            18       24         11  \n",
       "22            11        6          3  \n",
       "23            14        5         15  \n",
       "24             6        3          3  \n",
       "25             9       14          3  \n",
       "26            24       17         13  \n",
       "27            21       12          5  \n",
       "28            11        6          5  \n",
       "29            32        7          9  \n",
       "30            16        3          1  \n",
       "31            10       23         13  \n",
       "32            34        8         10  \n",
       "33            14       28         17  \n",
       "34            20       12         18  \n",
       "35            13        6          7  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIR1 = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/AGCT/'\n",
    "DIR2 = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC/'\n",
    "SAVE_TO = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/'\n",
    "FILTER = False\n",
    "DEL_MATCHES = True\n",
    "LINE = True\n",
    "\n",
    "dirs = os.listdir(DIR1)\n",
    "if LINE:\n",
    "    num_matched = pd.DataFrame(columns=['NAME', 'TUMOR_Alu1_MN_HaeIII', 'TUMOR_HaeIII_MN_Alu1', 'TUMOR_Alu1', 'TUMOR_HaeIII', 'MN_Alu1', 'MN_HaeIII'])\n",
    "else:\n",
    "    num_matched = pd.DataFrame(columns=['NAME', 'TUMOR_Alu1_MN_Rsa1', 'TUMOR_Rsa1_MN_Alu1', 'TUMOR_Alu1', 'TUMOR_Rsa1', 'MN_Alu1', 'MN_Rsa1'])\n",
    "for dirname in dirs:\n",
    "\n",
    "    curr_dir1 = DIR1 + dirname\n",
    "    curr_dir2 = DIR2 + dirname\n",
    "    \n",
    "    if not os.path.isdir(curr_dir1):\n",
    "        continue\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir1, show=False)\n",
    "    # tumor norm matches \n",
    "    num_matched.loc[len(num_matched.index)] = check_matches_tumor_mn(curr_dir1, curr_dir2, file_names, dirname, FILTER, DEL_MATCHES)\n",
    "    \n",
    "if FILTER:\n",
    "    num_matched.to_csv(SAVE_TO + f'matches_tumor_mn_filtr.csv', index=False)\n",
    "else:\n",
    "    num_matched.to_csv(SAVE_TO + f'matches_tumor_mn_nofiltr.csv', index=False)\n",
    "\n",
    "num_matched"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa3f20e6-cd1e-4945-a682-62566106ec39",
   "metadata": {},
   "source": [
    "### Get values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "632b9f28-18c9-4067-925e-53a1e0efb32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coords(curr_dir1, curr_dir2, file_names, save_to, filter, is_LINE):\n",
    "\n",
    "    filtration = ''\n",
    "    if not filter:\n",
    "        filtration = '/nofiltration'\n",
    "\n",
    "    for name in file_names:\n",
    "        df = pd.read_table(curr_dir1 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        coord_df1 = df[['QUERY', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', f'{name}_NUM_READS']].copy(deep=True)\n",
    "        coord_df1.columns = ['READ_NAME', 'INS_CHR', 'INS_COORD', 'INS_STRAND', 'NUM_READS']\n",
    "        coord_df1['RESTRICTASE'] = 'Alu1'\n",
    "        \n",
    "        df = pd.read_table(curr_dir2 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        coord_df2 = df[['QUERY', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', f'{name}_NUM_READS']].copy(deep=True)\n",
    "        coord_df2.columns = ['READ_NAME', 'INS_CHR', 'INS_COORD', 'INS_STRAND', 'NUM_READS']\n",
    "        if is_LINE:\n",
    "            coord_df2['RESTRICTASE'] = 'HaeIII'\n",
    "        else:\n",
    "            coord_df2['RESTRICTASE'] = 'Rsa1'\n",
    "\n",
    "        for i, row1 in coord_df1.iterrows():\n",
    "            for j, row2 in coord_df2.iterrows():\n",
    "                if (row1['INS_CHR'] == row2['INS_CHR']) and (abs(row1['INS_COORD'] - row2['INS_COORD']) <= 10):\n",
    "                    if is_LINE:\n",
    "                        coord_df1.at[i, 'RESTRICTASE'] = 'Alu1, HaeIII'\n",
    "                    else:\n",
    "                        coord_df1.at[i, 'RESTRICTASE'] = 'Alu1, Rsa1'\n",
    "                    coord_df1.at[i, 'NUM_READS'] = row1['NUM_READS'] + row2['NUM_READS']\n",
    "                    coord_df2.drop(j, axis=0, inplace=True)\n",
    "        \n",
    "        df_to_save = pd.concat([coord_df1, coord_df2], ignore_index=True).drop_duplicates()\n",
    "        \n",
    "        df_to_save.to_csv(save_to + f'{name}_coords.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aa059d51-7635-4485-9056-151f8fe18e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_matched_coords(curr_dir1, curr_dir2, file_names, dirname, filter):\n",
    "\n",
    "    num_matched = [dirname]\n",
    "\n",
    "    filtration = ''\n",
    "    if not filter:\n",
    "        filtration = '/nofiltration'\n",
    "\n",
    "    if file_names[0].find('MN-') != -1:\n",
    "        tmp = str(file_names[1])\n",
    "        file_names[1] = str(file_names[0])\n",
    "        file_names[0] = str(tmp)\n",
    "    \n",
    "    for name in file_names:\n",
    "        \n",
    "        df1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        df2 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "\n",
    "        df_to_save = pd.DataFrame(columns=df2.columns)\n",
    "        \n",
    "        for i, row1 in df1.iterrows():\n",
    "            for j, row2 in df2.iterrows():\n",
    "                if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= 10):\n",
    "                    df_to_save.loc[len(df_to_save.index)] = row1\n",
    "\n",
    "        num_matched.append(len(df_to_save['QUERY'].unique()))\n",
    "        num_matched.append(len(df1['QUERY'].unique()))\n",
    "        num_matched.append(len(df2['QUERY'].unique()))\n",
    "    \n",
    "    return num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "521dc870-20c3-46ed-b97c-29065ccdecd3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>MATCHED</th>\n",
       "      <th>Alu1</th>\n",
       "      <th>HaeIII</th>\n",
       "      <th>MN_MATCHED</th>\n",
       "      <th>MN_Alu1</th>\n",
       "      <th>MN_HaeIII</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BC60</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BC66</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC70</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLC1</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CC107</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CC127-2</td>\n",
       "      <td>8</td>\n",
       "      <td>51</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CC18</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>37</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CC46</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC5</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CC82</td>\n",
       "      <td>9</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CC94</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CC96</td>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "      <td>11</td>\n",
       "      <td>83</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CC99</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GC39</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GC41</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LS1</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LS6</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M14</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M21</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M22</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC5</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OC14</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OC33</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OC40</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OC42</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OC48</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OC50</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OC54</td>\n",
       "      <td>7</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OT1</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PC12</td>\n",
       "      <td>8</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PC14</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PC21</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PC32</td>\n",
       "      <td>8</td>\n",
       "      <td>72</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SkC6</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>StS1</td>\n",
       "      <td>5</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>UC2</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NAME  MATCHED  Alu1  HaeIII  MN_MATCHED  MN_Alu1  MN_HaeIII\n",
       "0      BC60        1    19      12           5       27         18\n",
       "1      BC66        1    17      18           1       24         12\n",
       "2      BC70        1     5       3           3       30         20\n",
       "3      BLC1        1    20      10           0        7          9\n",
       "4     CC107        2    19      11           0       11          3\n",
       "5   CC127-2        8    51      38           0       14          6\n",
       "6      CC18        2    13      10           9       37         34\n",
       "7      CC46        0     7       3           2       21          9\n",
       "8       CC5        3    12      22           2       23          8\n",
       "9      CC82        9    47      34           0       15          2\n",
       "10     CC94        1    22       8           0        2          0\n",
       "11     CC96        5    30      20          11       83         48\n",
       "12     CC99        2    21      18           1        6          9\n",
       "13     GC39        1    17      13           3       30         19\n",
       "14     GC41        1    13       7           3       16          9\n",
       "15      LS1        0    15       5           0        3          1\n",
       "16      LS6        1    22       9           0       12          6\n",
       "17      M14        2    10      15           4       19         16\n",
       "18      M21        2    29      15           3       19         10\n",
       "19      M22        0    28       8           3       15         14\n",
       "20      NC5        0    32       6           1        8          9\n",
       "21     OC14        1    19      18           1       24         11\n",
       "22     OC33        2    17      11           0        6          3\n",
       "23     OC40        1    26      14           0        5         15\n",
       "24     OC42        2     9       6           0        3          3\n",
       "25     OC48        3    31       9           0       14          3\n",
       "26     OC50        2    30      24           3       17         13\n",
       "27     OC54        7    24      21           1       12          5\n",
       "28      OT1        0    16      11           0        6          5\n",
       "29     PC12        8    68      32           0        7          9\n",
       "30     PC14        2    59      16           0        3          1\n",
       "31     PC21        2    19      10           1       23         13\n",
       "32     PC32        8    72      34           0        8         10\n",
       "33     SkC6        3    36      14           2       28         17\n",
       "34     StS1        5    27      20           1       12         18\n",
       "35      UC2        0    19      13           0        6          7"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIR1 = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/AGCT/'\n",
    "DIR2 = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC/'\n",
    "SAVE_TO = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/'\n",
    "filter = False\n",
    "LINE = True\n",
    "\n",
    "dirs = os.listdir(DIR1)\n",
    "if LINE:\n",
    "    num_matched = pd.DataFrame(columns=['NAME', 'MATCHED', 'Alu1', 'HaeIII', 'MN_MATCHED', 'MN_Alu1', 'MN_HaeIII'])\n",
    "else:\n",
    "    num_matched = pd.DataFrame(columns=['NAME', 'MATCHED', 'Alu1', 'Rsa1', 'MN_MATCHED', 'MN_Alu1', 'MN_Rsa1'])\n",
    "general_ins_num = pd.DataFrame(columns=['NAME', 'TOTAL_INS', 'MN_TOTAL'])\n",
    "for dirname in dirs:\n",
    "\n",
    "    curr_dir1 = DIR1 + dirname\n",
    "    curr_dir2 = DIR2 + dirname\n",
    "    \n",
    "    if not os.path.isdir(curr_dir1):\n",
    "        continue\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir1, show=False)\n",
    "    # get coords for Maria\n",
    "    get_coords(curr_dir1, curr_dir2, file_names, SAVE_TO, filter, LINE)\n",
    "    # check\n",
    "    matches = check_matched_coords(curr_dir1, curr_dir2, file_names, dirname, filter)\n",
    "    num_matched.loc[len(num_matched.index)] = matches\n",
    "    general_ins_num.loc[len(general_ins_num.index)] = [matches[0], matches[2] + matches[3] - matches[1],\n",
    "                                                       matches[5] + matches[6] - matches[4],\n",
    "                                                      ]\n",
    "\n",
    "if filter:\n",
    "    num_matched.to_csv(SAVE_TO + f'matched_ins_filtr.csv', index=False)\n",
    "    general_ins_num.to_csv(SAVE_TO + f'total_ins_filtr.csv', index=False)\n",
    "else:\n",
    "    num_matched.to_csv(SAVE_TO + f'matched_ins_nofiltr.csv', index=False)\n",
    "    general_ins_num.to_csv(SAVE_TO + f'total_ins_nofiltr.csv', index=False)\n",
    "\n",
    "num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "53ca3f94-5676-4bf8-bb72-aea3a4235527",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>TOTAL_INS</th>\n",
       "      <th>MN_TOTAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BC60</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BC66</td>\n",
       "      <td>34</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC70</td>\n",
       "      <td>7</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLC1</td>\n",
       "      <td>29</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CC107</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CC127-2</td>\n",
       "      <td>81</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CC18</td>\n",
       "      <td>21</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CC46</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC5</td>\n",
       "      <td>31</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CC82</td>\n",
       "      <td>72</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CC94</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CC96</td>\n",
       "      <td>45</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CC99</td>\n",
       "      <td>37</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GC39</td>\n",
       "      <td>29</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GC41</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LS1</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LS6</td>\n",
       "      <td>30</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M14</td>\n",
       "      <td>23</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M21</td>\n",
       "      <td>42</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M22</td>\n",
       "      <td>36</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC5</td>\n",
       "      <td>38</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OC14</td>\n",
       "      <td>36</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OC33</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OC40</td>\n",
       "      <td>39</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OC42</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OC48</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OC50</td>\n",
       "      <td>52</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OC54</td>\n",
       "      <td>38</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OT1</td>\n",
       "      <td>27</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PC12</td>\n",
       "      <td>92</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PC14</td>\n",
       "      <td>73</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PC21</td>\n",
       "      <td>27</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PC32</td>\n",
       "      <td>98</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SkC6</td>\n",
       "      <td>47</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>StS1</td>\n",
       "      <td>42</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>UC2</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NAME  TOTAL_INS  MN_TOTAL\n",
       "0      BC60         30        40\n",
       "1      BC66         34        35\n",
       "2      BC70          7        47\n",
       "3      BLC1         29        16\n",
       "4     CC107         28        14\n",
       "5   CC127-2         81        20\n",
       "6      CC18         21        62\n",
       "7      CC46         10        28\n",
       "8       CC5         31        29\n",
       "9      CC82         72        17\n",
       "10     CC94         29         2\n",
       "11     CC96         45       120\n",
       "12     CC99         37        14\n",
       "13     GC39         29        46\n",
       "14     GC41         19        22\n",
       "15      LS1         20         4\n",
       "16      LS6         30        18\n",
       "17      M14         23        31\n",
       "18      M21         42        26\n",
       "19      M22         36        26\n",
       "20      NC5         38        16\n",
       "21     OC14         36        34\n",
       "22     OC33         26         9\n",
       "23     OC40         39        20\n",
       "24     OC42         13         6\n",
       "25     OC48         37        17\n",
       "26     OC50         52        27\n",
       "27     OC54         38        16\n",
       "28      OT1         27        11\n",
       "29     PC12         92        16\n",
       "30     PC14         73         4\n",
       "31     PC21         27        35\n",
       "32     PC32         98        18\n",
       "33     SkC6         47        43\n",
       "34     StS1         42        29\n",
       "35      UC2         32        13"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "general_ins_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c0b87d-8dd1-477b-a8a3-917ead651d0b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
