{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63c8357d-aca0-43ac-ad76-d8954b6d133d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/Komar/Lab/retroparty\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b91bd6f-b591-40e3-ad7e-c4d43ca38fa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['BC60-L', 'BC66-L', 'BC70-L', 'BCL1-L', 'CC107-L', 'CC127-2-L', 'CC18-L', 'CC46-L', 'CC5-L', 'CC82-L', 'CC94-L', 'CC96-L', 'CC99-L', 'GC39-L', 'GC41-L', 'LS1-L', 'LS6-L', 'M14-L', 'M21-L', 'M22-L', 'NC5-L', 'OC14-L', 'OC33-L', 'OC40-L', 'OC42-L', 'OC48-L', 'OC50-L', 'OC54-L', 'OT1-L', 'PC12-L', 'PC14-L', 'PC21-L', 'PC32-L', 'SkC6-L', 'StS1-L', 'UC2-L']\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC/')\n",
    "print(os.listdir())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48d2948a-2b26-472a-bcab-aa6d7dbc20ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC60-L ---------------------------------------------------\n",
      "metatable shape: (279, 32)\n",
      "sample file names: ['BC60-L_S106_ME_L001__', 'MN-BC60-L_S108_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 80   including test: 0\n",
      "somatic and quasi-somatic: (199, 32)   including test: 0 \n",
      "\n",
      "BC60-L_S106_ME_L001__ 97   test: 0\n",
      "MN-BC60-L_S108_ME_L001__ 102   test: 0\n",
      "overall num of somatic insertions:  199 \n",
      "\n",
      "BC60-L_S106_ME_L001__  was: 97  become: 16\n",
      "BC60-L_S106_ME_L001__ reads were written\n",
      "MN-BC60-L_S108_ME_L001__  was: 102  become: 23\n",
      "MN-BC60-L_S108_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC60-L_S106_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 1/1 [00:04<00:00,  4.77s/it]\n",
      "BC60-L_S106_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 1/1 [00:04<00:00,  4.98s/it]\n",
      "MN-BC60-L_S108_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.93s/it]\n",
      "MN-BC60-L_S108_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.33s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.86s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 9\n",
      "unique names: 9\n",
      "after: 9 \n",
      "\n",
      "before: 13\n",
      "unique names: 13\n",
      "after: 13 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC60-L_S106_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 9/9 [02:02<00:00, 13.58s/it]\n",
      "MN-BC60-L_S108_ME_L001__: 100%|█████████████████████████████████████████████████████████| 13/13 [03:03<00:00, 14.10s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BC66-L ---------------------------------------------------\n",
      "metatable shape: (275, 32)\n",
      "sample file names: ['BC66-L_S102_ME_L001__', 'MN-BC66-L_S104_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 75   including test: 0\n",
      "somatic and quasi-somatic: (200, 32)   including test: 0 \n",
      "\n",
      "BC66-L_S102_ME_L001__ 83   test: 0\n",
      "MN-BC66-L_S104_ME_L001__ 117   test: 0\n",
      "overall num of somatic insertions:  200 \n",
      "\n",
      "BC66-L_S102_ME_L001__  was: 83  become: 15\n",
      "BC66-L_S102_ME_L001__ reads were written\n",
      "MN-BC66-L_S104_ME_L001__  was: 117  become: 23\n",
      "MN-BC66-L_S104_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC66-L_S102_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 1/1 [00:06<00:00,  6.53s/it]\n",
      "BC66-L_S102_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 1/1 [00:05<00:00,  5.22s/it]\n",
      "MN-BC66-L_S104_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.50s/it]\n",
      "MN-BC66-L_S104_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.86s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.62s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 48\n",
      "unique names: 11\n",
      "after: 16 \n",
      "\n",
      "before: 14\n",
      "unique names: 12\n",
      "after: 14 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC66-L_S102_ME_L001__: 100%|████████████████████████████████████████████████████████████| 16/16 [02:04<00:00,  7.76s/it]\n",
      "MN-BC66-L_S104_ME_L001__: 100%|█████████████████████████████████████████████████████████| 14/14 [02:38<00:00, 11.30s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BC70-L ---------------------------------------------------\n",
      "metatable shape: (270, 32)\n",
      "sample file names: ['BC70-L_S142_ME_L001__', 'MN-BC70-L_S144_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 9   including test: 0\n",
      "somatic and quasi-somatic: (261, 32)   including test: 0 \n",
      "\n",
      "BC70-L_S142_ME_L001__ 26   test: 0\n",
      "MN-BC70-L_S144_ME_L001__ 235   test: 0\n",
      "overall num of somatic insertions:  261 \n",
      "\n",
      "BC70-L_S142_ME_L001__  was: 26  become: 2\n",
      "BC70-L_S142_ME_L001__ reads were written\n",
      "MN-BC70-L_S144_ME_L001__  was: 235  become: 46\n",
      "MN-BC70-L_S144_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC70-L_S142_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "BC70-L_S142_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "MN-BC70-L_S144_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.11s/it]\n",
      "MN-BC70-L_S144_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 4/4 [00:16<00:00,  4.22s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.70s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 2\n",
      "unique names: 2\n",
      "after: 2 \n",
      "\n",
      "before: 54\n",
      "unique names: 36\n",
      "after: 42 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BC70-L_S142_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 2/2 [00:42<00:00, 21.29s/it]\n",
      "MN-BC70-L_S144_ME_L001__: 100%|█████████████████████████████████████████████████████████| 42/42 [09:47<00:00, 13.98s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "BCL1-L ---------------------------------------------------\n",
      "metatable shape: (1059, 32)\n",
      "sample file names: ['BLC1-L_S74_ME_L001__', 'MN-BLC1-L_S76_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 402   including test: 0\n",
      "somatic and quasi-somatic: (657, 32)   including test: 0 \n",
      "\n",
      "BLC1-L_S74_ME_L001__ 441   test: 0\n",
      "MN-BLC1-L_S76_ME_L001__ 216   test: 0\n",
      "overall num of somatic insertions:  657 \n",
      "\n",
      "BLC1-L_S74_ME_L001__  was: 441  become: 93\n",
      "BLC1-L_S74_ME_L001__ reads were written\n",
      "MN-BLC1-L_S76_ME_L001__  was: 216  become: 31\n",
      "MN-BLC1-L_S76_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLC1-L_S74_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 9/9 [00:31<00:00,  3.49s/it]\n",
      "BLC1-L_S74_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 9/9 [00:40<00:00,  4.45s/it]\n",
      "MN-BLC1-L_S76_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:10<00:00,  3.44s/it]\n",
      "MN-BLC1-L_S76_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:13<00:00,  4.43s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:16<00:00,  8.07s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 41\n",
      "unique names: 35\n",
      "after: 41 \n",
      "\n",
      "before: 85\n",
      "unique names: 16\n",
      "after: 25 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLC1-L_S74_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 41/41 [08:45<00:00, 12.81s/it]\n",
      "MN-BLC1-L_S76_ME_L001__: 100%|██████████████████████████████████████████████████████████| 25/25 [05:39<00:00, 13.57s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC107-L ---------------------------------------------------\n",
      "metatable shape: (764, 32)\n",
      "sample file names: ['CC107-L_S130_ME_L001__', 'MN-CC107-L_S132_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 277   including test: 0\n",
      "somatic and quasi-somatic: (487, 32)   including test: 0 \n",
      "\n",
      "CC107-L_S130_ME_L001__ 298   test: 0\n",
      "MN-CC107-L_S132_ME_L001__ 189   test: 0\n",
      "overall num of somatic insertions:  487 \n",
      "\n",
      "CC107-L_S130_ME_L001__  was: 298  become: 60\n",
      "CC107-L_S130_ME_L001__ reads were written\n",
      "MN-CC107-L_S132_ME_L001__  was: 189  become: 22\n",
      "MN-CC107-L_S132_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC107-L_S130_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 6/6 [00:21<00:00,  3.52s/it]\n",
      "CC107-L_S130_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 6/6 [00:26<00:00,  4.41s/it]\n",
      "MN-CC107-L_S132_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.33s/it]\n",
      "MN-CC107-L_S132_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 2/2 [00:11<00:00,  5.52s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:26<00:00, 13.34s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 395\n",
      "unique names: 33\n",
      "after: 64 \n",
      "\n",
      "before: 39\n",
      "unique names: 9\n",
      "after: 17 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC107-L_S130_ME_L001__: 100%|███████████████████████████████████████████████████████████| 64/64 [11:39<00:00, 10.94s/it]\n",
      "MN-CC107-L_S132_ME_L001__: 100%|████████████████████████████████████████████████████████| 17/17 [02:27<00:00,  8.70s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC127-2-L ---------------------------------------------------\n",
      "metatable shape: (1161, 32)\n",
      "sample file names: ['CC127-2-L_S6_ME_L001__', 'MN-CC127-L_S8_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 346   including test: 0\n",
      "somatic and quasi-somatic: (815, 32)   including test: 0 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__ 578   test: 0\n",
      "MN-CC127-L_S8_ME_L001__ 237   test: 0\n",
      "overall num of somatic insertions:  815 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__  was: 578  become: 155\n",
      "CC127-2-L_S6_ME_L001__ reads were written\n",
      "MN-CC127-L_S8_ME_L001__  was: 237  become: 42\n",
      "MN-CC127-L_S8_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC127-2-L_S6_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 15/15 [00:59<00:00,  3.97s/it]\n",
      "CC127-2-L_S6_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 15/15 [01:13<00:00,  4.87s/it]\n",
      "MN-CC127-L_S8_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 4/4 [00:15<00:00,  3.77s/it]\n",
      "MN-CC127-L_S8_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 4/4 [00:19<00:00,  4.85s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:55<00:00, 27.51s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 293\n",
      "unique names: 87\n",
      "after: 142 \n",
      "\n",
      "before: 167\n",
      "unique names: 22\n",
      "after: 41 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC127-2-L_S6_ME_L001__: 100%|█████████████████████████████████████████████████████████| 142/142 [22:28<00:00,  9.49s/it]\n",
      "MN-CC127-L_S8_ME_L001__: 100%|██████████████████████████████████████████████████████████| 41/41 [05:33<00:00,  8.13s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC18-L ---------------------------------------------------\n",
      "metatable shape: (1035, 32)\n",
      "sample file names: ['CC18-L_S118_ME_L001__', 'MN-CC18-L_S120_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 334   including test: 0\n",
      "somatic and quasi-somatic: (701, 32)   including test: 0 \n",
      "\n",
      "CC18-L_S118_ME_L001__ 260   test: 0\n",
      "MN-CC18-L_S120_ME_L001__ 441   test: 0\n",
      "overall num of somatic insertions:  701 \n",
      "\n",
      "CC18-L_S118_ME_L001__  was: 260  become: 47\n",
      "CC18-L_S118_ME_L001__ reads were written\n",
      "MN-CC18-L_S120_ME_L001__  was: 441  become: 120\n",
      "MN-CC18-L_S120_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC18-L_S118_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 4/4 [00:13<00:00,  3.33s/it]\n",
      "CC18-L_S118_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 4/4 [00:16<00:00,  4.21s/it]\n",
      "MN-CC18-L_S120_ME_L001___R1: 100%|██████████████████████████████████████████████████████| 12/12 [00:43<00:00,  3.62s/it]\n",
      "MN-CC18-L_S120_ME_L001___R2: 100%|██████████████████████████████████████████████████████| 12/12 [00:56<00:00,  4.71s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:32<00:00, 16.04s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 50\n",
      "unique names: 24\n",
      "after: 29 \n",
      "\n",
      "before: 491\n",
      "unique names: 66\n",
      "after: 102 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC18-L_S118_ME_L001__: 100%|████████████████████████████████████████████████████████████| 29/29 [05:08<00:00, 10.65s/it]\n",
      "MN-CC18-L_S120_ME_L001__: 100%|███████████████████████████████████████████████████████| 102/102 [15:59<00:00,  9.41s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC46-L ---------------------------------------------------\n",
      "metatable shape: (805, 32)\n",
      "sample file names: ['CC46-L_S134_ME_L001__', 'MN-CC46-L_S136_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 252   including test: 0\n",
      "somatic and quasi-somatic: (553, 32)   including test: 0 \n",
      "\n",
      "CC46-L_S134_ME_L001__ 213   test: 0\n",
      "MN-CC46-L_S136_ME_L001__ 340   test: 0\n",
      "overall num of somatic insertions:  553 \n",
      "\n",
      "CC46-L_S134_ME_L001__  was: 213  become: 36\n",
      "CC46-L_S134_ME_L001__ reads were written\n",
      "MN-CC46-L_S136_ME_L001__  was: 340  become: 70\n",
      "MN-CC46-L_S136_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC46-L_S134_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:09<00:00,  3.07s/it]\n",
      "CC46-L_S134_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:11<00:00,  3.90s/it]\n",
      "MN-CC46-L_S136_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 7/7 [00:26<00:00,  3.84s/it]\n",
      "MN-CC46-L_S136_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 7/7 [00:32<00:00,  4.58s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:26<00:00, 13.27s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 66\n",
      "unique names: 15\n",
      "after: 27 \n",
      "\n",
      "before: 275\n",
      "unique names: 49\n",
      "after: 132 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC46-L_S134_ME_L001__: 100%|████████████████████████████████████████████████████████████| 27/27 [05:17<00:00, 11.76s/it]\n",
      "MN-CC46-L_S136_ME_L001__: 100%|███████████████████████████████████████████████████████| 132/132 [17:05<00:00,  7.77s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC5-L ---------------------------------------------------\n",
      "metatable shape: (361, 32)\n",
      "sample file names: ['CC5-L_S146_ME_L001__', 'MN-CC5-L_S148_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 6   including test: 0\n",
      "somatic and quasi-somatic: (355, 32)   including test: 0 \n",
      "\n",
      "CC5-L_S146_ME_L001__ 103   test: 0\n",
      "MN-CC5-L_S148_ME_L001__ 252   test: 0\n",
      "overall num of somatic insertions:  355 \n",
      "\n",
      "CC5-L_S146_ME_L001__  was: 103  become: 21\n",
      "CC5-L_S146_ME_L001__ reads were written\n",
      "MN-CC5-L_S148_ME_L001__  was: 252  become: 47\n",
      "MN-CC5-L_S148_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC5-L_S146_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.72s/it]\n",
      "CC5-L_S146_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.41s/it]\n",
      "MN-CC5-L_S148_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.04s/it]\n",
      "MN-CC5-L_S148_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 4/4 [00:17<00:00,  4.41s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.92s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 15\n",
      "unique names: 15\n",
      "after: 15 \n",
      "\n",
      "before: 176\n",
      "unique names: 34\n",
      "after: 47 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC5-L_S146_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 15/15 [02:59<00:00, 11.99s/it]\n",
      "MN-CC5-L_S148_ME_L001__: 100%|██████████████████████████████████████████████████████████| 47/47 [07:40<00:00,  9.80s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC82-L ---------------------------------------------------\n",
      "metatable shape: (651, 32)\n",
      "sample file names: ['CC82-L_S10_ME_L001__', 'MN-CC82-L_S12_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 103   including test: 0\n",
      "somatic and quasi-somatic: (548, 32)   including test: 0 \n",
      "\n",
      "CC82-L_S10_ME_L001__ 410   test: 0\n",
      "MN-CC82-L_S12_ME_L001__ 138   test: 0\n",
      "overall num of somatic insertions:  548 \n",
      "\n",
      "CC82-L_S10_ME_L001__  was: 410  become: 62\n",
      "CC82-L_S10_ME_L001__ reads were written\n",
      "MN-CC82-L_S12_ME_L001__  was: 138  become: 15\n",
      "MN-CC82-L_S12_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC82-L_S10_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.32s/it]\n",
      "CC82-L_S10_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:16<00:00,  2.76s/it]\n",
      "MN-CC82-L_S12_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.75s/it]\n",
      "MN-CC82-L_S12_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.71s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.77s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 672\n",
      "unique names: 46\n",
      "after: 85 \n",
      "\n",
      "before: 9\n",
      "unique names: 9\n",
      "after: 9 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC82-L_S10_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 85/85 [17:05<00:00, 12.07s/it]\n",
      "MN-CC82-L_S12_ME_L001__: 100%|████████████████████████████████████████████████████████████| 9/9 [01:51<00:00, 12.44s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC94-L ---------------------------------------------------\n",
      "metatable shape: (124, 32)\n",
      "sample file names: ['CC94-L_S14_ME_L001__', 'MN-CC94-L_S16_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 10   including test: 0\n",
      "somatic and quasi-somatic: (114, 32)   including test: 0 \n",
      "\n",
      "CC94-L_S14_ME_L001__ 88   test: 0\n",
      "MN-CC94-L_S16_ME_L001__ 26   test: 0\n",
      "overall num of somatic insertions:  114 \n",
      "\n",
      "CC94-L_S14_ME_L001__  was: 88  become: 8\n",
      "CC94-L_S14_ME_L001__ reads were written\n",
      "MN-CC94-L_S16_ME_L001__  was: 26  become: 5\n",
      "MN-CC94-L_S16_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC94-L_S14_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "CC94-L_S14_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "MN-CC94-L_S16_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "MN-CC94-L_S16_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  5.19it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 1\n",
      "unique names: 1\n",
      "after: 1 \n",
      "\n",
      "before: 11\n",
      "unique names: 4\n",
      "after: 9 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC94-L_S14_ME_L001__: 100%|███████████████████████████████████████████████████████████████| 1/1 [00:21<00:00, 21.16s/it]\n",
      "MN-CC94-L_S16_ME_L001__: 100%|████████████████████████████████████████████████████████████| 9/9 [02:05<00:00, 13.98s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC96-L ---------------------------------------------------\n",
      "metatable shape: (1660, 32)\n",
      "sample file names: ['CC96-L_S66_ME_L001__', 'MN-CC96-L_S68_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 606   including test: 0\n",
      "somatic and quasi-somatic: (1054, 32)   including test: 0 \n",
      "\n",
      "CC96-L_S66_ME_L001__ 383   test: 0\n",
      "MN-CC96-L_S68_ME_L001__ 671   test: 0\n",
      "overall num of somatic insertions:  1054 \n",
      "\n",
      "CC96-L_S66_ME_L001__  was: 383  become: 96\n",
      "CC96-L_S66_ME_L001__ reads were written\n",
      "MN-CC96-L_S68_ME_L001__  was: 671  become: 161\n",
      "MN-CC96-L_S68_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC96-L_S66_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 9/9 [00:18<00:00,  2.11s/it]\n",
      "CC96-L_S66_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 9/9 [00:24<00:00,  2.75s/it]\n",
      "MN-CC96-L_S68_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 16/16 [00:36<00:00,  2.27s/it]\n",
      "MN-CC96-L_S68_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 16/16 [00:45<00:00,  2.82s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:15<00:00,  7.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 192\n",
      "unique names: 51\n",
      "after: 68 \n",
      "\n",
      "before: 213\n",
      "unique names: 91\n",
      "after: 131 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC96-L_S66_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 68/68 [14:21<00:00, 12.67s/it]\n",
      "MN-CC96-L_S68_ME_L001__: 100%|████████████████████████████████████████████████████████| 131/131 [22:56<00:00, 10.51s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CC99-L ---------------------------------------------------\n",
      "metatable shape: (865, 32)\n",
      "sample file names: ['CC99-L_S70_ME_L001__', 'MN-CC99-L_S72_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 265   including test: 0\n",
      "somatic and quasi-somatic: (600, 32)   including test: 0 \n",
      "\n",
      "CC99-L_S70_ME_L001__ 380   test: 0\n",
      "MN-CC99-L_S72_ME_L001__ 220   test: 0\n",
      "overall num of somatic insertions:  600 \n",
      "\n",
      "CC99-L_S70_ME_L001__  was: 380  become: 65\n",
      "CC99-L_S70_ME_L001__ reads were written\n",
      "MN-CC99-L_S72_ME_L001__  was: 220  become: 40\n",
      "MN-CC99-L_S72_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC99-L_S70_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:12<00:00,  2.15s/it]\n",
      "CC99-L_S70_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.63s/it]\n",
      "MN-CC99-L_S72_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 4/4 [00:06<00:00,  1.70s/it]\n",
      "MN-CC99-L_S72_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 4/4 [00:09<00:00,  2.42s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.83s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 127\n",
      "unique names: 39\n",
      "after: 47 \n",
      "\n",
      "before: 17\n",
      "unique names: 17\n",
      "after: 17 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CC99-L_S70_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 47/47 [08:04<00:00, 10.30s/it]\n",
      "MN-CC99-L_S72_ME_L001__: 100%|██████████████████████████████████████████████████████████| 17/17 [04:24<00:00, 15.56s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GC39-L ---------------------------------------------------\n",
      "metatable shape: (1172, 32)\n",
      "sample file names: ['GC39-L_S62_ME_L001__', 'MN-GC39-L_S64_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 493   including test: 0\n",
      "somatic and quasi-somatic: (679, 32)   including test: 0 \n",
      "\n",
      "GC39-L_S62_ME_L001__ 301   test: 0\n",
      "MN-GC39-L_S64_ME_L001__ 378   test: 0\n",
      "overall num of somatic insertions:  679 \n",
      "\n",
      "GC39-L_S62_ME_L001__  was: 301  become: 68\n",
      "GC39-L_S62_ME_L001__ reads were written\n",
      "MN-GC39-L_S64_ME_L001__  was: 378  become: 65\n",
      "MN-GC39-L_S64_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC39-L_S62_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.23s/it]\n",
      "GC39-L_S62_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 6/6 [00:17<00:00,  2.88s/it]\n",
      "MN-GC39-L_S64_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.31s/it]\n",
      "MN-GC39-L_S64_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 6/6 [00:17<00:00,  2.96s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.10s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 91\n",
      "unique names: 29\n",
      "after: 39 \n",
      "\n",
      "before: 123\n",
      "unique names: 33\n",
      "after: 46 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC39-L_S62_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 39/39 [05:44<00:00,  8.83s/it]\n",
      "MN-GC39-L_S64_ME_L001__: 100%|██████████████████████████████████████████████████████████| 46/46 [07:10<00:00,  9.36s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GC41-L ---------------------------------------------------\n",
      "metatable shape: (971, 32)\n",
      "sample file names: ['GC41-L_S154_ME_L001__', 'MN-GC41-L_S156_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 398   including test: 0\n",
      "somatic and quasi-somatic: (573, 32)   including test: 0 \n",
      "\n",
      "GC41-L_S154_ME_L001__ 300   test: 0\n",
      "MN-GC41-L_S156_ME_L001__ 273   test: 0\n",
      "overall num of somatic insertions:  573 \n",
      "\n",
      "GC41-L_S154_ME_L001__  was: 300  become: 49\n",
      "GC41-L_S154_ME_L001__ reads were written\n",
      "MN-GC41-L_S156_ME_L001__  was: 273  become: 45\n",
      "MN-GC41-L_S156_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC41-L_S154_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.00s/it]\n",
      "GC41-L_S154_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 4/4 [00:10<00:00,  2.52s/it]\n",
      "MN-GC41-L_S156_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.14s/it]\n",
      "MN-GC41-L_S156_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 4/4 [00:10<00:00,  2.67s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.84s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 43\n",
      "unique names: 21\n",
      "after: 27 \n",
      "\n",
      "before: 42\n",
      "unique names: 24\n",
      "after: 36 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GC41-L_S154_ME_L001__: 100%|████████████████████████████████████████████████████████████| 27/27 [06:07<00:00, 13.61s/it]\n",
      "MN-GC41-L_S156_ME_L001__: 100%|█████████████████████████████████████████████████████████| 36/36 [08:06<00:00, 13.52s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LS1-L ---------------------------------------------------\n",
      "metatable shape: (186, 32)\n",
      "sample file names: ['LS1-L_S138_ME_L001__', 'MN-LS1-L_S140_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 5   including test: 0\n",
      "somatic and quasi-somatic: (181, 32)   including test: 0 \n",
      "\n",
      "LS1-L_S138_ME_L001__ 128   test: 0\n",
      "MN-LS1-L_S140_ME_L001__ 53   test: 0\n",
      "overall num of somatic insertions:  181 \n",
      "\n",
      "LS1-L_S138_ME_L001__  was: 128  become: 14\n",
      "LS1-L_S138_ME_L001__ reads were written\n",
      "MN-LS1-L_S140_ME_L001__  was: 53  become: 7\n",
      "MN-LS1-L_S140_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS1-L_S138_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.94s/it]\n",
      "LS1-L_S138_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.45s/it]\n",
      "MN-LS1-L_S140_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "MN-LS1-L_S140_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  3.64it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 7\n",
      "unique names: 7\n",
      "after: 7 \n",
      "\n",
      "before: 4\n",
      "unique names: 4\n",
      "after: 4 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS1-L_S138_ME_L001__: 100%|███████████████████████████████████████████████████████████████| 7/7 [01:25<00:00, 12.21s/it]\n",
      "MN-LS1-L_S140_ME_L001__: 100%|████████████████████████████████████████████████████████████| 4/4 [00:38<00:00,  9.60s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LS6-L ---------------------------------------------------\n",
      "metatable shape: (217, 32)\n",
      "sample file names: ['LS6-L_S46_ME_L001__', 'MN-LS6-L_S48_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 49   including test: 0\n",
      "somatic and quasi-somatic: (168, 32)   including test: 0 \n",
      "\n",
      "LS6-L_S46_ME_L001__ 100   test: 0\n",
      "MN-LS6-L_S48_ME_L001__ 68   test: 0\n",
      "overall num of somatic insertions:  168 \n",
      "\n",
      "LS6-L_S46_ME_L001__  was: 100  become: 11\n",
      "LS6-L_S46_ME_L001__ reads were written\n",
      "MN-LS6-L_S48_ME_L001__  was: 68  become: 8\n",
      "MN-LS6-L_S48_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS6-L_S46_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.13s/it]\n",
      "LS6-L_S46_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.08s/it]\n",
      "MN-LS6-L_S48_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "MN-LS6-L_S48_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  2.20it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 31\n",
      "unique names: 7\n",
      "after: 12 \n",
      "\n",
      "before: 6\n",
      "unique names: 6\n",
      "after: 6 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LS6-L_S46_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 12/12 [01:54<00:00,  9.56s/it]\n",
      "MN-LS6-L_S48_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 6/6 [01:04<00:00, 10.68s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M14-L ---------------------------------------------------\n",
      "metatable shape: (248, 32)\n",
      "sample file names: ['M14-L_S94_ME_L001__', 'MN-M14-L_S96_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 76   including test: 0\n",
      "somatic and quasi-somatic: (172, 32)   including test: 0 \n",
      "\n",
      "M14-L_S94_ME_L001__ 78   test: 0\n",
      "MN-M14-L_S96_ME_L001__ 94   test: 0\n",
      "overall num of somatic insertions:  172 \n",
      "\n",
      "M14-L_S94_ME_L001__  was: 78  become: 14\n",
      "M14-L_S94_ME_L001__ reads were written\n",
      "MN-M14-L_S96_ME_L001__  was: 94  become: 15\n",
      "MN-M14-L_S96_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M14-L_S94_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.14s/it]\n",
      "M14-L_S94_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.81s/it]\n",
      "MN-M14-L_S96_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.38s/it]\n",
      "MN-M14-L_S96_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.62s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.13it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 9\n",
      "unique names: 7\n",
      "after: 9 \n",
      "\n",
      "before: 31\n",
      "unique names: 11\n",
      "after: 15 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M14-L_S94_ME_L001__: 100%|████████████████████████████████████████████████████████████████| 9/9 [01:51<00:00, 12.44s/it]\n",
      "MN-M14-L_S96_ME_L001__: 100%|███████████████████████████████████████████████████████████| 15/15 [02:26<00:00,  9.77s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M21-L ---------------------------------------------------\n",
      "metatable shape: (1203, 32)\n",
      "sample file names: ['M21-L_S82_ME_L001__', 'MN-M21-L_S84_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 526   including test: 0\n",
      "somatic and quasi-somatic: (677, 32)   including test: 0 \n",
      "\n",
      "M21-L_S82_ME_L001__ 298   test: 0\n",
      "MN-M21-L_S84_ME_L001__ 379   test: 0\n",
      "overall num of somatic insertions:  677 \n",
      "\n",
      "M21-L_S82_ME_L001__  was: 298  become: 75\n",
      "M21-L_S82_ME_L001__ reads were written\n",
      "MN-M21-L_S84_ME_L001__  was: 379  become: 82\n",
      "MN-M21-L_S84_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M21-L_S82_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 7/7 [00:14<00:00,  2.09s/it]\n",
      "M21-L_S82_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 7/7 [00:19<00:00,  2.86s/it]\n",
      "MN-M21-L_S84_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 8/8 [00:16<00:00,  2.00s/it]\n",
      "MN-M21-L_S84_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 8/8 [00:20<00:00,  2.59s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:12<00:00,  6.08s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 404\n",
      "unique names: 43\n",
      "after: 79 \n",
      "\n",
      "before: 181\n",
      "unique names: 34\n",
      "after: 59 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M21-L_S82_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 79/79 [16:09<00:00, 12.27s/it]\n",
      "MN-M21-L_S84_ME_L001__: 100%|███████████████████████████████████████████████████████████| 59/59 [11:05<00:00, 11.28s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "M22-L ---------------------------------------------------\n",
      "metatable shape: (1248, 32)\n",
      "sample file names: ['M22-L_S98_ME_L001__', 'MN-M22-L_S100_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 546   including test: 0\n",
      "somatic and quasi-somatic: (702, 32)   including test: 0 \n",
      "\n",
      "M22-L_S98_ME_L001__ 319   test: 0\n",
      "MN-M22-L_S100_ME_L001__ 383   test: 0\n",
      "overall num of somatic insertions:  702 \n",
      "\n",
      "M22-L_S98_ME_L001__  was: 319  become: 64\n",
      "M22-L_S98_ME_L001__ reads were written\n",
      "MN-M22-L_S100_ME_L001__  was: 383  become: 93\n",
      "MN-M22-L_S100_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M22-L_S98_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.32s/it]\n",
      "M22-L_S98_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.64s/it]\n",
      "MN-M22-L_S100_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 9/9 [00:18<00:00,  2.01s/it]\n",
      "MN-M22-L_S100_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 9/9 [00:23<00:00,  2.56s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.89s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 395\n",
      "unique names: 24\n",
      "after: 37 \n",
      "\n",
      "before: 107\n",
      "unique names: 36\n",
      "after: 52 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "M22-L_S98_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 37/37 [05:47<00:00,  9.39s/it]\n",
      "MN-M22-L_S100_ME_L001__: 100%|██████████████████████████████████████████████████████████| 52/52 [09:28<00:00, 10.93s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NC5-L ---------------------------------------------------\n",
      "metatable shape: (776, 32)\n",
      "sample file names: ['MN-NC5-L_S88_ME_L001__', 'NC5-L_S86_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 299   including test: 0\n",
      "somatic and quasi-somatic: (477, 32)   including test: 0 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__ 163   test: 0\n",
      "NC5-L_S86_ME_L001__ 314   test: 0\n",
      "overall num of somatic insertions:  477 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__  was: 163  become: 40\n",
      "MN-NC5-L_S88_ME_L001__ reads were written\n",
      "NC5-L_S86_ME_L001__  was: 314  become: 60\n",
      "NC5-L_S86_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-NC5-L_S88_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.15s/it]\n",
      "MN-NC5-L_S88_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 4/4 [00:11<00:00,  2.83s/it]\n",
      "NC5-L_S86_ME_L001___R1: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:12<00:00,  2.00s/it]\n",
      "NC5-L_S86_ME_L001___R2: 100%|█████████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.56s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:08<00:00,  4.26s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 306\n",
      "unique names: 23\n",
      "after: 45 \n",
      "\n",
      "before: 500\n",
      "unique names: 22\n",
      "after: 48 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-NC5-L_S88_ME_L001__: 100%|███████████████████████████████████████████████████████████| 45/45 [05:45<00:00,  7.67s/it]\n",
      "NC5-L_S86_ME_L001__: 100%|██████████████████████████████████████████████████████████████| 48/48 [07:33<00:00,  9.45s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC14-L ---------------------------------------------------\n",
      "metatable shape: (1057, 32)\n",
      "sample file names: ['MN-OC14-L_S112_ME_L001__', 'OC14-L_S110_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 421   including test: 0\n",
      "somatic and quasi-somatic: (636, 32)   including test: 0 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__ 331   test: 0\n",
      "OC14-L_S110_ME_L001__ 305   test: 0\n",
      "overall num of somatic insertions:  636 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__  was: 331  become: 77\n",
      "MN-OC14-L_S112_ME_L001__ reads were written\n",
      "OC14-L_S110_ME_L001__  was: 305  become: 83\n",
      "OC14-L_S110_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC14-L_S112_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 7/7 [00:15<00:00,  2.15s/it]\n",
      "MN-OC14-L_S112_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 7/7 [00:19<00:00,  2.82s/it]\n",
      "OC14-L_S110_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 8/8 [00:17<00:00,  2.20s/it]\n",
      "OC14-L_S110_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 8/8 [00:21<00:00,  2.74s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.62s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 174\n",
      "unique names: 45\n",
      "after: 80 \n",
      "\n",
      "before: 222\n",
      "unique names: 43\n",
      "after: 65 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC14-L_S112_ME_L001__: 100%|█████████████████████████████████████████████████████████| 80/80 [12:48<00:00,  9.61s/it]\n",
      "OC14-L_S110_ME_L001__: 100%|████████████████████████████████████████████████████████████| 65/65 [11:39<00:00, 10.77s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC33-L ---------------------------------------------------\n",
      "metatable shape: (479, 32)\n",
      "sample file names: ['MN-OC33-L_S116_ME_L001__', 'OC33-L_S114_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 125   including test: 0\n",
      "somatic and quasi-somatic: (354, 32)   including test: 0 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__ 107   test: 0\n",
      "OC33-L_S114_ME_L001__ 247   test: 0\n",
      "overall num of somatic insertions:  354 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__  was: 107  become: 16\n",
      "MN-OC33-L_S116_ME_L001__ reads were written\n",
      "OC33-L_S114_ME_L001__  was: 247  become: 37\n",
      "OC33-L_S114_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC33-L_S116_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.04s/it]\n",
      "MN-OC33-L_S116_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.30s/it]\n",
      "OC33-L_S114_ME_L001___R1: 100%|███████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.21s/it]\n",
      "OC33-L_S114_ME_L001___R2: 100%|███████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.81s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.02s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 252\n",
      "unique names: 12\n",
      "after: 32 \n",
      "\n",
      "before: 121\n",
      "unique names: 20\n",
      "after: 34 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC33-L_S116_ME_L001__: 100%|█████████████████████████████████████████████████████████| 32/32 [10:01<00:00, 18.80s/it]\n",
      "OC33-L_S114_ME_L001__: 100%|████████████████████████████████████████████████████████████| 34/34 [04:43<00:00,  8.33s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC40-L ---------------------------------------------------\n",
      "metatable shape: (989, 32)\n",
      "sample file names: ['MN-OC40-L_S80_ME_L001__', 'OC40-L_S78_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 352   including test: 0\n",
      "somatic and quasi-somatic: (637, 32)   including test: 0 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__ 274   test: 0\n",
      "OC40-L_S78_ME_L001__ 363   test: 0\n",
      "overall num of somatic insertions:  637 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__  was: 274  become: 60\n",
      "MN-OC40-L_S80_ME_L001__ reads were written\n",
      "OC40-L_S78_ME_L001__  was: 363  become: 80\n",
      "OC40-L_S78_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC40-L_S80_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 6/6 [00:11<00:00,  1.91s/it]\n",
      "MN-OC40-L_S80_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 6/6 [00:14<00:00,  2.44s/it]\n",
      "OC40-L_S78_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 8/8 [00:17<00:00,  2.13s/it]\n",
      "OC40-L_S78_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 8/8 [00:21<00:00,  2.71s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.54s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 41\n",
      "unique names: 32\n",
      "after: 41 \n",
      "\n",
      "before: 148\n",
      "unique names: 48\n",
      "after: 74 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC40-L_S80_ME_L001__: 100%|██████████████████████████████████████████████████████████| 41/41 [07:37<00:00, 11.16s/it]\n",
      "OC40-L_S78_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 74/74 [11:01<00:00,  8.94s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC42-L ---------------------------------------------------\n",
      "metatable shape: (2375, 32)\n",
      "sample file names: ['MN-OC42-L_S40_ME_L001__', 'OC42-L_S38_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 25   including test: 0\n",
      "somatic and quasi-somatic: (2350, 32)   including test: 0 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__ 2346   test: 0\n",
      "OC42-L_S38_ME_L001__ 4   test: 0\n",
      "overall num of somatic insertions:  2350 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__  was: 2346  become: 603\n",
      "MN-OC42-L_S40_ME_L001__ reads were written\n",
      "OC42-L_S38_ME_L001__  was: 4  become: 2\n",
      "OC42-L_S38_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC42-L_S40_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 60/60 [01:41<00:00,  1.69s/it]\n",
      "MN-OC42-L_S40_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 60/60 [01:50<00:00,  1.84s/it]\n",
      "OC42-L_S38_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "OC42-L_S38_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [01:51<00:00, 55.90s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 8143\n",
      "unique names: 242\n",
      "after: 735 \n",
      "\n",
      "before: 2\n",
      "unique names: 2\n",
      "after: 2 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC42-L_S40_ME_L001__: 100%|██████████████████████████████████████████████████████| 735/735 [1:45:54<00:00,  8.65s/it]\n",
      "OC42-L_S38_ME_L001__: 100%|███████████████████████████████████████████████████████████████| 2/2 [00:26<00:00, 13.08s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC48-L ---------------------------------------------------\n",
      "metatable shape: (249, 32)\n",
      "sample file names: ['MN-OC48-L_S44_ME_L001__', 'OC48-L_S42_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 60   including test: 0\n",
      "somatic and quasi-somatic: (189, 32)   including test: 0 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__ 57   test: 0\n",
      "OC48-L_S42_ME_L001__ 132   test: 0\n",
      "overall num of somatic insertions:  189 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__  was: 57  become: 6\n",
      "MN-OC48-L_S44_ME_L001__ reads were written\n",
      "OC48-L_S42_ME_L001__  was: 132  become: 18\n",
      "OC48-L_S42_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC48-L_S44_ME_L001___R1: 0it [00:00, ?it/s]\n",
      "MN-OC48-L_S44_ME_L001___R2: 0it [00:00, ?it/s]\n",
      "OC48-L_S42_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.07s/it]\n",
      "OC48-L_S42_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.61s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.04it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 45\n",
      "unique names: 6\n",
      "after: 10 \n",
      "\n",
      "before: 9\n",
      "unique names: 9\n",
      "after: 9 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC48-L_S44_ME_L001__: 100%|██████████████████████████████████████████████████████████| 10/10 [01:26<00:00,  8.67s/it]\n",
      "OC48-L_S42_ME_L001__: 100%|███████████████████████████████████████████████████████████████| 9/9 [01:50<00:00, 12.32s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC50-L ---------------------------------------------------\n",
      "metatable shape: (1307, 32)\n",
      "sample file names: ['MN-OC50-L_S128_ME_L001__', 'OC50-L_S126_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 502   including test: 0\n",
      "somatic and quasi-somatic: (805, 32)   including test: 0 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__ 336   test: 0\n",
      "OC50-L_S126_ME_L001__ 469   test: 0\n",
      "overall num of somatic insertions:  805 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__  was: 336  become: 64\n",
      "MN-OC50-L_S128_ME_L001__ reads were written\n",
      "OC50-L_S126_ME_L001__  was: 469  become: 109\n",
      "OC50-L_S126_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC50-L_S128_ME_L001___R1: 100%|████████████████████████████████████████████████████████| 6/6 [00:12<00:00,  2.12s/it]\n",
      "MN-OC50-L_S128_ME_L001___R2: 100%|████████████████████████████████████████████████████████| 6/6 [00:15<00:00,  2.58s/it]\n",
      "OC50-L_S126_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 10/10 [00:20<00:00,  2.06s/it]\n",
      "OC50-L_S126_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 10/10 [00:28<00:00,  2.84s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.99s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 157\n",
      "unique names: 33\n",
      "after: 53 \n",
      "\n",
      "before: 270\n",
      "unique names: 58\n",
      "after: 86 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC50-L_S128_ME_L001__: 100%|█████████████████████████████████████████████████████████| 53/53 [10:03<00:00, 11.39s/it]\n",
      "OC50-L_S126_ME_L001__: 100%|████████████████████████████████████████████████████████████| 86/86 [19:27<00:00, 13.58s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OC54-L ---------------------------------------------------\n",
      "metatable shape: (1293, 32)\n",
      "sample file names: ['MN-OC54-L_S56_ME_L001__', 'OC54-L_S54_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 494   including test: 0\n",
      "somatic and quasi-somatic: (799, 32)   including test: 0 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__ 373   test: 0\n",
      "OC54-L_S54_ME_L001__ 426   test: 0\n",
      "overall num of somatic insertions:  799 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__  was: 373  become: 52\n",
      "MN-OC54-L_S56_ME_L001__ reads were written\n",
      "OC54-L_S54_ME_L001__  was: 426  become: 95\n",
      "OC54-L_S54_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC54-L_S56_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.34s/it]\n",
      "MN-OC54-L_S56_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 5/5 [00:14<00:00,  2.91s/it]\n",
      "OC54-L_S54_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 9/9 [00:18<00:00,  2.04s/it]\n",
      "OC54-L_S54_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 9/9 [00:24<00:00,  2.67s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.36s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 15\n",
      "unique names: 12\n",
      "after: 15 \n",
      "\n",
      "before: 358\n",
      "unique names: 54\n",
      "after: 80 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OC54-L_S56_ME_L001__: 100%|██████████████████████████████████████████████████████████| 15/15 [02:24<00:00,  9.65s/it]\n",
      "OC54-L_S54_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 80/80 [14:07<00:00, 10.60s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OT1-L ---------------------------------------------------\n",
      "metatable shape: (640, 32)\n",
      "sample file names: ['MN-OT1-L_S152_ME_L001__', 'OT1-L_S150_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 251   including test: 0\n",
      "somatic and quasi-somatic: (389, 32)   including test: 0 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__ 159   test: 0\n",
      "OT1-L_S150_ME_L001__ 230   test: 0\n",
      "overall num of somatic insertions:  389 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__  was: 159  become: 28\n",
      "MN-OT1-L_S152_ME_L001__ reads were written\n",
      "OT1-L_S150_ME_L001__  was: 230  become: 40\n",
      "OT1-L_S150_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OT1-L_S152_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.94s/it]\n",
      "MN-OT1-L_S152_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.83s/it]\n",
      "OT1-L_S150_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.11s/it]\n",
      "OT1-L_S150_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 4/4 [00:10<00:00,  2.56s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.59s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 227\n",
      "unique names: 13\n",
      "after: 22 \n",
      "\n",
      "before: 444\n",
      "unique names: 25\n",
      "after: 46 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-OT1-L_S152_ME_L001__: 100%|██████████████████████████████████████████████████████████| 22/22 [02:50<00:00,  7.75s/it]\n",
      "OT1-L_S150_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 46/46 [07:23<00:00,  9.64s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC12-L ---------------------------------------------------\n",
      "metatable shape: (1568, 32)\n",
      "sample file names: ['MN-PC12-L_S24_ME_L001__', 'PC12-L_S22_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 214   including test: 0\n",
      "somatic and quasi-somatic: (1354, 32)   including test: 0 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__ 434   test: 0\n",
      "PC12-L_S22_ME_L001__ 920   test: 0\n",
      "overall num of somatic insertions:  1354 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__  was: 434  become: 68\n",
      "MN-PC12-L_S24_ME_L001__ reads were written\n",
      "PC12-L_S22_ME_L001__  was: 920  become: 170\n",
      "PC12-L_S22_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC12-L_S24_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 6/6 [00:13<00:00,  2.30s/it]\n",
      "MN-PC12-L_S24_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 6/6 [00:16<00:00,  2.73s/it]\n",
      "PC12-L_S22_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 17/17 [00:37<00:00,  2.23s/it]\n",
      "PC12-L_S22_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 17/17 [00:53<00:00,  3.15s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.50s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 13\n",
      "unique names: 13\n",
      "after: 13 \n",
      "\n",
      "before: 296\n",
      "unique names: 90\n",
      "after: 123 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC12-L_S24_ME_L001__: 100%|██████████████████████████████████████████████████████████| 13/13 [02:28<00:00, 11.42s/it]\n",
      "PC12-L_S22_ME_L001__: 100%|███████████████████████████████████████████████████████████| 123/123 [19:07<00:00,  9.33s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC14-L ---------------------------------------------------\n",
      "metatable shape: (857, 32)\n",
      "sample file names: ['MN-PC14-L_S36_ME_L001__', 'PC14-L_S34_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 188   including test: 0\n",
      "somatic and quasi-somatic: (669, 32)   including test: 0 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__ 100   test: 0\n",
      "PC14-L_S34_ME_L001__ 569   test: 0\n",
      "overall num of somatic insertions:  669 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__  was: 100  become: 19\n",
      "MN-PC14-L_S36_ME_L001__ reads were written\n",
      "PC14-L_S34_ME_L001__  was: 569  become: 125\n",
      "PC14-L_S34_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC14-L_S36_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.07s/it]\n",
      "MN-PC14-L_S36_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.07s/it]\n",
      "PC14-L_S34_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 12/12 [00:26<00:00,  2.17s/it]\n",
      "PC14-L_S34_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 12/12 [00:44<00:00,  3.73s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:17<00:00,  8.62s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 110\n",
      "unique names: 8\n",
      "after: 20 \n",
      "\n",
      "before: 89\n",
      "unique names: 68\n",
      "after: 84 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC14-L_S36_ME_L001__: 100%|██████████████████████████████████████████████████████████| 20/20 [02:23<00:00,  7.19s/it]\n",
      "PC14-L_S34_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 84/84 [13:13<00:00,  9.45s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC21-L ---------------------------------------------------\n",
      "metatable shape: (1071, 32)\n",
      "sample file names: ['MN-PC21-L_S20_ME_L001__', 'PC21-L_S18_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 337   including test: 0\n",
      "somatic and quasi-somatic: (734, 32)   including test: 0 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__ 578   test: 0\n",
      "PC21-L_S18_ME_L001__ 156   test: 0\n",
      "overall num of somatic insertions:  734 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__  was: 578  become: 108\n",
      "MN-PC21-L_S20_ME_L001__ reads were written\n",
      "PC21-L_S18_ME_L001__  was: 156  become: 41\n",
      "PC21-L_S18_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC21-L_S20_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 10/10 [00:21<00:00,  2.13s/it]\n",
      "MN-PC21-L_S20_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 10/10 [00:28<00:00,  2.80s/it]\n",
      "PC21-L_S18_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.09s/it]\n",
      "PC21-L_S18_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 4/4 [00:11<00:00,  2.87s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.75s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 54\n",
      "unique names: 30\n",
      "after: 36 \n",
      "\n",
      "before: 52\n",
      "unique names: 25\n",
      "after: 33 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC21-L_S20_ME_L001__: 100%|██████████████████████████████████████████████████████████| 36/36 [07:46<00:00, 12.95s/it]\n",
      "PC21-L_S18_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 33/33 [05:39<00:00, 10.28s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC32-L ---------------------------------------------------\n",
      "metatable shape: (2116, 32)\n",
      "sample file names: ['MN-PC32-L_S32_ME_L001__', 'PC32-L_S30_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 110   including test: 0\n",
      "somatic and quasi-somatic: (2006, 32)   including test: 0 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__ 1803   test: 0\n",
      "PC32-L_S30_ME_L001__ 203   test: 0\n",
      "overall num of somatic insertions:  2006 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__  was: 1803  become: 360\n",
      "MN-PC32-L_S32_ME_L001__ reads were written\n",
      "PC32-L_S30_ME_L001__  was: 203  become: 30\n",
      "PC32-L_S30_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC32-L_S32_ME_L001___R1: 100%|███████████████████████████████████████████████████████| 36/36 [01:07<00:00,  1.89s/it]\n",
      "MN-PC32-L_S32_ME_L001___R2: 100%|███████████████████████████████████████████████████████| 36/36 [01:17<00:00,  2.14s/it]\n",
      "PC32-L_S30_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 3/3 [00:05<00:00,  1.67s/it]\n",
      "PC32-L_S30_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.29s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:36<00:00, 18.04s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 3122\n",
      "unique names: 155\n",
      "after: 356 \n",
      "\n",
      "before: 41\n",
      "unique names: 15\n",
      "after: 33 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-PC32-L_S32_ME_L001__: 100%|████████████████████████████████████████████████████████| 356/356 [50:44<00:00,  8.55s/it]\n",
      "PC32-L_S30_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 33/33 [09:12<00:00, 16.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SkC6-L ---------------------------------------------------\n",
      "metatable shape: (284, 32)\n",
      "sample file names: ['MN-SkC6-L_S60_ME_L001__', 'SkC6-L_S58_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 73   including test: 0\n",
      "somatic and quasi-somatic: (211, 32)   including test: 0 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__ 88   test: 0\n",
      "SkC6-L_S58_ME_L001__ 123   test: 0\n",
      "overall num of somatic insertions:  211 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__  was: 88  become: 15\n",
      "MN-SkC6-L_S60_ME_L001__ reads were written\n",
      "SkC6-L_S58_ME_L001__  was: 123  become: 16\n",
      "SkC6-L_S58_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-SkC6-L_S60_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.34s/it]\n",
      "MN-SkC6-L_S60_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.31s/it]\n",
      "SkC6-L_S58_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.02s/it]\n",
      "SkC6-L_S58_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.64s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.42s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 11\n",
      "unique names: 11\n",
      "after: 11 \n",
      "\n",
      "before: 104\n",
      "unique names: 11\n",
      "after: 27 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-SkC6-L_S60_ME_L001__: 100%|██████████████████████████████████████████████████████████| 11/11 [02:20<00:00, 12.74s/it]\n",
      "SkC6-L_S58_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 27/27 [04:42<00:00, 10.47s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "StS1-L ---------------------------------------------------\n",
      "metatable shape: (1234, 32)\n",
      "sample file names: ['MN-StS1-L_S92_ME_L001__', 'StS1-L_S90_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 448   including test: 0\n",
      "somatic and quasi-somatic: (786, 32)   including test: 0 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__ 330   test: 0\n",
      "StS1-L_S90_ME_L001__ 456   test: 0\n",
      "overall num of somatic insertions:  786 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__  was: 330  become: 76\n",
      "MN-StS1-L_S92_ME_L001__ reads were written\n",
      "StS1-L_S90_ME_L001__  was: 456  become: 101\n",
      "StS1-L_S90_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-StS1-L_S92_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 7/7 [00:20<00:00,  2.94s/it]\n",
      "MN-StS1-L_S92_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 7/7 [00:23<00:00,  3.37s/it]\n",
      "StS1-L_S90_ME_L001___R1: 100%|██████████████████████████████████████████████████████████| 10/10 [00:26<00:00,  2.62s/it]\n",
      "StS1-L_S90_ME_L001___R2: 100%|██████████████████████████████████████████████████████████| 10/10 [00:37<00:00,  3.78s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:15<00:00,  7.52s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 539\n",
      "unique names: 39\n",
      "after: 64 \n",
      "\n",
      "before: 198\n",
      "unique names: 48\n",
      "after: 70 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-StS1-L_S92_ME_L001__: 100%|██████████████████████████████████████████████████████████| 64/64 [09:34<00:00,  8.97s/it]\n",
      "StS1-L_S90_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 70/70 [12:29<00:00, 10.71s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "UC2-L ---------------------------------------------------\n",
      "metatable shape: (832, 32)\n",
      "sample file names: ['MN-UC2-L_S124_ME_L001__', 'UC2-L_S122_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 315   including test: 0\n",
      "somatic and quasi-somatic: (517, 32)   including test: 0 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__ 199   test: 0\n",
      "UC2-L_S122_ME_L001__ 318   test: 0\n",
      "overall num of somatic insertions:  517 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__  was: 199  become: 32\n",
      "MN-UC2-L_S124_ME_L001__ reads were written\n",
      "UC2-L_S122_ME_L001__  was: 318  become: 56\n",
      "UC2-L_S122_ME_L001__ reads were written\n",
      "\n",
      "Human BLAT Search\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-UC2-L_S124_ME_L001___R1: 100%|█████████████████████████████████████████████████████████| 3/3 [00:06<00:00,  2.20s/it]\n",
      "MN-UC2-L_S124_ME_L001___R2: 100%|█████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.60s/it]\n",
      "UC2-L_S122_ME_L001___R1: 100%|████████████████████████████████████████████████████████████| 5/5 [00:12<00:00,  2.49s/it]\n",
      "UC2-L_S122_ME_L001___R2: 100%|████████████████████████████████████████████████████████████| 5/5 [00:16<00:00,  3.28s/it]\n",
      "blat analysis:: 100%|█████████████████████████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.75s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "before: 80\n",
      "unique names: 19\n",
      "after: 28 \n",
      "\n",
      "before: 278\n",
      "unique names: 32\n",
      "after: 50 \n",
      "\n",
      "Repeat Annotation Request Form\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MN-UC2-L_S124_ME_L001__: 100%|██████████████████████████████████████████████████████████| 28/28 [04:51<00:00, 10.41s/it]\n",
      "UC2-L_S122_ME_L001__: 100%|█████████████████████████████████████████████████████████████| 50/50 [08:32<00:00, 10.26s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "dirs = os.listdir()[8:]\n",
    "for dirname in dirs:\n",
    "    print(dirname, '---------------------------------------------------')\n",
    "    curr_dir = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC/' + dirname\n",
    "    if not os.path.isdir(curr_dir):\n",
    "        continue\n",
    "    os.chdir(curr_dir)\n",
    "    DIR = os.getcwd() + '/nofiltration'\n",
    "    t = time.time()\n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir)\n",
    "    # 2 Filtration\n",
    "    writedown_filtered(DIR, file_names, df_somatic_list)\n",
    "    # 3 UCSC BLAT parsing\n",
    "    blat_parsing(DIR, file_names)\n",
    "    # 4 BLAT analysis\n",
    "    blat_analysis(DIR, file_names)\n",
    "    # 5 RepeatMasker parsing\n",
    "    blat_res, RM_res = get_10best_alings(DIR, file_names)\n",
    "    RM_parsing(DIR, file_names, blat_res, RM_res)\n",
    "    # 6 Merging\n",
    "    merging(DIR, file_names)\n",
    "\n",
    "    test = telegram_bot_sendtext(f\"Postprocessing done for {dirname}_AGCT\\n\" + f\"Duration: {int(time.time() - t)} sec\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db2dbb9b-3e55-4447-97e7-506a91f3733c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Output data filtration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "854976dc-f742-4bb7-8642-b650a576d7dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def get_somatic(path, show=True):\n",
    "    # read pipeline output\n",
    "    res = pd.read_table(path + '/metatable_ts.txt')\n",
    "    if show:\n",
    "        print('metatable shape:', res.shape)\n",
    "    #print(res.head())\n",
    "\n",
    "    # make list of file names\n",
    "    file_names = []\n",
    "    for col_name in res.columns:\n",
    "        if col_name.find('_NUM_READS') != -1:\n",
    "            file_names.append(col_name[:-10])\n",
    "    if show:\n",
    "        print('sample file names:', file_names)\n",
    "    \n",
    "    # checking for null occurrences \n",
    "    df_0 = res.copy(deep=True)\n",
    "    for i in range(len(file_names)):\n",
    "        df_0 = df_0[df_0[file_names[i] + '_NUM_READS'] == 0]\n",
    "    if show:\n",
    "        print('null occurrences:', df_0.shape[0])\n",
    "    \n",
    "    # not somatic (germline) insertions (exists in all samples)\n",
    "    df_germ = res.copy(deep=True)\n",
    "    for i in range(len(file_names)):\n",
    "        df_germ = df_germ[df_germ[file_names[i] + '_NUM_READS'] > 0]\n",
    "    if show:\n",
    "        print('all not somatic:', df_germ.shape[0],  '  including test:', df_germ[df_germ['READNAME'].str.contains('ARTIFICIAL')].shape[0])\n",
    "    \n",
    "    # somatic and quasi-somatic (exists for example in 3/4 samples) insertions\n",
    "    df_qsom = res.copy(deep=True)\n",
    "    df_qsom = df_qsom.drop(df_germ.index, axis=0)\n",
    "    if show:\n",
    "        print('somatic and quasi-somatic:', df_qsom.shape, '  including test:', df_qsom[df_qsom['READNAME'].str.contains('ARTIFICIAL')].shape[0], '\\n')\n",
    "    \n",
    "    # insertions specific only for this tissue/sample(file) \n",
    "    df_somatic_list = []\n",
    "    overall_num = 0\n",
    "    for i in range(len(file_names)):\n",
    "        df = df_qsom.copy(deep=True)\n",
    "        df = df[df[file_names[i] + '_NUM_READS'] > 0]\n",
    "        for j in range(1, len(file_names)):\n",
    "            df = df[df[file_names[i - j] + '_NUM_READS'] == 0]\n",
    "        if show:\n",
    "            print(file_names[i], df.shape[0], '  test:', df[df['READNAME'].str.contains('ARTIFICIAL')].shape[0])\n",
    "        overall_num += df.shape[0]\n",
    "        df_somatic_list.append(df.copy(deep=True))\n",
    "    if show:\n",
    "        print('overall num of somatic insertions: ', overall_num, '\\n')\n",
    "    return file_names, df_somatic_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "50dc1107-a2f3-49bc-b811-c5ae6289b393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# preparing for copying into blat and filtering \n",
    "def writedown_filtered(DIR, file_names, df_somatic_list):\n",
    "    if not os.path.exists(DIR + '/filtred'):\n",
    "        os.makedirs(DIR + '/filtred')\n",
    "    \n",
    "    for i in range(len(file_names)):\n",
    "        # filtration\n",
    "        df_tmp = df_somatic_list[i].copy(deep=True)\n",
    "        #df_tmp = df_tmp[df_tmp[file_names[i] + '_NUM_READS'] > 1].copy(deep=True)\n",
    "        df_tmp = df_tmp[df_tmp['R1'].str.len() > 25].copy(deep=True)  # because Blat can aling seq > 25nt\n",
    "        df_tmp = df_tmp[df_tmp['R2'].str.len() > 25].copy(deep=True)\n",
    "        df_tmp = df_tmp[df_tmp['MDMATCH'] < (52+6+14 - 2)].copy(deep=True)  # template_switch_filtration: 12 primer + 6 dALU + 14 flank = 32\n",
    "        print(file_names[i], ' was:', df_somatic_list[i].shape[0], ' become:', df_tmp.shape[0])\n",
    "    \n",
    "        df_tmp.to_csv(DIR + '/filtred/' + f'{file_names[i]}_filtered.csv')\n",
    "        print(f'{file_names[i]} reads were written')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4bec163-6591-4b65-b8e4-39d72b250720",
   "metadata": {},
   "source": [
    "### UCSC Blat parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e2b30646-5aad-4efa-bec0-32b37eaf7db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement(char):\n",
    "    if char == 'A':\n",
    "        return 'T'\n",
    "    if char == 'T':\n",
    "        return 'A'\n",
    "    if char == 'C':\n",
    "        return 'G'\n",
    "    if char == 'G':\n",
    "        return 'C'\n",
    "    if char == 'N':\n",
    "        return 'N'\n",
    "\n",
    "def reverse_comp(seq):\n",
    "    seq_ls = list(seq)\n",
    "    new_seq = []\n",
    "    for i in reversed(seq_ls):\n",
    "        new_seq.append(complement(i))\n",
    "    return \"\".join(new_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "660b628e-5486-464d-8562-6231870c1c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reads_fa(reads_list, reads_names, R_n=1):\n",
    "    line = ''\n",
    "    for read, name in zip(reads_list, reads_names):\n",
    "        if R_n == 1:\n",
    "            line = line + '>' + name + f'_R{R_n}' + '\\n' + reverse_comp(read) + '\\n'  # because if insertion is \"+\" then R1 is rev_comp of + strand dna\n",
    "        else:\n",
    "            line = line + '>' + name + f'_R{R_n}' + '\\n' + read + '\\n'\n",
    "    return line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e54449ea-8362-44b4-b448-57d336351ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_blat_output(driver, filtered_flanks, i, k, j, length=10):\n",
    "    # set flank coords and submit\n",
    "    range_input = driver.find_element(By.NAME, \"userSeq\")\n",
    "    range_input.clear()\n",
    "    #push 10 reads at once\n",
    "    read_list = filtered_flanks[i][f'R{k}'].tolist()[j*10 : j*10 + length]\n",
    "    name_list = filtered_flanks[i]['READNAME'].tolist()[j*10 : j*10 + length]\n",
    "    range_input.send_keys(get_reads_fa(read_list, name_list, R_n=k))\n",
    "    \n",
    "    button = driver.find_element(By.NAME, \"Submit\")\n",
    "    button.click()\n",
    "    # get results \n",
    "    blat_out = ''\n",
    "    try:\n",
    "        #print('waiting')\n",
    "        element = WebDriverWait(driver, 60).until(\n",
    "                            EC.presence_of_element_located((By.TAG_NAME, \"pre\")) \n",
    "                            )\n",
    "        blat_out = element.text\n",
    "    except:\n",
    "        print(f'exception for {file_names[i]}')\n",
    "    # get back to request form       \n",
    "    driver.back()\n",
    "    return blat_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a3b1ba3e-b957-4e9f-a196-317bbded61e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import trange\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.wait import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.support.select import Select\n",
    "\n",
    "def blat_parsing(DIR, file_names):\n",
    "    service = Service(executable_path=\"/usr/bin/chromedriver\")\n",
    "    chrome_options = Options()\n",
    "    chrome_options.add_argument('--headless')\n",
    "    chrome_options.add_argument('--no-sandbox')\n",
    "    chrome_options.add_argument('--disable-dev-shm-usage')\n",
    "    driver = webdriver.Chrome(service=service, options=chrome_options)\n",
    "    \n",
    "    filtered_flanks = []\n",
    "    for i in range(len(file_names)):\n",
    "        df_tmp = pd.read_csv(DIR + '/filtred/' + f'{file_names[i]}_filtered.csv')\n",
    "        filtered_flanks.append(df_tmp.copy(deep=True))\n",
    "    \n",
    "    # go to the UCSC Blat\n",
    "    driver.get(\"https://genome.ucsc.edu/cgi-bin/hgBlat?hgsid=1711647570_KRW9QfFHAHaje8cp8sWT4EtpGupE&command=start\")\n",
    "    print(driver.title)\n",
    "    \n",
    "    # set Hg38\n",
    "    selects = driver.find_elements(By.TAG_NAME, \"select\")\n",
    "    for i in selects:\n",
    "        if i.get_property(\"name\") == \"db\":\n",
    "            element = i\n",
    "            break\n",
    "    dropdown = Select(element)\n",
    "    dropdown.select_by_visible_text(\"Dec. 2013 (GRCh38/hg38)\")\n",
    "    \n",
    "    if not os.path.exists(DIR + '/blat_output'):\n",
    "        os.makedirs(DIR + '/blat_output')\n",
    "    \n",
    "    for i in range(len(filtered_flanks)):\n",
    "        for k in [1, 2]:\n",
    "            output = ''\n",
    "            for j in trange(int(filtered_flanks[i].shape[0] / 10), desc=f'{file_names[i]}_R{k}'):  \n",
    "                output = output + '\\n' + get_blat_output(driver, filtered_flanks, i, k, j)\n",
    "            if filtered_flanks[i].shape[0] % 10 != 0:\n",
    "                output = output + '\\n' + get_blat_output(driver, filtered_flanks, i, k, int(filtered_flanks[i].shape[0] / 10), \n",
    "                                                         length=filtered_flanks[i].shape[0] % 10)\n",
    "                \n",
    "            with open(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out_tmp.txt', 'w') as f:\n",
    "                f.write(output)\n",
    "                #print('res written')\n",
    "    \n",
    "    driver.quit()\n",
    "\n",
    "    for i in range(len(filtered_flanks)):\n",
    "        for k in [1, 2]:\n",
    "            blat_out = pd.DataFrame(columns=['QUERY', 'SCORE', 'QSTART', 'QEND', 'QSIZE', 'IDENTITY', 'CHROM', 'STRAND', 'START', 'END', 'SPAN'])\n",
    "            with open(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out_tmp.txt', 'r') as f:\n",
    "                output = f.read()\n",
    "            output = output.split('\\n')\n",
    "            for line in output:\n",
    "                    if line.count('browser'):\n",
    "                        tmp = line.split(' ')\n",
    "                        tmp = ' '.join(tmp).split()\n",
    "                        blat_out.loc[len(blat_out.index)] = [tmp[2], tmp[3], tmp[4], tmp[5], tmp[6], tmp[7], tmp[8], tmp[9], tmp[10], tmp[11], tmp[12]]\n",
    "            blat_out.to_csv(DIR + '/blat_output/' + f'{file_names[i]}_R{k}_blat_out.csv') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81a7be26-c0e6-4f9b-965b-5109ac089de3",
   "metadata": {},
   "source": [
    "### Blat analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7251a234-dbfa-471c-8fd1-e468b75f6aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import trange\n",
    "\n",
    "def blat_merge_and_filtration(filename):\n",
    "\n",
    "    FRAGMENT_LEN = 1000  # from electroforeses flanks will be from ~60 to ~585\n",
    "\n",
    "    #chomosome list\n",
    "    chromosomes = list(range(1, 23))\n",
    "    chromosomes.append('X')\n",
    "    chromosomes.append('Y')\n",
    "    chromosomes = ['chr' + str(x) for x in chromosomes]\n",
    "    \n",
    "    df1 = pd.read_table(filename + '_R1_blat_out.csv', sep=',')\n",
    "    df1 = df1.drop(['Unnamed: 0'], axis=1)\n",
    "    \n",
    "    df2 = pd.read_table(filename + '_R2_blat_out.csv', sep=',')\n",
    "    df2 = df2.drop(['Unnamed: 0'], axis=1)\n",
    "    \n",
    "    read_names = df1['QUERY'].unique()\n",
    "    merget_df = pd.DataFrame(columns=['QUERY', \n",
    "                                      'SCORE_1', 'QSIZE_1', 'QSTART_1', 'QEND_1', 'IDENTITY_1', 'CHROM_1', 'STRAND_1', 'START_1', 'END_1', 'SPAN_1',\n",
    "                                      'SCORE_2', 'QSIZE_2', 'QSTART_2', 'QEND_2', 'IDENTITY_2', 'CHROM_2', 'STRAND_2', 'START_2', 'END_2', 'SPAN_2'\n",
    "                                     ]\n",
    "                            )\n",
    "    for name in read_names:\n",
    "        df1_1 = df1[df1['QUERY'].str.contains(name)].copy(deep=True)\n",
    "        df2_1 = df2[df2['QUERY'].str.contains(name[:-3])].copy(deep=True)\n",
    "        for i, row_1 in df1_1.iterrows():\n",
    "            tmp_df = df2_1[df2_1['CHROM'] == row_1['CHROM']]\n",
    "            for i, row_tmp in tmp_df.iterrows():\n",
    "                if abs(row_tmp['START'] - row_1['START']) <= FRAGMENT_LEN and abs(row_tmp['END'] - row_1['END']) <= FRAGMENT_LEN:\n",
    "                    merget_df.loc[len(merget_df.index)] = [row_1['QUERY'][:-3], \n",
    "                                                           row_1['SCORE'], row_1['QSIZE'], row_1['QSTART'], row_1['QEND'], row_1['IDENTITY'],\n",
    "                                                           row_1['CHROM'], row_1['STRAND'], row_1['START'], row_1['END'], row_1['SPAN'],\n",
    "                                                           row_tmp['SCORE'], row_tmp['QSIZE'], row_tmp['QSTART'], row_tmp['QEND'], row_tmp['IDENTITY'], \n",
    "                                                           row_tmp['CHROM'], row_tmp['STRAND'], row_tmp['START'], row_tmp['END'], row_tmp['SPAN'] \n",
    "                                                          ]\n",
    "    #print(name[:-3], df1_1.shape[0], df2_1.shape[0], merget_df.shape[0])\n",
    "    \n",
    "    #drop alt chr\n",
    "    merget_df.drop(merget_df[ ~merget_df['CHROM_1'].isin(chromosomes) ].index, axis=0, inplace=True)\n",
    "    #drop matches with different directions \n",
    "    merget_df = merget_df[merget_df['STRAND_1'] == merget_df['STRAND_2']].copy(deep=True)\n",
    "\n",
    "    # get only high score\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df = merget_df[merget_df['SCORE_1'] > 0.8 * merget_df['QSIZE_1']].copy(deep=True)\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df = merget_df[merget_df['SCORE_2'] > 0.8 * merget_df['QSIZE_2']].copy(deep=True)\n",
    "\n",
    "    # 5' Alu\n",
    "    # drop aligns which are not close to RE 5'end \n",
    "    # if merget_df.shape[0] > 1:\n",
    "    #     merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['QEND_1'] <= merget_df['QSIZE_1'] - 3) ].index, axis=0, inplace=True)   \n",
    "    # if merget_df.shape[0] > 1:\n",
    "    #     merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['QSTART_1'] >= 3) ].index, axis=0, inplace=True)\n",
    "    \n",
    "    # 3' LINE\n",
    "    # drop aligns which are not close to RE 3'end \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['QEND_1'] <= merget_df['QSIZE_1'] - 3) ].index, axis=0, inplace=True)   \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['QSTART_1'] >= 3) ].index, axis=0, inplace=True)\n",
    "    \n",
    "\n",
    "    # 5' Alu\n",
    "    # drop strange aligns: \"+\" R2_start >= R1_end; \"-\" R1_start >= R2_end\n",
    "    # if merget_df.shape[0] > 1:\n",
    "    #     merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['START_2'] >= merget_df['END_1']) ].index, axis=0, inplace=True)   \n",
    "    # if merget_df.shape[0] > 1:\n",
    "    #     merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['START_1'] >= merget_df['END_2']) ].index, axis=0, inplace=True)\n",
    "\n",
    "    # 3' LINE\n",
    "    # drop strange aligns: \"+\" R2_start >= R1_end; \"-\" R1_start >= R2_end\n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='+') & (merget_df['START_1'] >= merget_df['END_2']) ].index, axis=0, inplace=True)   \n",
    "    if merget_df.shape[0] > 1:\n",
    "        merget_df.drop(merget_df[ (merget_df['STRAND_1']=='-') & (merget_df['START_2'] >= merget_df['END_1']) ].index, axis=0, inplace=True)\n",
    "    \n",
    "    # calc flank metrics and insertion coord \n",
    "    if merget_df.shape[0] > 0:\n",
    "        for i, row in merget_df.iterrows():\n",
    "            merget_df.loc[i, 'F_CHR'] = row['CHROM_1']\n",
    "            start = int(min(row['START_1'], row['START_2'], row['END_1'], row['END_2']))\n",
    "            end = int(max(row['START_1'], row['START_2'], row['END_1'], row['END_2']))\n",
    "            merget_df.loc[i, 'F_START'] = start\n",
    "            merget_df.loc[i, 'F_END'] = end\n",
    "            merget_df.loc[i, 'F_LEN'] = int(end - start) + 1\n",
    "            if row['STRAND_1'] == '+':\n",
    "                merget_df.loc[i, 'INS_COORD'] = start - 1  # for 3' LINE\n",
    "            else:\n",
    "                merget_df.loc[i, 'INS_COORD'] = end   # for 3' LINE\n",
    "            \n",
    "    return merget_df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "257ca7e7-032a-4492-a2e8-6a659b1641b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def blat_analysis(DIR, file_names):\n",
    "    if not os.path.exists(DIR + '/blat_filtred'):\n",
    "        os.makedirs(DIR + '/blat_filtred')\n",
    "    \n",
    "    for i in trange(len(file_names), desc='blat analysis:'):\n",
    "        df_tmp = blat_merge_and_filtration(DIR + '/blat_output/' + file_names[i])\n",
    "        df_tmp.to_csv(DIR + '/blat_filtred/' + f'{file_names[i]}_blat_filtred.csv')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3128433c-7822-4342-bb60-f966cd19ecf3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### RepeatMasker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "39de57f6-133d-4996-bb6b-9fd374e765a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get 5 best aligns for each read\n",
    "import pandas as pd\n",
    "\n",
    "def get_10best_alings(DIR, file_names):\n",
    "    blat_res = []\n",
    "    RM_res = []\n",
    "    for name in file_names:\n",
    "        blat_res.append(pd.read_table(DIR + '/blat_filtred/' + f'{name}_blat_filtred.csv', sep=','))\n",
    "        RM_res.append(pd.DataFrame(columns=['QUERY', 'EXTENSION', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', 'REPEAT_NAME', 'R_STRAND', 'R_CHR', \n",
    "                                            'R_START', 'R_END', 'INS_OVERLAP', 'DESC', 'F_LEN', 'R_LEN', 'F_START', 'F_END',\n",
    "                                            'SCORE_1', 'QSIZE_1', 'SCORE_2', 'QSIZE_2', 'CHR_1', 'START_1', 'END_1', 'CHR_2', 'START_2', 'END_2', \n",
    "                                            ]))\n",
    "\n",
    "    for blat_res_part in blat_res:\n",
    "        print('before:', blat_res_part.shape[0])\n",
    "        amount_list = blat_res_part.groupby(['QUERY']).count().sort_values(by=['QUERY'])['SCORE_1'].tolist()\n",
    "        read_names = blat_res_part['QUERY'].unique()\n",
    "        read_names.sort()\n",
    "        print('unique names:', len(read_names))\n",
    "        for num, name in zip(amount_list, read_names):\n",
    "            if num > 5:\n",
    "                indexes = blat_res_part[blat_res_part['QUERY'].str.contains(name)].index.tolist()[5:]\n",
    "                blat_res_part.drop(indexes, axis=0, inplace=True)\n",
    "                \n",
    "        print('after:', blat_res_part.shape[0], '\\n')\n",
    "\n",
    "    return blat_res, RM_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d3bfd31e-1ec0-4191-aee1-2ed383ae2017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return distanse from the insertion to the repeat nearest end and description \n",
    "# if they overlap - positive number, else - negative\n",
    "def calc_overlap(ins_coord, r_start, r_end):\n",
    "    ins_coord = int(ins_coord)\n",
    "    r_start = int(r_start)\n",
    "    r_end = int(r_end)\n",
    "    if r_start < ins_coord:\n",
    "        if r_end < ins_coord:\n",
    "            return r_end - ins_coord, 'ins after the repeat'  # negative - insertion after the repeat\n",
    "        else:\n",
    "            return min(r_end - ins_coord, ins_coord - r_start), 'ins in the repeat'  # positive - insertion in the repeat\n",
    "    else:\n",
    "        # r_start >= ins_coord\n",
    "        return ins_coord - r_start, 'ins before the repeat'  # negative - insertion before the repeat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1a898c34-d74d-47c4-9b7e-b931db426167",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import trange\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.wait import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.support.select import Select\n",
    "\n",
    "def RM_parsing(DIR, file_names, blat_res, RM_res):\n",
    "    service = Service(executable_path=\"/usr/bin/chromedriver\")\n",
    "    chrome_options = Options()\n",
    "    chrome_options.add_argument('--headless')\n",
    "    chrome_options.add_argument('--no-sandbox')\n",
    "    chrome_options.add_argument('--disable-dev-shm-usage')\n",
    "    driver = webdriver.Chrome(service=service, options=chrome_options)\n",
    "    \n",
    "    # go to the RepeatMasker\n",
    "    driver.get(\"https://www.repeatmasker.org/cgi-bin/AnnotationRequest\")\n",
    "    print(driver.title)\n",
    "    \n",
    "    # set Hg38\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(\n",
    "            EC.presence_of_element_located((By.NAME, \"genome\"))\n",
    "        )\n",
    "        dropdown = Select(element)\n",
    "        dropdown.select_by_visible_text(\"Human - Dec 2013 - hg38\")\n",
    "    except:\n",
    "        print(\"can't select genome\")\n",
    "    \n",
    "    \n",
    "    if not os.path.exists(DIR + '/RepeatMasker'):\n",
    "        os.makedirs(DIR + '/RepeatMasker')\n",
    "    \n",
    "    # flank coords annotating\n",
    "    j = 0\n",
    "    EXTENSION = 20\n",
    "    for blat_res_part, RM_res_part in zip(blat_res, RM_res):\n",
    "        blat_res_part = blat_res_part.reset_index()\n",
    "        for i in trange(blat_res_part.shape[0], desc=f'{file_names[j]}'):\n",
    "            # set flank coords and submit\n",
    "            try:\n",
    "                range_input = WebDriverWait(driver, 15).until(\n",
    "                              EC.presence_of_element_located((By.NAME, \"range\")) \n",
    "                              )\n",
    "                #driver.find_element(By.NAME, \"range\")\n",
    "                range_input.clear()\n",
    "    \n",
    "                low_coord = int(blat_res_part.loc[i, 'INS_COORD'] - EXTENSION)\n",
    "                high_coord = int(blat_res_part.loc[i, 'INS_COORD'] + EXTENSION)\n",
    "                chr = blat_res_part.loc[i, 'F_CHR']\n",
    "                coords = f\"{chr}:{low_coord}-{high_coord}\" \n",
    "                #print(coords)\n",
    "                range_input.send_keys(coords)\n",
    "               \n",
    "                button = driver.find_element(By.NAME, \".submit\")\n",
    "                button.click()\n",
    "            except:\n",
    "                print(\"can't pass coords\", blat_res_part.loc[i, 'QUERY'])\n",
    "        \n",
    "            # get results \n",
    "            try:\n",
    "                element = WebDriverWait(driver, 15).until(\n",
    "                          EC.presence_of_element_located((By.TAG_NAME, \"pre\")) \n",
    "                          )\n",
    "                #link = element.find_element(By.TAG_NAME, \"a\").get_attribute(\"href\")\n",
    "                output = element.text.split('\\n')[2:]\n",
    "                #print(output)\n",
    "                for line in output:\n",
    "                    tmp_line = line.split(' ')\n",
    "                    tmp_line = ' '.join(tmp_line).split()\n",
    "                    #print(tmp_line)\n",
    "                    overlapping = calc_overlap(blat_res_part.loc[i, 'INS_COORD'], tmp_line[5], tmp_line[6])\n",
    "                    #print(overlapping)\n",
    "                    RM_res_part.loc[len(RM_res_part.index)] = [blat_res_part.loc[i, 'QUERY'], EXTENSION, blat_res_part.loc[i, 'F_CHR'],\n",
    "                                                               blat_res_part.loc[i, 'INS_COORD'], blat_res_part.loc[i, 'STRAND_1'], tmp_line[9], \n",
    "                                                               tmp_line[8], tmp_line[4], tmp_line[5], tmp_line[6], overlapping[0], overlapping[1], \n",
    "                                                               blat_res_part.loc[i, 'F_LEN'], int(tmp_line[6]) - int(tmp_line[5]) + 1, \n",
    "                                                               blat_res_part.loc[i, 'F_START'], blat_res_part.loc[i, 'F_END'],\n",
    "                                                               blat_res_part.loc[i, 'SCORE_1'], blat_res_part.loc[i, 'QSIZE_1'], \n",
    "                                                               blat_res_part.loc[i, 'SCORE_2'], blat_res_part.loc[i, 'QSIZE_2'],\n",
    "                                                               blat_res_part.loc[i, 'CHROM_1'], blat_res_part.loc[i, 'START_1'], \n",
    "                                                               blat_res_part.loc[i, 'END_1'], blat_res_part.loc[i, 'CHROM_2'], \n",
    "                                                               blat_res_part.loc[i, 'START_2'], blat_res_part.loc[i, 'END_2'],\n",
    "                                                              ]\n",
    "            except:\n",
    "                #print('no matching')\n",
    "                RM_res_part.loc[len(RM_res_part.index)] = [blat_res_part.loc[i, 'QUERY'], EXTENSION, blat_res_part.loc[i, 'F_CHR'],\n",
    "                                                           blat_res_part.loc[i, 'INS_COORD'], blat_res_part.loc[i, 'STRAND_1'], '-', '-', \n",
    "                                                           '-', 0, 0, 0, 'no repeat', blat_res_part.loc[i, 'F_LEN'],\n",
    "                                                           0, blat_res_part.loc[i, 'F_START'], blat_res_part.loc[i, 'F_END'],\n",
    "                                                           blat_res_part.loc[i, 'SCORE_1'], blat_res_part.loc[i, 'QSIZE_1'], \n",
    "                                                           blat_res_part.loc[i, 'SCORE_2'], blat_res_part.loc[i, 'QSIZE_2'],\n",
    "                                                           blat_res_part.loc[i, 'CHROM_1'], blat_res_part.loc[i, 'START_1'], \n",
    "                                                           blat_res_part.loc[i, 'END_1'], blat_res_part.loc[i, 'CHROM_2'], \n",
    "                                                           blat_res_part.loc[i, 'START_2'], blat_res_part.loc[i, 'END_2'],\n",
    "                                                          ]\n",
    "            # get back to request form\n",
    "            driver.back()\n",
    "        RM_res_part.to_csv(DIR + '/RepeatMasker/' + f'{file_names[j]}_RM_res.csv')\n",
    "        j += 1\n",
    "    \n",
    "    #quit from site\n",
    "    driver.quit()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4412bf-df02-4c68-8bde-2c7cff267be6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Merging with pipeline info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "698dba1a-3f78-4ad1-86e9-509a6e2f66cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging with pipeline info\n",
    "def attach_pipeline_info(filename, df_res_part):\n",
    "    filtered_flank = pd.read_csv(DIR + '/filtred/' + f'{filename}_filtered.csv')\n",
    "    \n",
    "    right_df = filtered_flank[['READNAME', 'CHR', 'POS', 'INS_STRAND', f'{filename}_NUM_READS', 'R1', 'R2']].copy(deep=True)\n",
    "    left_df = df_res_part.copy(deep=True)\n",
    "    \n",
    "    return left_df.merge(right_df, left_on = 'QUERY', right_on = 'READNAME').drop(columns='READNAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f9c1b72-05d8-4cbd-b02d-99a99e69d61b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_dist(x1, x2):\n",
    "    j = 0\n",
    "    for i in range(len(x1)):\n",
    "        if x1[i] != x2[i]:\n",
    "            j = j + 1\n",
    "    return j\n",
    "\n",
    "# return length of overlapping part\n",
    "def find_overlap_len(str1, str2):\n",
    "    if len(str1) < len(str2):\n",
    "        str_tmp = str1\n",
    "        str1 = str2\n",
    "        str2 = str_tmp\n",
    "    if len(str1) >= len(str2):\n",
    "        delta = len(str1) - len(str2)\n",
    "        for i in range(len(str1)):\n",
    "            if delta - i >= 0:\n",
    "                part1 = str1[i:len(str2) + i]\n",
    "                part2 = str2\n",
    "            else:\n",
    "                part1 = str1[i:]\n",
    "                part2 = str2[0:len(str2) + delta - i]\n",
    "            # print(part1, part2)\n",
    "            if hamming_dist(part1, part2) == 0:\n",
    "                #print(part1, len(part1))\n",
    "                return len(part1)\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c8f7b0ca-e77f-434e-911b-fb373afac65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging with pipeline info\n",
    "# reads overlap calculation\n",
    "def merging(DIR, file_names):\n",
    "    RM_res_list = []\n",
    "    i = 0\n",
    "    for name in file_names:\n",
    "        RM_res_list.append(pd.read_table(DIR + '/RepeatMasker/' + f'{name}_RM_res.csv', sep=','))\n",
    "        RM_res_list[i] = attach_pipeline_info(name, RM_res_list[i])\n",
    "        for j, row in RM_res_list[i].iterrows():\n",
    "            RM_res_list[i].loc[j, 'READS_OVERLAP'] = find_overlap_len(row['R1'], reverse_comp(row['R2']))\n",
    "        RM_res_list[i].to_csv(DIR + '/RepeatMasker/' + f'{name}_RM_res_merged.csv')\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5705f834-5d69-4b35-baca-1a0ba86510b7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e639e5c-4b65-49e3-bfee-6bf25aefb4a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def text_replace(text):\n",
    "    if len(text) == 0:\n",
    "        return ' '\n",
    "    text = text.replace('_', '\\\\_')\n",
    "    text = text.replace('*', '\\\\*')\n",
    "    text = text.replace('[', '\\\\[')\n",
    "    text = text.replace(']', '\\\\]')\n",
    "    text = text.replace('(', '\\\\(')\n",
    "    text = text.replace(')', '\\\\)')\n",
    "    text = text.replace('~', '\\\\~')\n",
    "    text = text.replace('`', '\\\\`')\n",
    "    text = text.replace('>', '\\\\>')\n",
    "    text = text.replace('#', '\\\\#')\n",
    "    text = text.replace('+', '\\\\+')\n",
    "    text = text.replace('-', '\\\\-')\n",
    "    text = text.replace('=', '\\\\=')\n",
    "    text = text.replace('|', '\\\\|')\n",
    "    text = text.replace('{', '\\\\{')\n",
    "    text = text.replace('}', '\\\\}')\n",
    "    text = text.replace('.', '\\\\.')\n",
    "    text = text.replace('!', '\\\\!')\n",
    "    return text\n",
    "\n",
    "def telegram_bot_sendtext(bot_message):\n",
    "    bot_message = text_replace(bot_message)\n",
    "    BOT_TOKEN = \"6608136916:AAHB_1JsP8lsFLD7Uh4WoiajrUFCZToQ2cQ\"\n",
    "    admin_id = str(360072308)\n",
    "    send_text = 'https://api.telegram.org/bot' + BOT_TOKEN + '/sendMessage?chat_id=' + admin_id + '&parse_mode=Markdown&text=' + bot_message\n",
    "    response = requests.get(send_text)\n",
    "    return response.json()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "761b34db-4dbb-4e0f-8b66-48ffdec1590b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Drop questionable insertions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c27a97c4-861b-4ad1-9ea3-e2dc61c49957",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "def is_questionable(df_row):\n",
    "    # no repeat - ok\n",
    "    if df_row['DESC'] == 'no repeat':\n",
    "        return False\n",
    "    # ins in the repeat\n",
    "    if df_row['DESC'] == 'ins in the repeat':\n",
    "        # repeat is in L1HS, <=L1PA7 or Alu => drop\n",
    "        if df_row['REPEAT_NAME'] == 'L1HS' or \\\n",
    "           ( bool(re.search('L1PA[1-7]', df_row['REPEAT_NAME'])) and (not bool(re.search('L1PA[1-9][0-9]', df_row['REPEAT_NAME']))) ) or \\\n",
    "           bool(re.search('Alu', df_row['REPEAT_NAME'])):\n",
    "            return True\n",
    "    # ins near the repeat\n",
    "    else:\n",
    "        # near Alu in 5nt => drop\n",
    "        if bool(re.search('Alu', df_row['REPEAT_NAME'])) and abs(df_row['INS_OVERLAP']) <= 5:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "\n",
    "def drop_questionable(DIR, file_names):\n",
    "    \n",
    "    for name in file_names:\n",
    "        df = pd.read_table(DIR + '/RepeatMasker/' + f'{name}_RM_res_merged.csv', sep=',')\n",
    "\n",
    "        read_names = df['QUERY'].unique()\n",
    "        print(name, ' candidates num:', len(read_names))\n",
    "        num = 0\n",
    "        for r_name in read_names:\n",
    "            df_tmp = df[df['QUERY'] == r_name].copy(deep=True)\n",
    "            # uniaue match\n",
    "            if len(df_tmp['INS_COORD_B'].unique()) == 1:\n",
    "                # 1 repeat or no repeat\n",
    "                if df_tmp.shape[0] == 1:\n",
    "                    if is_questionable(df_tmp.iloc[0]):\n",
    "                        df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                        num += 1\n",
    "                        #print(r_name)\n",
    "                # 2 repeats in the area of the ins\n",
    "                else:\n",
    "                    for i, row in df_tmp.iterrows():\n",
    "                        if is_questionable(row):\n",
    "                            df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                            num += 1\n",
    "                            #print(r_name)\n",
    "                            break\n",
    "            # not unique => drop\n",
    "            else:\n",
    "                df.drop(df_tmp.index, axis=0, inplace=True)\n",
    "                num += 1\n",
    "                #print(r_name)\n",
    "        print(name, ' dropped:', num)\n",
    "        df.to_csv(DIR + '/result/' + f'{name}_result.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e7cb77e8-6adb-41d0-950b-ce6bf9ea62f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC60-L ---------------------------------------------------\n",
      "metatable shape: (279, 32)\n",
      "sample file names: ['BC60-L_S106_ME_L001__', 'MN-BC60-L_S108_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 80   including test: 0\n",
      "somatic and quasi-somatic: (199, 32)   including test: 0 \n",
      "\n",
      "BC60-L_S106_ME_L001__ 97   test: 0\n",
      "MN-BC60-L_S108_ME_L001__ 102   test: 0\n",
      "overall num of somatic insertions:  199 \n",
      "\n",
      "BC60-L_S106_ME_L001__  candidates num: 9\n",
      "BC60-L_S106_ME_L001__  dropped: 0\n",
      "MN-BC60-L_S108_ME_L001__  candidates num: 13\n",
      "MN-BC60-L_S108_ME_L001__  dropped: 0\n",
      "BC66-L ---------------------------------------------------\n",
      "metatable shape: (275, 32)\n",
      "sample file names: ['BC66-L_S102_ME_L001__', 'MN-BC66-L_S104_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 75   including test: 0\n",
      "somatic and quasi-somatic: (200, 32)   including test: 0 \n",
      "\n",
      "BC66-L_S102_ME_L001__ 83   test: 0\n",
      "MN-BC66-L_S104_ME_L001__ 117   test: 0\n",
      "overall num of somatic insertions:  200 \n",
      "\n",
      "BC66-L_S102_ME_L001__  candidates num: 11\n",
      "BC66-L_S102_ME_L001__  dropped: 3\n",
      "MN-BC66-L_S104_ME_L001__  candidates num: 12\n",
      "MN-BC66-L_S104_ME_L001__  dropped: 1\n",
      "BC70-L ---------------------------------------------------\n",
      "metatable shape: (270, 32)\n",
      "sample file names: ['BC70-L_S142_ME_L001__', 'MN-BC70-L_S144_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 9   including test: 0\n",
      "somatic and quasi-somatic: (261, 32)   including test: 0 \n",
      "\n",
      "BC70-L_S142_ME_L001__ 26   test: 0\n",
      "MN-BC70-L_S144_ME_L001__ 235   test: 0\n",
      "overall num of somatic insertions:  261 \n",
      "\n",
      "BC70-L_S142_ME_L001__  candidates num: 2\n",
      "BC70-L_S142_ME_L001__  dropped: 0\n",
      "MN-BC70-L_S144_ME_L001__  candidates num: 36\n",
      "MN-BC70-L_S144_ME_L001__  dropped: 3\n",
      "BCL1-L ---------------------------------------------------\n",
      "metatable shape: (1059, 32)\n",
      "sample file names: ['BLC1-L_S74_ME_L001__', 'MN-BLC1-L_S76_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 402   including test: 0\n",
      "somatic and quasi-somatic: (657, 32)   including test: 0 \n",
      "\n",
      "BLC1-L_S74_ME_L001__ 441   test: 0\n",
      "MN-BLC1-L_S76_ME_L001__ 216   test: 0\n",
      "overall num of somatic insertions:  657 \n",
      "\n",
      "BLC1-L_S74_ME_L001__  candidates num: 35\n",
      "BLC1-L_S74_ME_L001__  dropped: 3\n",
      "MN-BLC1-L_S76_ME_L001__  candidates num: 16\n",
      "MN-BLC1-L_S76_ME_L001__  dropped: 3\n",
      "CC107-L ---------------------------------------------------\n",
      "metatable shape: (764, 32)\n",
      "sample file names: ['CC107-L_S130_ME_L001__', 'MN-CC107-L_S132_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 277   including test: 0\n",
      "somatic and quasi-somatic: (487, 32)   including test: 0 \n",
      "\n",
      "CC107-L_S130_ME_L001__ 298   test: 0\n",
      "MN-CC107-L_S132_ME_L001__ 189   test: 0\n",
      "overall num of somatic insertions:  487 \n",
      "\n",
      "CC107-L_S130_ME_L001__  candidates num: 33\n",
      "CC107-L_S130_ME_L001__  dropped: 10\n",
      "MN-CC107-L_S132_ME_L001__  candidates num: 9\n",
      "MN-CC107-L_S132_ME_L001__  dropped: 2\n",
      "CC127-2-L ---------------------------------------------------\n",
      "metatable shape: (1161, 32)\n",
      "sample file names: ['CC127-2-L_S6_ME_L001__', 'MN-CC127-L_S8_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 346   including test: 0\n",
      "somatic and quasi-somatic: (815, 32)   including test: 0 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__ 578   test: 0\n",
      "MN-CC127-L_S8_ME_L001__ 237   test: 0\n",
      "overall num of somatic insertions:  815 \n",
      "\n",
      "CC127-2-L_S6_ME_L001__  candidates num: 87\n",
      "CC127-2-L_S6_ME_L001__  dropped: 19\n",
      "MN-CC127-L_S8_ME_L001__  candidates num: 22\n",
      "MN-CC127-L_S8_ME_L001__  dropped: 8\n",
      "CC18-L ---------------------------------------------------\n",
      "metatable shape: (1035, 32)\n",
      "sample file names: ['CC18-L_S118_ME_L001__', 'MN-CC18-L_S120_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 334   including test: 0\n",
      "somatic and quasi-somatic: (701, 32)   including test: 0 \n",
      "\n",
      "CC18-L_S118_ME_L001__ 260   test: 0\n",
      "MN-CC18-L_S120_ME_L001__ 441   test: 0\n",
      "overall num of somatic insertions:  701 \n",
      "\n",
      "CC18-L_S118_ME_L001__  candidates num: 24\n",
      "CC18-L_S118_ME_L001__  dropped: 2\n",
      "MN-CC18-L_S120_ME_L001__  candidates num: 66\n",
      "MN-CC18-L_S120_ME_L001__  dropped: 11\n",
      "CC46-L ---------------------------------------------------\n",
      "metatable shape: (805, 32)\n",
      "sample file names: ['CC46-L_S134_ME_L001__', 'MN-CC46-L_S136_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 252   including test: 0\n",
      "somatic and quasi-somatic: (553, 32)   including test: 0 \n",
      "\n",
      "CC46-L_S134_ME_L001__ 213   test: 0\n",
      "MN-CC46-L_S136_ME_L001__ 340   test: 0\n",
      "overall num of somatic insertions:  553 \n",
      "\n",
      "CC46-L_S134_ME_L001__  candidates num: 15\n",
      "CC46-L_S134_ME_L001__  dropped: 3\n",
      "MN-CC46-L_S136_ME_L001__  candidates num: 49\n",
      "MN-CC46-L_S136_ME_L001__  dropped: 23\n",
      "CC5-L ---------------------------------------------------\n",
      "metatable shape: (361, 32)\n",
      "sample file names: ['CC5-L_S146_ME_L001__', 'MN-CC5-L_S148_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 6   including test: 0\n",
      "somatic and quasi-somatic: (355, 32)   including test: 0 \n",
      "\n",
      "CC5-L_S146_ME_L001__ 103   test: 0\n",
      "MN-CC5-L_S148_ME_L001__ 252   test: 0\n",
      "overall num of somatic insertions:  355 \n",
      "\n",
      "CC5-L_S146_ME_L001__  candidates num: 15\n",
      "CC5-L_S146_ME_L001__  dropped: 0\n",
      "MN-CC5-L_S148_ME_L001__  candidates num: 34\n",
      "MN-CC5-L_S148_ME_L001__  dropped: 5\n",
      "CC82-L ---------------------------------------------------\n",
      "metatable shape: (651, 32)\n",
      "sample file names: ['CC82-L_S10_ME_L001__', 'MN-CC82-L_S12_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 103   including test: 0\n",
      "somatic and quasi-somatic: (548, 32)   including test: 0 \n",
      "\n",
      "CC82-L_S10_ME_L001__ 410   test: 0\n",
      "MN-CC82-L_S12_ME_L001__ 138   test: 0\n",
      "overall num of somatic insertions:  548 \n",
      "\n",
      "CC82-L_S10_ME_L001__  candidates num: 46\n",
      "CC82-L_S10_ME_L001__  dropped: 12\n",
      "MN-CC82-L_S12_ME_L001__  candidates num: 9\n",
      "MN-CC82-L_S12_ME_L001__  dropped: 0\n",
      "CC94-L ---------------------------------------------------\n",
      "metatable shape: (124, 32)\n",
      "sample file names: ['CC94-L_S14_ME_L001__', 'MN-CC94-L_S16_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 10   including test: 0\n",
      "somatic and quasi-somatic: (114, 32)   including test: 0 \n",
      "\n",
      "CC94-L_S14_ME_L001__ 88   test: 0\n",
      "MN-CC94-L_S16_ME_L001__ 26   test: 0\n",
      "overall num of somatic insertions:  114 \n",
      "\n",
      "CC94-L_S14_ME_L001__  candidates num: 1\n",
      "CC94-L_S14_ME_L001__  dropped: 0\n",
      "MN-CC94-L_S16_ME_L001__  candidates num: 4\n",
      "MN-CC94-L_S16_ME_L001__  dropped: 2\n",
      "CC96-L ---------------------------------------------------\n",
      "metatable shape: (1660, 32)\n",
      "sample file names: ['CC96-L_S66_ME_L001__', 'MN-CC96-L_S68_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 606   including test: 0\n",
      "somatic and quasi-somatic: (1054, 32)   including test: 0 \n",
      "\n",
      "CC96-L_S66_ME_L001__ 383   test: 0\n",
      "MN-CC96-L_S68_ME_L001__ 671   test: 0\n",
      "overall num of somatic insertions:  1054 \n",
      "\n",
      "CC96-L_S66_ME_L001__  candidates num: 51\n",
      "CC96-L_S66_ME_L001__  dropped: 7\n",
      "MN-CC96-L_S68_ME_L001__  candidates num: 91\n",
      "MN-CC96-L_S68_ME_L001__  dropped: 11\n",
      "CC99-L ---------------------------------------------------\n",
      "metatable shape: (865, 32)\n",
      "sample file names: ['CC99-L_S70_ME_L001__', 'MN-CC99-L_S72_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 265   including test: 0\n",
      "somatic and quasi-somatic: (600, 32)   including test: 0 \n",
      "\n",
      "CC99-L_S70_ME_L001__ 380   test: 0\n",
      "MN-CC99-L_S72_ME_L001__ 220   test: 0\n",
      "overall num of somatic insertions:  600 \n",
      "\n",
      "CC99-L_S70_ME_L001__  candidates num: 39\n",
      "CC99-L_S70_ME_L001__  dropped: 2\n",
      "MN-CC99-L_S72_ME_L001__  candidates num: 17\n",
      "MN-CC99-L_S72_ME_L001__  dropped: 0\n",
      "GC39-L ---------------------------------------------------\n",
      "metatable shape: (1172, 32)\n",
      "sample file names: ['GC39-L_S62_ME_L001__', 'MN-GC39-L_S64_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 493   including test: 0\n",
      "somatic and quasi-somatic: (679, 32)   including test: 0 \n",
      "\n",
      "GC39-L_S62_ME_L001__ 301   test: 0\n",
      "MN-GC39-L_S64_ME_L001__ 378   test: 0\n",
      "overall num of somatic insertions:  679 \n",
      "\n",
      "GC39-L_S62_ME_L001__  candidates num: 29\n",
      "GC39-L_S62_ME_L001__  dropped: 4\n",
      "MN-GC39-L_S64_ME_L001__  candidates num: 33\n",
      "MN-GC39-L_S64_ME_L001__  dropped: 5\n",
      "GC41-L ---------------------------------------------------\n",
      "metatable shape: (971, 32)\n",
      "sample file names: ['GC41-L_S154_ME_L001__', 'MN-GC41-L_S156_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 398   including test: 0\n",
      "somatic and quasi-somatic: (573, 32)   including test: 0 \n",
      "\n",
      "GC41-L_S154_ME_L001__ 300   test: 0\n",
      "MN-GC41-L_S156_ME_L001__ 273   test: 0\n",
      "overall num of somatic insertions:  573 \n",
      "\n",
      "GC41-L_S154_ME_L001__  candidates num: 21\n",
      "GC41-L_S154_ME_L001__  dropped: 2\n",
      "MN-GC41-L_S156_ME_L001__  candidates num: 24\n",
      "MN-GC41-L_S156_ME_L001__  dropped: 3\n",
      "LS1-L ---------------------------------------------------\n",
      "metatable shape: (186, 32)\n",
      "sample file names: ['LS1-L_S138_ME_L001__', 'MN-LS1-L_S140_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 5   including test: 0\n",
      "somatic and quasi-somatic: (181, 32)   including test: 0 \n",
      "\n",
      "LS1-L_S138_ME_L001__ 128   test: 0\n",
      "MN-LS1-L_S140_ME_L001__ 53   test: 0\n",
      "overall num of somatic insertions:  181 \n",
      "\n",
      "LS1-L_S138_ME_L001__  candidates num: 7\n",
      "LS1-L_S138_ME_L001__  dropped: 0\n",
      "MN-LS1-L_S140_ME_L001__  candidates num: 4\n",
      "MN-LS1-L_S140_ME_L001__  dropped: 0\n",
      "LS6-L ---------------------------------------------------\n",
      "metatable shape: (217, 32)\n",
      "sample file names: ['LS6-L_S46_ME_L001__', 'MN-LS6-L_S48_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 49   including test: 0\n",
      "somatic and quasi-somatic: (168, 32)   including test: 0 \n",
      "\n",
      "LS6-L_S46_ME_L001__ 100   test: 0\n",
      "MN-LS6-L_S48_ME_L001__ 68   test: 0\n",
      "overall num of somatic insertions:  168 \n",
      "\n",
      "LS6-L_S46_ME_L001__  candidates num: 7\n",
      "LS6-L_S46_ME_L001__  dropped: 2\n",
      "MN-LS6-L_S48_ME_L001__  candidates num: 6\n",
      "MN-LS6-L_S48_ME_L001__  dropped: 0\n",
      "M14-L ---------------------------------------------------\n",
      "metatable shape: (248, 32)\n",
      "sample file names: ['M14-L_S94_ME_L001__', 'MN-M14-L_S96_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 76   including test: 0\n",
      "somatic and quasi-somatic: (172, 32)   including test: 0 \n",
      "\n",
      "M14-L_S94_ME_L001__ 78   test: 0\n",
      "MN-M14-L_S96_ME_L001__ 94   test: 0\n",
      "overall num of somatic insertions:  172 \n",
      "\n",
      "M14-L_S94_ME_L001__  candidates num: 7\n",
      "M14-L_S94_ME_L001__  dropped: 1\n",
      "MN-M14-L_S96_ME_L001__  candidates num: 11\n",
      "MN-M14-L_S96_ME_L001__  dropped: 1\n",
      "M21-L ---------------------------------------------------\n",
      "metatable shape: (1203, 32)\n",
      "sample file names: ['M21-L_S82_ME_L001__', 'MN-M21-L_S84_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 526   including test: 0\n",
      "somatic and quasi-somatic: (677, 32)   including test: 0 \n",
      "\n",
      "M21-L_S82_ME_L001__ 298   test: 0\n",
      "MN-M21-L_S84_ME_L001__ 379   test: 0\n",
      "overall num of somatic insertions:  677 \n",
      "\n",
      "M21-L_S82_ME_L001__  candidates num: 43\n",
      "M21-L_S82_ME_L001__  dropped: 11\n",
      "MN-M21-L_S84_ME_L001__  candidates num: 34\n",
      "MN-M21-L_S84_ME_L001__  dropped: 7\n",
      "M22-L ---------------------------------------------------\n",
      "metatable shape: (1248, 32)\n",
      "sample file names: ['M22-L_S98_ME_L001__', 'MN-M22-L_S100_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 546   including test: 0\n",
      "somatic and quasi-somatic: (702, 32)   including test: 0 \n",
      "\n",
      "M22-L_S98_ME_L001__ 319   test: 0\n",
      "MN-M22-L_S100_ME_L001__ 383   test: 0\n",
      "overall num of somatic insertions:  702 \n",
      "\n",
      "M22-L_S98_ME_L001__  candidates num: 24\n",
      "M22-L_S98_ME_L001__  dropped: 5\n",
      "MN-M22-L_S100_ME_L001__  candidates num: 36\n",
      "MN-M22-L_S100_ME_L001__  dropped: 4\n",
      "NC5-L ---------------------------------------------------\n",
      "metatable shape: (776, 32)\n",
      "sample file names: ['MN-NC5-L_S88_ME_L001__', 'NC5-L_S86_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 299   including test: 0\n",
      "somatic and quasi-somatic: (477, 32)   including test: 0 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__ 163   test: 0\n",
      "NC5-L_S86_ME_L001__ 314   test: 0\n",
      "overall num of somatic insertions:  477 \n",
      "\n",
      "MN-NC5-L_S88_ME_L001__  candidates num: 23\n",
      "MN-NC5-L_S88_ME_L001__  dropped: 6\n",
      "NC5-L_S86_ME_L001__  candidates num: 22\n",
      "NC5-L_S86_ME_L001__  dropped: 8\n",
      "OC14-L ---------------------------------------------------\n",
      "metatable shape: (1057, 32)\n",
      "sample file names: ['MN-OC14-L_S112_ME_L001__', 'OC14-L_S110_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 421   including test: 0\n",
      "somatic and quasi-somatic: (636, 32)   including test: 0 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__ 331   test: 0\n",
      "OC14-L_S110_ME_L001__ 305   test: 0\n",
      "overall num of somatic insertions:  636 \n",
      "\n",
      "MN-OC14-L_S112_ME_L001__  candidates num: 45\n",
      "MN-OC14-L_S112_ME_L001__  dropped: 11\n",
      "OC14-L_S110_ME_L001__  candidates num: 43\n",
      "OC14-L_S110_ME_L001__  dropped: 6\n",
      "OC33-L ---------------------------------------------------\n",
      "metatable shape: (479, 32)\n",
      "sample file names: ['MN-OC33-L_S116_ME_L001__', 'OC33-L_S114_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 125   including test: 0\n",
      "somatic and quasi-somatic: (354, 32)   including test: 0 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__ 107   test: 0\n",
      "OC33-L_S114_ME_L001__ 247   test: 0\n",
      "overall num of somatic insertions:  354 \n",
      "\n",
      "MN-OC33-L_S116_ME_L001__  candidates num: 12\n",
      "MN-OC33-L_S116_ME_L001__  dropped: 5\n",
      "OC33-L_S114_ME_L001__  candidates num: 20\n",
      "OC33-L_S114_ME_L001__  dropped: 5\n",
      "OC40-L ---------------------------------------------------\n",
      "metatable shape: (989, 32)\n",
      "sample file names: ['MN-OC40-L_S80_ME_L001__', 'OC40-L_S78_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 352   including test: 0\n",
      "somatic and quasi-somatic: (637, 32)   including test: 0 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__ 274   test: 0\n",
      "OC40-L_S78_ME_L001__ 363   test: 0\n",
      "overall num of somatic insertions:  637 \n",
      "\n",
      "MN-OC40-L_S80_ME_L001__  candidates num: 32\n",
      "MN-OC40-L_S80_ME_L001__  dropped: 4\n",
      "OC40-L_S78_ME_L001__  candidates num: 48\n",
      "OC40-L_S78_ME_L001__  dropped: 11\n",
      "OC42-L ---------------------------------------------------\n",
      "metatable shape: (2375, 32)\n",
      "sample file names: ['MN-OC42-L_S40_ME_L001__', 'OC42-L_S38_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 25   including test: 0\n",
      "somatic and quasi-somatic: (2350, 32)   including test: 0 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__ 2346   test: 0\n",
      "OC42-L_S38_ME_L001__ 4   test: 0\n",
      "overall num of somatic insertions:  2350 \n",
      "\n",
      "MN-OC42-L_S40_ME_L001__  candidates num: 242\n",
      "MN-OC42-L_S40_ME_L001__  dropped: 135\n",
      "OC42-L_S38_ME_L001__  candidates num: 2\n",
      "OC42-L_S38_ME_L001__  dropped: 0\n",
      "OC48-L ---------------------------------------------------\n",
      "metatable shape: (249, 32)\n",
      "sample file names: ['MN-OC48-L_S44_ME_L001__', 'OC48-L_S42_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 60   including test: 0\n",
      "somatic and quasi-somatic: (189, 32)   including test: 0 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__ 57   test: 0\n",
      "OC48-L_S42_ME_L001__ 132   test: 0\n",
      "overall num of somatic insertions:  189 \n",
      "\n",
      "MN-OC48-L_S44_ME_L001__  candidates num: 6\n",
      "MN-OC48-L_S44_ME_L001__  dropped: 1\n",
      "OC48-L_S42_ME_L001__  candidates num: 9\n",
      "OC48-L_S42_ME_L001__  dropped: 1\n",
      "OC50-L ---------------------------------------------------\n",
      "metatable shape: (1307, 32)\n",
      "sample file names: ['MN-OC50-L_S128_ME_L001__', 'OC50-L_S126_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 502   including test: 0\n",
      "somatic and quasi-somatic: (805, 32)   including test: 0 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__ 336   test: 0\n",
      "OC50-L_S126_ME_L001__ 469   test: 0\n",
      "overall num of somatic insertions:  805 \n",
      "\n",
      "MN-OC50-L_S128_ME_L001__  candidates num: 33\n",
      "MN-OC50-L_S128_ME_L001__  dropped: 7\n",
      "OC50-L_S126_ME_L001__  candidates num: 58\n",
      "OC50-L_S126_ME_L001__  dropped: 8\n",
      "OC54-L ---------------------------------------------------\n",
      "metatable shape: (1293, 32)\n",
      "sample file names: ['MN-OC54-L_S56_ME_L001__', 'OC54-L_S54_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 494   including test: 0\n",
      "somatic and quasi-somatic: (799, 32)   including test: 0 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__ 373   test: 0\n",
      "OC54-L_S54_ME_L001__ 426   test: 0\n",
      "overall num of somatic insertions:  799 \n",
      "\n",
      "MN-OC54-L_S56_ME_L001__  candidates num: 12\n",
      "MN-OC54-L_S56_ME_L001__  dropped: 1\n",
      "OC54-L_S54_ME_L001__  candidates num: 54\n",
      "OC54-L_S54_ME_L001__  dropped: 9\n",
      "OT1-L ---------------------------------------------------\n",
      "metatable shape: (640, 32)\n",
      "sample file names: ['MN-OT1-L_S152_ME_L001__', 'OT1-L_S150_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 251   including test: 0\n",
      "somatic and quasi-somatic: (389, 32)   including test: 0 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__ 159   test: 0\n",
      "OT1-L_S150_ME_L001__ 230   test: 0\n",
      "overall num of somatic insertions:  389 \n",
      "\n",
      "MN-OT1-L_S152_ME_L001__  candidates num: 13\n",
      "MN-OT1-L_S152_ME_L001__  dropped: 3\n",
      "OT1-L_S150_ME_L001__  candidates num: 25\n",
      "OT1-L_S150_ME_L001__  dropped: 6\n",
      "PC12-L ---------------------------------------------------\n",
      "metatable shape: (1568, 32)\n",
      "sample file names: ['MN-PC12-L_S24_ME_L001__', 'PC12-L_S22_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 214   including test: 0\n",
      "somatic and quasi-somatic: (1354, 32)   including test: 0 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__ 434   test: 0\n",
      "PC12-L_S22_ME_L001__ 920   test: 0\n",
      "overall num of somatic insertions:  1354 \n",
      "\n",
      "MN-PC12-L_S24_ME_L001__  candidates num: 13\n",
      "MN-PC12-L_S24_ME_L001__  dropped: 0\n",
      "PC12-L_S22_ME_L001__  candidates num: 90\n",
      "PC12-L_S22_ME_L001__  dropped: 9\n",
      "PC14-L ---------------------------------------------------\n",
      "metatable shape: (857, 32)\n",
      "sample file names: ['MN-PC14-L_S36_ME_L001__', 'PC14-L_S34_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 188   including test: 0\n",
      "somatic and quasi-somatic: (669, 32)   including test: 0 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__ 100   test: 0\n",
      "PC14-L_S34_ME_L001__ 569   test: 0\n",
      "overall num of somatic insertions:  669 \n",
      "\n",
      "MN-PC14-L_S36_ME_L001__  candidates num: 8\n",
      "MN-PC14-L_S36_ME_L001__  dropped: 3\n",
      "PC14-L_S34_ME_L001__  candidates num: 68\n",
      "PC14-L_S34_ME_L001__  dropped: 7\n",
      "PC21-L ---------------------------------------------------\n",
      "metatable shape: (1071, 32)\n",
      "sample file names: ['MN-PC21-L_S20_ME_L001__', 'PC21-L_S18_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 337   including test: 0\n",
      "somatic and quasi-somatic: (734, 32)   including test: 0 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__ 578   test: 0\n",
      "PC21-L_S18_ME_L001__ 156   test: 0\n",
      "overall num of somatic insertions:  734 \n",
      "\n",
      "MN-PC21-L_S20_ME_L001__  candidates num: 30\n",
      "MN-PC21-L_S20_ME_L001__  dropped: 2\n",
      "PC21-L_S18_ME_L001__  candidates num: 25\n",
      "PC21-L_S18_ME_L001__  dropped: 4\n",
      "PC32-L ---------------------------------------------------\n",
      "metatable shape: (2116, 32)\n",
      "sample file names: ['MN-PC32-L_S32_ME_L001__', 'PC32-L_S30_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 110   including test: 0\n",
      "somatic and quasi-somatic: (2006, 32)   including test: 0 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__ 1803   test: 0\n",
      "PC32-L_S30_ME_L001__ 203   test: 0\n",
      "overall num of somatic insertions:  2006 \n",
      "\n",
      "MN-PC32-L_S32_ME_L001__  candidates num: 155\n",
      "MN-PC32-L_S32_ME_L001__  dropped: 58\n",
      "PC32-L_S30_ME_L001__  candidates num: 15\n",
      "PC32-L_S30_ME_L001__  dropped: 6\n",
      "SkC6-L ---------------------------------------------------\n",
      "metatable shape: (284, 32)\n",
      "sample file names: ['MN-SkC6-L_S60_ME_L001__', 'SkC6-L_S58_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 73   including test: 0\n",
      "somatic and quasi-somatic: (211, 32)   including test: 0 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__ 88   test: 0\n",
      "SkC6-L_S58_ME_L001__ 123   test: 0\n",
      "overall num of somatic insertions:  211 \n",
      "\n",
      "MN-SkC6-L_S60_ME_L001__  candidates num: 11\n",
      "MN-SkC6-L_S60_ME_L001__  dropped: 1\n",
      "SkC6-L_S58_ME_L001__  candidates num: 11\n",
      "SkC6-L_S58_ME_L001__  dropped: 4\n",
      "StS1-L ---------------------------------------------------\n",
      "metatable shape: (1234, 32)\n",
      "sample file names: ['MN-StS1-L_S92_ME_L001__', 'StS1-L_S90_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 448   including test: 0\n",
      "somatic and quasi-somatic: (786, 32)   including test: 0 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__ 330   test: 0\n",
      "StS1-L_S90_ME_L001__ 456   test: 0\n",
      "overall num of somatic insertions:  786 \n",
      "\n",
      "MN-StS1-L_S92_ME_L001__  candidates num: 39\n",
      "MN-StS1-L_S92_ME_L001__  dropped: 6\n",
      "StS1-L_S90_ME_L001__  candidates num: 48\n",
      "StS1-L_S90_ME_L001__  dropped: 9\n",
      "UC2-L ---------------------------------------------------\n",
      "metatable shape: (832, 32)\n",
      "sample file names: ['MN-UC2-L_S124_ME_L001__', 'UC2-L_S122_ME_L001__']\n",
      "null occurrences: 0\n",
      "all not somatic: 315   including test: 0\n",
      "somatic and quasi-somatic: (517, 32)   including test: 0 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__ 199   test: 0\n",
      "UC2-L_S122_ME_L001__ 318   test: 0\n",
      "overall num of somatic insertions:  517 \n",
      "\n",
      "MN-UC2-L_S124_ME_L001__  candidates num: 19\n",
      "MN-UC2-L_S124_ME_L001__  dropped: 3\n",
      "UC2-L_S122_ME_L001__  candidates num: 32\n",
      "UC2-L_S122_ME_L001__  dropped: 6\n"
     ]
    }
   ],
   "source": [
    "DIR = '/mnt/c/Users/Komar/Lab/OutPutDir/LINE/GGCC/'\n",
    "os.chdir(DIR)\n",
    "\n",
    "dirs = os.listdir()\n",
    "for dirname in dirs:\n",
    "    print(dirname, '---------------------------------------------------')\n",
    "    curr_dir = DIR + dirname\n",
    "    if not os.path.isdir(curr_dir):\n",
    "        continue\n",
    "    os.chdir(curr_dir)\n",
    "    \n",
    "    if not os.path.exists(curr_dir + '/result'):\n",
    "        os.makedirs(curr_dir + '/result')\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir)\n",
    "    # drop\n",
    "    drop_questionable(curr_dir, file_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea975fec-b6ac-439a-9eb9-67904652447c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Tumor norm matches "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "0620e0b4-2fed-48d3-937e-7cecd37d1aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# вставки в раке Алю1 - вставки в норме Рса; \n",
    "# Вставки в раке Рса - вставки в норме Алю1, для того же пациента\n",
    "def check_matches_tumor_mn(curr_dir1, curr_dir2, file_names, dirname, filter, del_matches):\n",
    "\n",
    "    num_matched = [dirname]\n",
    "\n",
    "    if file_names[0].find('MN-') != -1:\n",
    "        tmp = str(file_names[1])\n",
    "        file_names[1] = str(file_names[0])\n",
    "        file_names[0] = str(tmp)\n",
    "    # 0 - tumor, 1 - matched norm\n",
    "    \n",
    "    filtration = ''\n",
    "    if not filter:\n",
    "        filtration = '/nofiltration'\n",
    "    #\n",
    "    df_tumor_Alu1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{file_names[0]}_result.csv', sep=',')\n",
    "    df_tumor_Rsa1 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{file_names[0]}_result.csv', sep=',')\n",
    "    df_norm_Alu1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{file_names[1]}_result.csv', sep=',')\n",
    "    df_norm_Rsa1 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{file_names[1]}_result.csv', sep=',')\n",
    "\n",
    "    df_t_Alu1_mn_Rsa1 = pd.DataFrame(columns=df_tumor_Alu1.columns)\n",
    "    for i, row1 in df_tumor_Alu1.iterrows():\n",
    "        for j, row2 in df_norm_Rsa1.iterrows():\n",
    "            if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= 10):\n",
    "                if del_matches:\n",
    "                    df_tumor_Alu1.drop(i, axis=0, inplace=True)\n",
    "                    df_norm_Rsa1.drop(j, axis=0, inplace=True)\n",
    "                else:\n",
    "                    df_t_Alu1_mn_Rsa1.loc[len(df_t_Alu1_mn_Rsa1.index)] = row1\n",
    "\n",
    "    df_t_Rsa1_mn_Alu1 = pd.DataFrame(columns=df_tumor_Alu1.columns)\n",
    "    for i, row1 in df_tumor_Rsa1.iterrows():\n",
    "        for j, row2 in df_norm_Alu1.iterrows():\n",
    "            if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= 10):\n",
    "                if del_matches:\n",
    "                    df_tumor_Rsa1.drop(i, axis=0, inplace=True)\n",
    "                    df_norm_Alu1.drop(j, axis=0, inplace=True)\n",
    "                else:\n",
    "                    df_t_Rsa1_mn_Alu1.loc[len(df_t_Rsa1_mn_Alu1.index)] = row1\n",
    "\n",
    "    if del_matches:\n",
    "        df_tumor_Alu1.to_csv(curr_dir1 + filtration + '/result/' + f'{file_names[0]}_result.csv', index=False)\n",
    "        df_tumor_Rsa1.to_csv(curr_dir2 + filtration + '/result/' + f'{file_names[0]}_result.csv', index=False)\n",
    "        df_norm_Alu1.to_csv(curr_dir1 + filtration + '/result/' + f'{file_names[1]}_result.csv', index=False)\n",
    "        df_norm_Rsa1.to_csv(curr_dir2 + filtration + '/result/' + f'{file_names[1]}_result.csv', index=False)\n",
    "    \n",
    "    num_matched.append(len(df_t_Alu1_mn_Rsa1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_t_Rsa1_mn_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_tumor_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_tumor_Rsa1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_norm_Alu1['QUERY'].unique()))\n",
    "    num_matched.append(len(df_norm_Rsa1['QUERY'].unique()))\n",
    "    \n",
    "    return num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "fabd6f39-47e3-480c-8ee3-454054513023",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>TUMOR_Alu1_MN_Rsa1</th>\n",
       "      <th>TUMOR_Rsa1_MN_Alu1</th>\n",
       "      <th>TUMOR_Alu1</th>\n",
       "      <th>TUMOR_Rsa1</th>\n",
       "      <th>MN_Alu1</th>\n",
       "      <th>MN_Rsa1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BC60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BC66</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>33</td>\n",
       "      <td>47</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC70</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>36</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>33</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CC107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CC127-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>39</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CC18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CC46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>35</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CC82</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>44</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CC94</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CC96</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CC99</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GC39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GC41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LS6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>52</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OC14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OC33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OC40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>35</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OC42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OC48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OC50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OC54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OT1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PC12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>41</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PC14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PC21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PC32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SkC6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>StS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>UC2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NAME  TUMOR_Alu1_MN_Rsa1  ...  MN_Alu1  MN_Rsa1\n",
       "0      BC60                   0  ...       38       42\n",
       "1      BC66                   0  ...       47       41\n",
       "2      BC70                   0  ...       36       21\n",
       "3      BLC1                   0  ...       12       11\n",
       "4     CC107                   0  ...       11       11\n",
       "5   CC127-2                   0  ...       26       22\n",
       "6      CC18                   0  ...       21       21\n",
       "7      CC46                   0  ...       19       12\n",
       "8       CC5                   0  ...       16        3\n",
       "9      CC82                   0  ...        9        5\n",
       "10     CC94                   0  ...        4        3\n",
       "11     CC96                   0  ...       20       21\n",
       "12     CC99                   0  ...        9       13\n",
       "13     GC39                   0  ...       14       13\n",
       "14     GC41                   0  ...       10        4\n",
       "15      LS1                   0  ...        6        6\n",
       "16      LS6                   0  ...       52       18\n",
       "17      M14                   0  ...       25       23\n",
       "18      M21                   0  ...       15       16\n",
       "19      M22                   0  ...       11        3\n",
       "20      NC5                   0  ...        9       11\n",
       "21     OC14                   0  ...       14       18\n",
       "22     OC33                   0  ...        9       14\n",
       "23     OC40                   0  ...       15        6\n",
       "24     OC42                   0  ...        1       10\n",
       "25     OC48                   0  ...       14       13\n",
       "26     OC50                   0  ...        9        5\n",
       "27     OC54                   0  ...        6        7\n",
       "28      OT1                   0  ...       16       13\n",
       "29     PC12                   0  ...        8        8\n",
       "30     PC14                   0  ...        3        3\n",
       "31     PC21                   0  ...       11       15\n",
       "32     PC32                   0  ...        2       14\n",
       "33     SkC6                   0  ...       21       19\n",
       "34     StS1                   0  ...        8       12\n",
       "35      UC2                   0  ...        8       14\n",
       "\n",
       "[36 rows x 7 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIR1 = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/Alu_AGCT/'\n",
    "DIR2 = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/Alu_GTAC/'\n",
    "SAVE_TO = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/'\n",
    "FILTER = True\n",
    "DEL_MATCHES = True\n",
    "\n",
    "dirs = os.listdir(DIR1)\n",
    "num_matched = pd.DataFrame(columns=['NAME', 'TUMOR_Alu1_MN_Rsa1', 'TUMOR_Rsa1_MN_Alu1', 'TUMOR_Alu1', 'TUMOR_Rsa1', 'MN_Alu1', 'MN_Rsa1'])\n",
    "for dirname in dirs:\n",
    "\n",
    "    curr_dir1 = DIR1 + dirname\n",
    "    curr_dir2 = DIR2 + dirname\n",
    "    \n",
    "    if not os.path.isdir(curr_dir1):\n",
    "        continue\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir1, show=False)\n",
    "    # tumor norm matches \n",
    "    num_matched.loc[len(num_matched.index)] = check_matches_tumor_mn(curr_dir1, curr_dir2, file_names, dirname, FILTER, DEL_MATCHES)\n",
    "    \n",
    "if FILTER:\n",
    "    num_matched.to_csv(SAVE_TO + f'matches_tumor_mn_filtr.csv')\n",
    "else:\n",
    "    num_matched.to_csv(SAVE_TO + f'matches_tumor_mn_nofiltr.csv')\n",
    "\n",
    "num_matched"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa3f20e6-cd1e-4945-a682-62566106ec39",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Get values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "632b9f28-18c9-4067-925e-53a1e0efb32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coords(curr_dir1, curr_dir2, file_names, save_to):\n",
    "    \n",
    "    for name in file_names:\n",
    "        df = pd.read_table(curr_dir1 + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        coord_df1 = df[['QUERY', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', f'{name}_NUM_READS']].copy(deep=True)\n",
    "        coord_df1.columns = ['READ_NAME', 'INS_CHR', 'INS_COORD', 'INS_STRAND', 'NUM_READS']\n",
    "        coord_df1['RESTRICTASE'] = 'Alu1'\n",
    "\n",
    "        df = pd.read_table(curr_dir2 + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        coord_df2 = df[['QUERY', 'INS_CHR_B', 'INS_COORD_B', 'INS_STRAND_B', f'{name}_NUM_READS']].copy(deep=True)\n",
    "        coord_df2.columns = ['READ_NAME', 'INS_CHR', 'INS_COORD', 'INS_STRAND', 'NUM_READS']\n",
    "        coord_df2['RESTRICTASE'] = 'Rsa1'\n",
    "        \n",
    "        df_to_save = pd.concat([coord_df1, coord_df2], ignore_index=True).drop_duplicates()\n",
    "        df_to_save.to_csv(save_to + f'{name}_coords.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "aa059d51-7635-4485-9056-151f8fe18e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_matched_coords(curr_dir1, curr_dir2, file_names, dirname, filter):\n",
    "\n",
    "    num_matched = [dirname]\n",
    "    \n",
    "    for name in file_names:\n",
    "        filtration = ''\n",
    "        if not filter:\n",
    "            filtration = '/nofiltration'\n",
    "        df1 = pd.read_table(curr_dir1 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "        df2 = pd.read_table(curr_dir2 + filtration + '/result/' + f'{name}_result.csv', sep=',')\n",
    "\n",
    "        df_to_save = pd.DataFrame(columns=df2.columns)\n",
    "        \n",
    "        for i, row1 in df1.iterrows():\n",
    "            for j, row2 in df2.iterrows():\n",
    "                if (row1['INS_CHR_B'] == row2['INS_CHR_B']) and (abs(row1['INS_COORD_B'] - row2['INS_COORD_B']) <= 10):\n",
    "                    df_to_save.loc[len(df_to_save.index)] = row1\n",
    "\n",
    "        num_matched.append(len(df_to_save['QUERY'].unique()))\n",
    "        num_matched.append(len(df1['QUERY'].unique()))\n",
    "        num_matched.append(len(df2['QUERY'].unique()))\n",
    "    \n",
    "    return num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "521dc870-20c3-46ed-b97c-29065ccdecd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>MATCHED</th>\n",
       "      <th>Alu1</th>\n",
       "      <th>Rsa1</th>\n",
       "      <th>MN_MATCHED</th>\n",
       "      <th>MN_Alu1</th>\n",
       "      <th>MN_Rsa1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BC60</td>\n",
       "      <td>11</td>\n",
       "      <td>36</td>\n",
       "      <td>30</td>\n",
       "      <td>12</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BC66</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>47</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC70</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLC1</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CC107</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CC127-2</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CC18</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CC46</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC5</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CC82</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CC94</td>\n",
       "      <td>17</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CC96</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CC99</td>\n",
       "      <td>4</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GC39</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GC41</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LS1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LS6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>52</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M14</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M21</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M22</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC5</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OC14</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OC33</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OC40</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OC42</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>48</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OC48</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OC50</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OC54</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OT1</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PC12</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PC14</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>37</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PC21</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PC32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>92</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SkC6</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>StS1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>UC2</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NAME  MATCHED  Alu1  Rsa1  MN_MATCHED  MN_Alu1  MN_Rsa1\n",
       "0      BC60       11    36    30          12       38       42\n",
       "1      BC66        8    31    33          13       47       41\n",
       "2      BC70        0     5    11           5       36       21\n",
       "3      BLC1        6    29    33           0       12       11\n",
       "4     CC107        3    28    18           0       11       11\n",
       "5   CC127-2        3    28    39           0       26       22\n",
       "6      CC18        1    11    13           1       21       21\n",
       "7      CC46        0     4    11           1       19       12\n",
       "8       CC5        3    23    35           0       16        3\n",
       "9      CC82        6    34    44           0        9        5\n",
       "10     CC94       17    52    48           1        4        3\n",
       "11     CC96        0    11     7           1       20       21\n",
       "12     CC99        4    21    17           1        9       13\n",
       "13     GC39        0    11    12           0       14       13\n",
       "14     GC41        1    13    13           0       10        4\n",
       "15      LS1        0    23    12           1        6        6\n",
       "16      LS6        0     0    21           5       52       18\n",
       "17      M14        0     9     6           4       25       23\n",
       "18      M21        0     8    15           2       15       16\n",
       "19      M22        0     8     5           1       11        3\n",
       "20      NC5        0     9    11           2       19       21\n",
       "21     OC14        1    14    18           1        9        9\n",
       "22     OC33        1     9    14           2       41       14\n",
       "23     OC40        0    15     6           5       36       35\n",
       "24     OC42        0     1    10           3       48        9\n",
       "25     OC48        1    14    13           0        9        7\n",
       "26     OC50        0     9     5           4       18       25\n",
       "27     OC54        0     6     7           1       19       19\n",
       "28      OT1        1    16    13           1       21       16\n",
       "29     PC12        2     8     8           6       34       41\n",
       "30     PC14        1     3     3          10       37       46\n",
       "31     PC21        1    11    15           1        8       10\n",
       "32     PC32        0     2    14           4       92       23\n",
       "33     SkC6        2    21    19           1       12       21\n",
       "34     StS1        0     8    12           1       20       16\n",
       "35      UC2        0     8    14           0       14       14"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIR1 = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/Alu_AGCT/'\n",
    "DIR2 = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/Alu_GTAC/'\n",
    "SAVE_TO = '/mnt/c/Users/Komar/Lab/OutPutDir/experimental/'\n",
    "filter = True\n",
    "\n",
    "dirs = os.listdir(DIR1)\n",
    "num_matched = pd.DataFrame(columns=['NAME', 'MATCHED', 'Alu1', 'Rsa1', 'MN_MATCHED', 'MN_Alu1', 'MN_Rsa1'])\n",
    "for dirname in dirs:\n",
    "\n",
    "    curr_dir1 = DIR1 + dirname\n",
    "    curr_dir2 = DIR2 + dirname\n",
    "    \n",
    "    if not os.path.isdir(curr_dir1):\n",
    "        continue\n",
    "    \n",
    "    # 1 Get somatic insertions\n",
    "    file_names, df_somatic_list = get_somatic(curr_dir1, show=False)\n",
    "    # get coords for Maria\n",
    "    get_coords(curr_dir1, curr_dir2, file_names, SAVE_TO)\n",
    "    # check\n",
    "    num_matched.loc[len(num_matched.index)] = check_matched_coords(curr_dir1, curr_dir2, file_names, dirname, filter)\n",
    "    \n",
    "if filter:\n",
    "    num_matched.to_csv(SAVE_TO + f'matched_ins_filtr.csv')\n",
    "else:\n",
    "    num_matched.to_csv(SAVE_TO + f'matched_ins_nofiltr.csv')\n",
    "num_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53ca3f94-5676-4bf8-bb72-aea3a4235527",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
